---
title: '*telomers and centromers definition'
output: html_document
---

# New approach: Contigs

```{bash}
# We are considering using the contigs instead of the scaffolds to determine whether some regions are or are not telomers-centromers.

# Fisrt make a folder:

mkdir /home/mlucena/telomers_centromers_definition
cd /home/mlucena/telomers_centromers_definition

# We count the lines of the cat file:
cat /home/fabascal/FEATURES/tel10m.bed  /home/fabascal/FEATURES/tel2m.bed | wc -l
2640677

# Now create the file with the positions of telo2 plus telo10, as we know that telo2 goes from 0 to 2 and telo10 goes from 2 to 10.
# We also used bedtools to merge in case that there's some overlapping.
cat /home/fabascal/FEATURES/tel10m.bed  /home/fabascal/FEATURES/tel2m.bed | sort - -k1,1 -k2,2n | bedtools merge > tel0-10_merged.bed 

# It seems that there was, so we will use this merged file
wc -l tel0-10_merged.bed 
2628133 tel0-10_merged.bed 

# We will also merged the centromers file.

cat /home/fabascal/FEATURES/centr.bed  | sort - -k1,1 -k2,2n | bedtools merge > centr_merged.bed

wc -l /home/fabascal/FEATURES/centr.bed 
2211773 /home/fabascal/FEATURES/centr.bed
wc -l centr_merged.bed 
2200105 centr_merged.bed


# Check the overlapping with the contigs:

# First!

# We have 177602 positions covered by contigs. 
awk '{sum+=($3-$2)} END {print sum }' lp23contigs.bed 
177602
?¿?¿?¿?¿?¿?¿?

bedtools intersect -a /home/mlucena/grupolince/reference_genomes/lynx_pardinus_genome/lp23contigs.bed -b centr_merged.bed  -wa -wb > contig_centromers 

# Con esto sólo nos quedaríamos con un contig, el lp23.s27575c001 

# Hago lo mismo para telómeros. 

bedtools intersect -a /home/mlucena/grupolince/reference_genomes/lynx_pardinus_genome/lp23contigs.bed -b tel0-10_merged.bed   -wa -wb > contig_telomers

# Comprobamos cuantos contigs se nos quedarían en cada caso:

awk '{a[$4]+=($7-$6); b[$4]=($3-$2)} END {for(i in a) print i,a[i], b[i]}' contig_telomers | awk '{if (($2/$3)>0.25) print $0,"YES"; else print $0,"NO"}'

awk '{a[$4]+=($7-$6); b[$4]=($3-$2)} END {for(i in a) print i,a[i], b[i]}' contig_centromers | awk '{if (($2/$3)>0.25) print $0,"YES"; else print $0,"NO"}'

# Se nos uedan sólo dos contigs de telómeros y uno de centrómeros. Tengo que comprobar si está bien el archivo de contigs. Si es así, quizás no podamos aplicar estos filtros. 

```


# Old approach: Scaffolds. 

## Telomers.sh

```{bash}
# I extract a list of the scaffolds that have at least some bases in the subtelomeric region or in the centromere.

awk '{$1}' tel2m.bed | sort | uniq > scaffolds_tel2m.list
awk '{$1}' tel10m.bed | sort | uniq > scaffolds_tel10m.list
awk '{$1}' centr.bed | sort | uniq > scaffolds_centr.list

# How long is the longest scaffold?

join scaffolds_tel2m.list Length_scaffolds_lp23 | sort -k2n | tail

# lp23.s20863 3296142
# lp23.s31290 3370745
# lp23.s31287 3871840
# lp23.s00008 4038649
# lp23.s36495 4262991
# lp23.s00006 4480212
# lp23.s36493 4643072
# lp23.s26068 4986501
# lp23.s36488 7826823
# lp23.s31278 7976960

# I has 7.9M, and that is not possible as I am looking at a region of 2M.


join scaffolds_tel10m.list Length_scaffolds_lp23 | sort -k2n | tail

# lp23.s10428 5537880
# lp23.s00003 5556527
# lp23.s05215 5566334
# lp23.s15639 5809641
# lp23.s20852 6074816
# lp23.s05214 6149468
# lp23.s10426 6347329
# lp23.s20851 6824652
# lp23.s26065 6924262
# lp23.s26064 8029254



join scaffolds_centr.list Length_scaffolds_lp23 | sort -k2n | tail

# lp23.s20852 6074816
# lp23.s00002 6085556
# lp23.s10426 6347329
# lp23.s20851 6824652
# lp23.s26065 6924262
# lp23.s36488 7826823
# lp23.s31278 7976960
# lp23.s15637 8424978
# lp23.s05213 12220470
# lp23.s00001 13188378
 

# It looks like some scaffolds have just some bp homologous. 

# But before that...

# Is there a scaffold present in the centromere and the subtelomeric region?

grep -Fxf scaffolds_tel2m.list scaffolds_centr.list  > scaffolds_present_in_tel2_and_centr.list
# Yes! 41.

grep -Fxf scaffolds_tel10m.list scaffolds_centr.list  > scaffolds_present_in_tel10_and_centr.list
# Yes! 107.

# Are all the scaffolds present in tel2m in tel10m?

wc -l scaffolds_tel2m.list
# 575 scaffolds_tel2m.list

wc -l scaffolds_tel10m.list
# 1161 scaffolds_tel10m.list

grep -Fxf scaffolds_tel10m.list scaffolds_tel2m.list | wc -l
# No! only 54


# I asked Fede why and he mentioned that tel10m include subtelomeric region from 2 to 10Mb, therefore, from now on, we will include tel0-10 in our analysis. 

cat scaffolds_tel2m.list scaffolds_tel10m.list | sort | uniq > scaffolds_tel0-10m.list

wc -l scaffolds_tel0-10m.list 
# 1682 scaffolds_tel0-10m.list

join scaffolds_tel0-10m.list  Length_scaffolds_lp23 | sort -k2n | tail

# lp23.s05215 5566334
# lp23.s15639 5809641
# lp23.s20852 6074816
# lp23.s05214 6149468
# lp23.s10426 6347329
# lp23.s20851 6824652
# lp23.s26065 6924262
# lp23.s36488 7826823
# lp23.s31278 7976960
# lp23.s26064 8029254

#########################################

# Definimos tres niveles:

# 1. Scaffolds y longitud de los mismos: ya lo tenemos : Length_scaffolds_lp23 y scaffolds_tel2m.list/tel0-10m.list/centr.list
 
# 2. Región con homología:

## En este caso determinamos la región potencialmente subtelomérica:

sort -k1,1 -k2,2n /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed | bedtools merge -d 7976960 | awk -v OFS='\t'  '{print $1, $3-$2}' > tel2m_merged.bed
cat /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed /home/mlucena/grupolince/copia_fabascal/FEATURES/tel10m.bed | sort - -k1,1 -k2,2n | bedtools merge -d 8029254 | awk -v OFS='\t'  '{print $1, $3-$2}' > tel0-10m_merged.bed
sort -k1,1 -k2,2n /home/mlucena/grupolince/copia_fabascal/FEATURES/centr.bed | bedtools merge -d 13188378 | awk -v OFS='\t'  '{print $1, $3-$2}' > centr_merged.bed


# 3. Bases con homología. 

awk '{print $1, $3-$2}' /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed | awk '{a[$1]+=$2}END{for(i in a) print i,a[i]}' | sort -k1 | join - tel2m_merged.bed | sort -k1 | join - Length_scaffolds_lp23 | sort -k1 > tel2m_bp_region_length.info 
cat /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed /home/mlucena/grupolince/copia_fabascal/FEATURES/tel10m.bed | awk '{print $1, $3-$2}' | awk '{a[$1]+=$2}END{for(i in a) print i,a[i]}' | sort -k1 | join - tel0-10m_merged.bed | sort -k1 | join - Length_scaffolds_lp23 | sort -k1 > tel0-10m_bp_region_length.info 
awk '{print $1, $3-$2}' /home/mlucena/grupolince/copia_fabascal/FEATURES/centr.bed | awk '{a[$1]+=$2}END{for(i in a) print i,a[i]}' | sort -k1 | join - centr_merged.bed | sort -k1 | join - Length_scaffolds_lp23 | sort -k1 > centr_bp_region_length.info 


# En este último archivo ya tenemos todo:
# -------------------------------- Scaffold
#     -----------------			   Región
#     ..     .   ..   .			   Bases con homología.


# ---------------------------------------------------------------------------------#

# Con esta información exploramos un poquito los datos:

head scaffolds_present_in_tel2_and_centr.list
lp23.s00008
lp23.s00224
lp23.s00387
lp23.s10464

######################################

# Example: lp23.s00008: Aparece en ambos, tel2m y centr. 

grep lp23.s00008 tel2m_bp_region_length.info 
lp23.s00008 32 220 4038649

grep lp23.s00008 centr_bp_region_length.info 
lp23.s00008 7 2630019 4038649

grep lp23.s00008 /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed 
lp23.s00008	2579970	2579993	chrF1:67576422
lp23.s00008	2580181	2580190	chrF1:67576225

grep lp23.s00008 /home/mlucena/grupolince/copia_fabascal/FEATURES/centr.bed 
lp23.s00008	531713	531715	chrA3:40467745
lp23.s00008	531831	531834	chrA3:40467629
lp23.s00008	3161730	3161732	chrC2:65834393

# Now we try to estimate % of the bp that have sinteny in the region and in the whole scaffold --> código para hacer histograma. 

# awk '{print $1, $2, $3, $4, ($2/$3)*100, ($3/$4)*100 ,($2/$4)*100}' centr_bp_region_length.info | awk '{if ($7==0) print $0}' | sort -k7,7 -d | head
awk '{print $1, $2, $3, $4, ($2/$3)*100, ($3/$4)*100 ,($2/$4)*100}' centr_bp_region_length.info | awk '{print $7}' | sort -g | awk '{split ($1,a, "."); print a[1]}' | uniq -c | awk '{print $2, $1}' | perl -pe 's/ (\d+)$/"="x$1/e'
awk '{print $1, $2, $3, $4, ($2/$3)*100, ($3/$4)*100 ,($2/$4)*100}' tel0-10m_bp_region_length.info | awk '{print $7}' | sort -g | awk '{split ($1,a, "."); print a[1]}' | uniq -c | awk '{print $2, $1}' | perl -pe 's/ (\d+)$/"="x$1/e'
awk '{print $1, $2, $3, $4, ($2/$3)*100, ($3/$4)*100 ,($2/$4)*100}' tel2_bp_region_length.info | awk '{print $5}' | sort -g | awk '{split ($1,a, "."); print a[1]}' | uniq -c | awk '{print $2, $1}' | perl -pe 's/ (\d+)$/"="x$1/e'


# Tras hablar con Fede check email (21/02/2017) voy a hacer un merge con dos filtros -d 1000 y -d 5000, a ver cual funciona mejor. Posteriormente voy a hacer un histograma y luego decidir un filtro de numero de bases del scaffold. 


## Aquí determinamos regiones separadas por un máximo de 1000pb y posteriormente las filtramos para número de bases 1000 y 5000. 

sort -k1,1 -k2,2n /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed | bedtools merge -d 1000  > tel2m_merged_1000.bed
cat /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed /home/mlucena/grupolince/copia_fabascal/FEATURES/tel10m.bed | sort - -k1,1 -k2,2n  | bedtools merge -d 1000 > tel0-10m_merged_1000.bed
sort -k1,1 -k2,2n /home/mlucena/grupolince/copia_fabascal/FEATURES/centr.bed | bedtools merge -d 1000  > centr_merged_1000.bed

sort -k1,1 -k2,2n /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed | bedtools merge -d 5000  > tel2m_merged_5000.bed
cat /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed /home/mlucena/grupolince/copia_fabascal/FEATURES/tel10m.bed | sort - -k1,1 -k2,2n  | bedtools merge -d 5000 > tel0-10m_merged_5000.bed
sort -k1,1 -k2,2n /home/mlucena/grupolince/copia_fabascal/FEATURES/centr.bed | bedtools merge -d 5000  > centr_merged_5000.bed

sort -k1,1 -k2,2n /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed | bedtools merge -d 500  > tel2m_merged_500.bed
cat /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed /home/mlucena/grupolince/copia_fabascal/FEATURES/tel10m.bed | sort - -k1,1 -k2,2n  | bedtools merge -d 500 > tel0-10m_merged_500.bed
sort -k1,1 -k2,2n /home/mlucena/grupolince/copia_fabascal/FEATURES/centr.bed | bedtools merge -d 500  > centr_merged_500.bed


########

cat /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed /home/mlucena/grupolince/copia_fabascal/FEATURES/tel10m.bed | wc -l
2640677

wc -l tel0-10m_merged_1000.bed
53329 tel0-10m_merged_1000.bed

########

wc -l /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed 
401564 /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed
[mlucena@genomics_a Lyp_annotation_Apr14_final]$ wc -l tel2m_merged_1000.bed
11801 tel2m_merged_1000.bed

########

cat /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed /home/mlucena/grupolince/copia_fabascal/FEATURES/tel10m.bed  | grep lp23.s05217

wc -l /home/mlucena/grupolince/copia_fabascal/FEATURES/centr.bed
2211773 

wc -l centr_merged_1000.bed
38532 centr_merged_1000.bed

# --> 385327 / 2211773 = 0.17

scp /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed .
cat /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed /home/mlucena/grupolince/copia_fabascal/FEATURES/tel10m.bed | sort - -k1,1 -k2,2n > tel0-10m.bed
scp /home/mlucena/grupolince/copia_fabascal/FEATURES/centr.bed .

## Correr desde aquí:

FILES=($(ls *merged_1000.bed | uniq))

rm *_density_regions.info

for FILE in "${FILES[@]}"
do
echo $FILE

while read SCAFFOLD START FINISH
do
echo ${FILE/_merged_1000.bed/.bed}
echo $SCAFFOLD
echo $START
echo $FINISH

NUM_BASES=$(awk -v SCAFFOLD=$SCAFFOLD -v START=$START -v FINISH=$FINISH '$1==SCAFFOLD && $2>=START && $3<=FINISH {print $1, $3-$2}' ${FILE/_merged_1000.bed/.bed} | awk -v OFS='\t' '{for(i=1;i<=NF;i++) {sum[i] += $i; sumsq[i] += ($i)^2}} END {for (i=1;i<=NF;i++) {printf  sum[i]}}')
paste \
<(echo $SCAFFOLD ) \
<(echo $START ) \
<(echo $FINISH ) \
<(echo $NUM_BASES ) | \
sed 's/ /\t/g'| sed 's/\t\+/\t/g' >>  ${FILE/.bed/_density_regions.info}

done < <( cat `echo $FILE` )
done


FILES=($(ls *_1000_density_regions.info | uniq))

for FILE in "${FILES[@]}"
do
echo $FILE
awk -v OFS='\t' '{print $1,$3-$2,$4,($4*100)/($3-$2)}' ${FILE} > ${FILE/.info/.info.counts}
done
##########

FILES=($(ls *merged_5000.bed | uniq))

rm *_density_regions.info

for FILE in "${FILES[@]}"
do
echo $FILE

while read SCAFFOLD START FINISH
do
echo ${FILE/_merged_5000.bed/.bed}
echo $SCAFFOLD
echo $START
echo $FINISH

NUM_BASES=$(awk -v SCAFFOLD=$SCAFFOLD -v START=$START -v FINISH=$FINISH '$1==SCAFFOLD && $2>=START && $3<=FINISH {print $1, $3-$2}' ${FILE/_merged_5000.bed/.bed} | awk -v OFS='\t' '{for(i=1;i<=NF;i++) {sum[i] += $i; sumsq[i] += ($i)^2}} END {for (i=1;i<=NF;i++) {printf  sum[i]}}')
paste \
<(echo $SCAFFOLD ) \
<(echo $START ) \
<(echo $FINISH ) \
<(echo $NUM_BASES ) | \
sed 's/ /\t/g'| sed 's/\t\+/\t/g' >>  ${FILE/.bed/_density_regions.info}

done < <( cat `echo $FILE` )
done


FILES=($(ls *_5000_density_regions.info | uniq))

for FILE in "${FILES[@]}"
do
echo $FILE
awk -v OFS='\t' '{print $1,$3-$2,$4,($4*100)/($3-$2)}' ${FILE} > ${FILE/.info/.info.counts}
done


##########
scp mlucena@genomics-a.ebd.csic.es:/home/mlucena/grupolince/Lynx_annotation/Lyp_annotation_Apr14_final/*density_regions.info.counts /Users/marialucenaperez/Dropbox/PhD/contemporary/subtelomeric_centromeric_analysis/
##########

FILES=($(ls *merged_500.bed | uniq))

rm *_density_regions.info

for FILE in "${FILES[@]}"
do
echo $FILE

while read SCAFFOLD START FINISH
do
echo ${FILE/_merged_500.bed/.bed}
echo $SCAFFOLD
echo $START
echo $FINISH

NUM_BASES=$(awk -v SCAFFOLD=$SCAFFOLD -v START=$START -v FINISH=$FINISH '$1==SCAFFOLD && $2>=START && $3<=FINISH {print $1, $3-$2}' ${FILE/_merged_500.bed/.bed} | awk -v OFS='\t' '{for(i=1;i<=NF;i++) {sum[i] += $i; sumsq[i] += ($i)^2}} END {for (i=1;i<=NF;i++) {printf  sum[i]}}')
paste \
<(echo $SCAFFOLD ) \
<(echo $START ) \
<(echo $FINISH ) \
<(echo $NUM_BASES ) | \
sed 's/ /\t/g'| sed 's/\t\+/\t/g' >>  ${FILE/.bed/_density_regions.info}

done < <( cat `echo $FILE` )
done


FILES=($(ls *_500_density_regions.info | uniq))

for FILE in "${FILES[@]}"
do
echo $FILE
awk -v OFS='\t' '{print $1,$3-$2,$4,($4*100)/($3-$2)}' ${FILE} > ${FILE/.info/.info.counts}
done

##########
scp mlucena@genomics-a.ebd.csic.es:/home/mlucena/grupolince/Lynx_annotation/Lyp_annotation_Apr14_final/*density_regions.info.counts /Users/marialucenaperez/Dropbox/PhD/contemporary/subtelomeric_centromeric_analysis/
##########

# -> Filtering:

awk '{}'tel2m_merged_1000.bed
awk -v OFS='\t' '{for(i=1;i<=NF;i++) {sum[i] += $i; sumsq[i] += ($i)^2}} END {for (i=1;i<=NF;i++) {printf ("%.10e\n", sqrt((sumsq[i]-sum[i]^2/NR)/NR ))}}'


## Aquí determinamos regiones separadas por un máximo de 5000pb y posteriormente las filtramos para número de bases 1000 y 2000. 

sort -k1,1 -k2,2n /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed | bedtools merge -d 5000  > tel2m_merged_5000.bed
cat /home/mlucena/grupolince/copia_fabascal/FEATURES/tel2m.bed /home/mlucena/grupolince/copia_fabascal/FEATURES/tel10m.bed | sort - -k1,1 -k2,2n |  bedtools merge -d 5000 > tel0-10m_merged_5000.bed
sort -k1,1 -k2,2n /home/mlucena/grupolince/copia_fabascal/FEATURES/centr.bed | bedtools merge -d 5000  > centr_merged_5000.bed

# -> Filtering:

```

## subtelomeric-centromeric_filtering.R

```{r, engine=bash, eval=FALSE}
library (ggplot2) 
library(dplyr)

df_tel2 <- read.delim ("/Users/marialucenaperez/Dropbox/PhD/contemporary/subtelomeric_centromeric_analysis/tel2m_bp_region_length.info", sep =" ", header=F) 

pdf ("/Users/marialucenaperez/Dropbox/PhD/contemporary/subtelomeric_centromeric_analysis/tel2.pdf")
ggplot(data=df_tel2,mapping = aes(x=V2))+
  geom_histogram(aes(y=cumsum(..count..)), binwidth = 100)+
 # stat_bin(aes(y=cumsum(..count..)),geom="line",color="green", binwidth= 50)+
  ggtitle("tel2m")
dev.off()

ggplot(data=df_tel2,mapping = aes(x=V2, y=V3))+
  geom_point()

df_tel0_10 <- read.delim ("/Users/marialucenaperez/Dropbox/PhD/contemporary/subtelomeric_centromeric_analysis/tel0-10m_bp_region_length.info", sep =" ", header=F) 

pdf ("/Users/marialucenaperez/Dropbox/PhD/contemporary/subtelomeric_centromeric_analysis/tel0-10.pdf")
ggplot(data=df_tel0_10,mapping = aes(x=V2))+
  geom_histogram(aes(y=cumsum(..count..)), binwidth = 100)+
#  stat_bin(aes(y=cumsum(..count..)),geom="line",color="green")+
  ggtitle("tel0-10m")
dev.off()

df_centr <- read.delim ("/Users/marialucenaperez/Dropbox/PhD/contemporary/subtelomeric_centromeric_analysis/centr_bp_region_length.info", sep =" ", header=F) 


pdf ("/Users/marialucenaperez/Dropbox/PhD/contemporary/subtelomeric_centromeric_analysis/centr.pdf")
ggplot(data=df_centr ,mapping = aes(x=V2))+
  geom_histogram(aes(y=cumsum(..count..)), binwidth = 100)+
#  stat_bin(aes(y=cumsum(..count..)),geom="line",color="green")+
  ggtitle("centr")
dev.off()

```


## subtelomeric_centromeric_histogram_representation

```{r}
library("ggplot2")
library ("ggExtra")

setwd ("/Users/marialucenaperez/Dropbox/PhD/contemporary/subtelomeric_centromeric_analysis/")
fins = list.files(pattern="*_density_regions.info.counts") 

for (i in 1:length(fins)) 
{
  dat <- read.csv (fins[i], sep = '\t',stringsAsFactors = FALSE, dec = ".",header=F)
  name <- unlist(strsplit(fins[i], "[.]"))
#  pdf(file = paste(name[1], '.pdf', sep=''))
#  print(ggplot (dat, aes(x=V4))+
#    geom_histogram(binwidth =1) +
#    ggtitle(label = name[1]))
#  dev.off()
  sp2 <- ggplot(dat,aes(x=V4,y=V2)) + geom_point(size=0.3) +  xlab("Density of nt in the region (%)") +  ylab("Length of the region") 
  plot <- ggMarginal(sp2, type = 'histogram', xparams = list(bins=200), yparams = list(bins=1000))
  pdf(file = paste(name[1], '_density.pdf', sep=''))
  # Marginal density plot
  print(plot)
  dev.off()
}


a<-ggMarginal (sp2, type = 'histogram', xparams = list(bins=200), yparams = list(bins=1000))
print(a)

```





