---
title: "14.GO_enrichment"
output: html_document
---



# Load library and required information

Este es el script donde realizo el GO term annotation & enrichment analysis y el pathway enrichment analysis. 

Lo primero que hago es pedirle a Godoy una lista donde cada gen tenga asociado su GO-term. El me pasa una carpeta donde hay varios archivos.

GO2LYNPC.txt
GO2LYPA.LYPA23C.APPRIS.fatigo.genes.txt
GO2LYPA.LYPA23C.APPRIS.fatigo.txt
GO_enrichment.docx
genes2transcript2go.list.txt
lp23.allgenes.fatigo.txt
lp23.allgenes.goterms
lp23.allgenes.id

En el archivo docx, se explica los pasos que siguió Godoy para hacer el enrichment. Lo que deduzco es que las que se llaman lp23* son listas antiguas, y despues están las GO2LYPA que sólo contienen las isoformas principales. 

Lo distintos archivos son:

GO2LYNPC.txt --> Tiene el nombre del gen, el GO_term, el GO_name y el GO_type; parece que incluye todas las isoformas.
GO2LYPA.LYPA23C.APPRIS.fatigo.genes.txt --> Creo q es el que tengo que usar. Nombre de gen, GO_ACC, GO_name y el GO_type.
GO2LYPA.LYPA23C.APPRIS.fatigo.txt --> Igual pero con proteinas.
GO_enrichment.docx --> Archivo docx donde se explica el procedimiento.
genes2transcript2go.list.txt --> Creo que es la carpeta que tiene el nombre de gen, isoforma principal y proteina y despues GO term y sus características.

Los lp23 son antiguos sin la isoforma principal.

Lo primero que voy a hacer es cargar librerias y nuestros archivos:

## Load information for GO annotation & enrichment

```{r}
library(topGO)
library (plyr)
library(tidyr)
library (dplyr)


# Working directory

wd <- "/Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/Lypa23c_GO_terms_KO_terms_annotation/"

# Tabla con la información de GO term para cada gen.

tabla_all_genes_GO_term_complete_info <- read.table (paste0(wd, "GO2LYPA.LYPA23C.APPRIS.fatigo.genes_minor_modif.txt"), sep="\t", header = T) %>% mutate (GO_ACC=as.character(GO_ACC)) %>% filter(GO_TYPE=="biological_process") %>% dplyr::select (-"GO_TYPE")
# El minor_modif tiene ' cambiadas por empty space pq si no R entiende que es un string. 


## Modifo el archivo para poder generar un gmt con los nombres de cada GO term, su descripción, y a que genes va asociado. 
# Este archivo lo vamos a necesitar para cargar los datos en CYTOSCAPE. 
tabla_all_genes_GO_term_complete_info_modif <- tabla_all_genes_GO_term_complete_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(GENE_NAME, collapse=','))  

# Guardo la tabla
write.table(tabla_all_genes_GO_term_complete_info_modif, paste0(wd, "GO2LYPA.LYPA23C.APPRIS.fatigo.genes_modif.NAMES.gmt"), quote = F, row.names = F, col.names = F, sep = "\t" )


# Modificaciones para topGo.
list_of_list_all_genes_GO_term <- dlply(tabla_all_genes_GO_term_complete_info, .(GENE_NAME), function(x)x$GO_ACC)
list_of_list_all_genes_GO_term_chr <- lapply(list_of_list_all_genes_GO_term, function(x) as.character(unlist(x)))

# Guardo los nombres de los genes en cuestión. 
geneNames <- names(list_of_list_all_genes_GO_term_chr)

```

## Load information for pathway enrichment

No hace falta hacer este chunk, pq solo se tiene que hacer una vez.

- Previous work

Voy a seleccionar los KO que son los nombres de proteinas comunes a todos


<!-- ```{bash} -->

<!-- cd /GRUPOS/grupolince/Lyp_annotation_Apr14_final -->

<!-- grep "ko_group" LYPA23C.all.fix.nr.gff3 > all_mRNA -->

<!-- # sanity check -->
<!-- # all ko groups -->
<!-- grep -o '\bko_group=\w*' all_mRNA | wc -l -->
<!-- #10213 -->
<!-- # All parent_id -->
<!-- grep -o '\bParent=\w*' all_mRNA | wc -l -->
<!-- #10213 -->

<!-- # Igual! perfecto, los podemos juntar -->

<!-- paste <(grep -o '\bParent=\w*' all_mRNA) <(grep -o '\bko_group=\w*' all_mRNA) | sed 's/Parent=//g' | sed 's/ko_group=//g' > genes_to_KO.txt -->

<!-- rm all_mRNA -->

<!-- ``` -->

<!-- Me lo bajo a mi carpeta -->

<!-- ```{bash} -->
<!-- scp mlucena@genomics-b.ebd.csic.es:/GRUPOS/grupolince/Lyp_annotation_Apr14_final/genes_to_KO.txt /Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/Lypa23c_GO_terms_KO_terms_annotation -->
<!-- ``` -->

- Load KO information for pathway enrichment

From reactome webpage: https://reactome.org/userguide/analysis
Entities:
If your data is a single column of identifiers such as UniProt IDs, gene symbols or ChEBI IDs, they are mapped to pathways and over-representation and pathway-topology analyses are run. Over-representation analysis is a statistical (hypergeometric distribution) test that determines whether certain Reactome pathways are over-represented (enriched) in the submitted data. It answers the question ‘Does my list contain more proteins for pathway X than would be expected by chance?’ This test produces a probability score, which is corrected for false discovery rate using the Benjamani-Hochberg method.

Reactions:
Pathway topology analysis considers the connectivity between molecules that is represented by the pathway steps (which we refer to as reactions) in the pathway. It groups all the molecules represented in each reaction as a pathway ‘unit’. If any of these molecules are represented in your query set, this is considered a match to that reaction. This may give a better indication of the proportion of the pathway that matches your data, rather than the number of molecules that are common between your data and the pathway. It may also indicate that your data matches the start, end or a particular branch of a pathway process. This test does not have a probability score.



```{r}
# Archivo de equivalencias KO con genes

genes_to_KO.list <- read.table("/Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/Lypa23c_GO_terms_KO_terms_annotation/genes_to_KO.txt", header=F)

```

# GO term annotation

Tengo tablas individuales y las quiero anotar con los GO term correspondientes


Lo uno a la tabla de GO term


```{r}
library("stringr")

wd <- "/Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/Lypa23c_GO_terms_KO_terms_annotation/"


c_ll_ki_n013_c_ll_no_n008 <- dplyr::inner_join(read.table (paste0(wd,"c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise", sep=""), header = T, stringsAsFactors = F) %>% gsub( "P", "", as.character(.$Gen) n)
                                               
                                               gsub("%", "", paste(LOAN$RATE))
                                                %>%  sub("P1$", "", .$Gen)
                                                lstrip(., char = "P") %>% head()
                                               
                                               
                                               
                                               tabla_all_genes_GO_term_complete_info, by=c("V1"="GENE_NAME")) 

c_ll_ki_n013_c_ll_no_n008.gProfile <-  c_ll_ki_n013_c_ll_no_n008 %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(V1, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008.gProfile <- c_ll_ki_n013_c_ll_no_n008.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()
write.table(c_ll_ki_n013_c_ll_no_n008.gProfile, paste0(wd, "c_ll_ki_n013_c_ll_no_n008.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )


######
c_ll_ki_n013_c_ll_po_n008 <- dplyr::inner_join(read.table (paste0(wd,"list_genes_of_interest_c_ll_ki_n013_c_ll_po_n008_high_diversity_ratio_1_watterson_trimmed.txt", sep="")), tabla_all_genes_GO_term_complete_info, by=c("V1"="GENE_NAME")) 

c_ll_ki_n013_c_ll_po_n008.gProfile<-  c_ll_ki_n013_c_ll_po_n008 %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(V1, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_po_n008.gProfile <- c_ll_ki_n013_c_ll_po_n008.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()
write.table(c_ll_ki_n013_c_ll_po_n008.gProfile, paste0(wd, "c_ll_ki_n013_c_ll_po_n008.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )


######

c_lp_sm_n019_c_lp_do_n012 <- dplyr::inner_join(read.table (paste0(wd,"list_genes_of_interest_c_lp_sm_n019_c_lp_do_n012_high_diversity_ratio_1_watterson_trimmed.txt", sep="")), tabla_all_genes_GO_term_complete_info, by=c("V1"="GENE_NAME")) 

c_lp_sm_n019_c_lp_do_n012.gProfile<-  c_lp_sm_n019_c_lp_do_n012 %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(V1, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_lp_sm_n019_c_lp_do_n012.gProfile <- c_lp_sm_n019_c_lp_do_n012.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()
write.table(c_lp_sm_n019_c_lp_do_n012.gProfile, paste0(wd, "c_lp_sm_n019_c_lp_do_n012.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

```






# Files of interest

```{r}

# Tabla de genes de interes

# Tenemos 4 listas. 

# 1. Aquellas unidades con ratios mayores de uno  comunes a las tres comparaciones (con dos especies).
# /Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/ANGSD/sfs/list_for_venn_lynx_lynx_lynx_pardinus_ratio_above_1_CDS.txt 

# 2. Aquellas unidades con ratios mayores de uno en la comparación de las dos poblaciones de lynx lynx. 
# /Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/ANGSD/sfs/list_for_venn_lynx_lynx_ratio_above_1_CDS.txt

# 3. Unidades con un ratio similar a 1 y diversidad alta comunes a las tres comparaciones (con dos especies).
# /Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/ANGSD/sfs/list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1.txt

# 4. Unidades con un ratio similar a 1 y diversidad alta en la comparación de las dos poblaciones de lynx lynx. 
# /Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/ANGSD/sfs/list_genes_of_interest_lynx_lynx_high_diversity_ratio_1.txt

```

Voy a procesarla para que esté disponible para que se la trague R más facil

```{bash}

cd /Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/ANGSD/sfs/

#tail -n+2 list_for_venn_lynx_lynx_lynx_pardinus_ratio_above_1_CDS.txt | cut -d"_" -f1 | sed 's/.\{2\}$//' | sort | uniq > list_for_venn_lynx_lynx_lynx_pardinus_ratio_above_1_CDS_trimmed_noduplicates.txt

#tail -n+2 list_for_venn_lynx_lynx_ratio_above_1_CDS.txt | cut -d"_" -f1 | sed 's/.\{2\}$//' | sort | uniq > list_for_venn_lynx_lynx_ratio_above_1_CDS_trimmed_noduplicates.txt

tail -n+2 list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1.txt | cut -d"_" -f1 | sed 's/.\{2\}$//' | sort | uniq > list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_trimmed_noduplicates.txt

tail -n+2 list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1.txt | cut -d"_" -f1 | sed 's/.\{2\}$//' | sort > list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_trimmed.txt


tail -n+2 list_genes_of_interest_lynx_lynx_high_diversity_ratio_1.txt | cut -d"_" -f1 | sed 's/.\{2\}$//' | sort | uniq > list_genes_of_interest_lynx_lynx_high_diversity_ratio_1_trimmed_noduplicates.txt

tail -n+2 list_genes_of_interest_lynx_lynx_high_diversity_ratio_1.txt | cut -d"_" -f1 | sed 's/.\{2\}$//' | sort > list_genes_of_interest_lynx_lynx_high_diversity_ratio_1_trimmed.txt


# Ahora separo los que me han salido con pi y los que me han salido con watterson. 

tail -n+2 list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_pairwise.txt | cut -d"_" -f1 | sed 's/.\{2\}$//' | sort | uniq > list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_pairwise_trimmed_noduplicates.txt

tail -n+2 list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_watterson.txt | cut -d"_" -f1 | sed 's/.\{2\}$//' | sort | uniq > list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_watterson_trimmed_noduplicates.txt

tail -n+2 list_genes_of_interest_lynx_lynx_high_diversity_ratio_1_watterson.txt | cut -d"_" -f1 | sed 's/.\{2\}$//' | sort > list_genes_of_interest_lynx_lynx_high_diversity_ratio_1_watterson_trimmed.txt

tail -n+2 list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_watterson.txt | cut -d"_" -f1 | sed 's/.\{2\}$//' | sort > list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_watterson_trimmed.txt


```


## 1. Aquellas unidades con ratios mayores de uno comunes a las tres comparaciones (con dos especies).




###  Loading my files

```{r}
file_input <- "list_for_venn_lynx_lynx_lynx_pardinus_ratio_above_1_CDS_trimmed_noduplicates.txt"
file_input <- "list_for_venn_lynx_lynx_lynx_pardinus_ratio_above_1_CDS_trimmed.txt"

file_output <- "list_for_venn_lynx_lynx_lynx_pardinus_ratio_above_1_CDS_trimmed_noduplicates"
file_output <- "list_for_venn_lynx_lynx_lynx_pardinus_ratio_above_1_CDS_trimmed"

description <- "Analysis of CDS with a ratio above 1 lynx lynx & lynx pardinus"
```


```{r}

# Genes de interes
gene_of_interest <- as.data.frame(read.table(paste0(wd, file_input), header=F)) %>% mutate (V1=as.character(V1)) 

# Lo convierto en tabla y lo proceso para GO analysis.
gene_of_interest_list <- unlist(gene_of_interest$V1)

# Hago una lista con 0 y 1 con los genes que tengo que filtrar
geneList <- factor(as.integer(geneNames %in% gene_of_interest_list))

# Asigno esa lista a cada gen (pongo header)
names(geneList) <- as.factor(geneNames)

str(geneList)

```

###  GO analysis 

```{r}

GOdata <- new("topGOdata", ontology = "BP", description = "Analysis of CDS with a ratio above 1 lynx lynx & lynx pardinus", allGenes = geneList,
               annot = annFUN.gene2GO, gene2GO = list_of_list_all_genes_GO_term_chr)

description (GOdata)
termStat(GOdata)

test.stat <- new("weight01Count", testStatistic = GOFisherTest, name = "Fisher test")
resultFisher <- getSigGroups(GOdata, test.stat)

head(score(resultFisher))
hist(score(resultFisher))
geneData(resultFisher)

allRes <- GenTable(GOdata, weighted01=resultFisher, orderBy="weight01", ranksOf="weight01", topNodes=50 )

write.table(allRes, paste0(wd, file_output, "_results_GO.txt"), quote = F, row.names = F)

```

###  Pathway enrichment


Intersección con mi archivos de genes de interes

```{r}
## Uno las dos tablas genes de interes, y KO term. 
genes_of_interest_KO <- inner_join(genes_to_KO.list, gene_of_interest) %>%  distinct(V1,V2, .keep_all = TRUE)
write.table(genes_of_interest_KO$V2, paste0(wd, file_output, "_results_KO.txt"), quote = F, row.names = F, col.names = F)

```



## 2. Aquellas unidades con ratios mayores de uno en la comparación de las dos poblaciones de lynx lynx. 


### Loading my files

```{r}
file_input <- "list_for_venn_lynx_lynx_ratio_above_1_CDS_trimmed_noduplicates.txt"
file_output <- "list_for_venn_lynx_lynx_ratio_above_1_CDS_trimmed_noduplicates"
description <- "Analysis of CDS with a ratio above 1 lynx lynx"

```



```{r}

# Genes de interes

gene_of_interest <- as.data.frame(read.table(paste0(wd, file_input), header=F)) %>% mutate (V1=as.character(V1)) 


# Lo convierto en tabla y lo proceso para GO analysis.
gene_of_interest_list <- unlist(gene_of_interest$V1)

# Hago una lista con 0 y 1 con los genes que tengo que filtrar
geneList <- factor(as.integer(geneNames %in% gene_of_interest_list))

# Asigno esa lista a cada gen (pongo header)
names(geneList) <- as.factor(geneNames)

str(geneList)

```

### GO analysis 

```{r}

GOdata <- new("topGOdata", ontology = "BP", description = description, allGenes = geneList,
               annot = annFUN.gene2GO, gene2GO = list_of_list_all_genes_GO_term_chr)

description (GOdata)
termStat(GOdata)

test.stat <- new("weight01Count", testStatistic = GOFisherTest, name = "Fisher test")
resultFisher <- getSigGroups(GOdata, test.stat)

head(score(resultFisher))
hist(score(resultFisher))
geneData(resultFisher)

allRes <- GenTable(GOdata, weighted01=resultFisher, orderBy="weight01", ranksOf="weight01", topNodes=50 )

write.table(allRes, paste0(wd, file_output, "_results_GO.txt"), quote = F, row.names = F)

```

###  Pathway enrichment


Intersección con mi archivos de genes de interes

```{r}
## Uno las dos tablas genes de interes, y KO term. 
genes_of_interest_KO <- inner_join(genes_to_KO.list, gene_of_interest) %>%  distinct(V1,V2, .keep_all = TRUE)
write.table(genes_of_interest_KO$V2, paste0(wd, file_output, "_results_KO.txt"), quote = F, row.names = F, col.names = F)

```


## 3. Aquellas unidades con ratios ~1 y diversidad en las tres comparaciones

### Watterson:

#### Loading my files

```{r}
#file_input <- "list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_trimmed_noduplicates.txt"
file_input <- "list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_watterson_trimmed_noduplicates.txt"
#file_output <- "list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_trimmed_noduplicates"
file_output <- "list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_watterson_trimmed_noduplicates"

description <- "Analysis of CDS with a ratio over 1 and high Watterson diversity lynx lynx & lynx pardinus"

```


```{r}

# Genes de interes
gene_of_interest <- as.data.frame(read.table(paste0(wd, file_input), header=F)) %>% mutate (V1=as.character(V1)) 

# Lo convierto en tabla y lo proceso para GO analysis.
gene_of_interest_list <- unlist(gene_of_interest$V1)

# Hago una lista con 0 y 1 con los genes que tengo que filtrar
geneList <- factor(as.integer(geneNames %in% gene_of_interest_list))

# Asigno esa lista a cada gen (pongo header)
names(geneList) <- as.factor(geneNames)

str(geneList)

```

#### GO analysis 

```{r}
# Puedes cambiar ontology y poner BP (biological process), CC (celular component), or MF (molecular function). Cambia el nombre del archivo como se guarda. 

GOdata <- new("topGOdata", ontology = "CC", description = description, allGenes = geneList,
               annot = annFUN.gene2GO, gene2GO = list_of_list_all_genes_GO_term_chr)

description (GOdata)
termStat(GOdata)

test.stat <- new("weight01Count", testStatistic = GOFisherTest, name = "Fisher test")
resultFisher <- getSigGroups(GOdata, test.stat)

head(score(resultFisher))
hist(score(resultFisher))
geneData(resultFisher)

allRes <- GenTable(GOdata, weighted01=resultFisher, orderBy="weight01", ranksOf="weight01", topNodes=50, numChar=300 )

write.table(allRes, paste0(wd, file_output, "_results_GO_CC_Watterson.txt"), quote = F, row.names = F)

```

###  Pathway enrichment


Intersección con mi archivos de genes de interes

```{r}
## Uno las dos tablas genes de interes, y KO term. 
genes_of_interest_KO <- inner_join(genes_to_KO.list, gene_of_interest) %>%  distinct(V1,V2, .keep_all = TRUE)
write.table(genes_of_interest_KO$V2, paste0(wd, file_output, "_results_KO_Watterson.txt"), quote = F, row.names = F, col.names = F)

```

### Pi:

#### Loading my files

```{r}
#file_input <- "list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_trimmed_noduplicates.txt"
file_input <- "list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_pairwise_trimmed_noduplicates.txt"
#file_output <- "list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_trimmed_noduplicates"
file_output <- "list_genes_of_interest_lynx_lynx_lynx_pardinus_high_diversity_ratio_1_pairwise_trimmed_noduplicates"

description <- "Analysis of CDS with a ratio over 1 and high Pi diversity lynx lynx & lynx pardinus"

```


```{r}

# Genes de interes
gene_of_interest <- as.data.frame(read.table(paste0(wd, file_input), header=F)) %>% mutate (V1=as.character(V1)) 

# Lo convierto en tabla y lo proceso para GO analysis.
gene_of_interest_list <- unlist(gene_of_interest$V1)

# Hago una lista con 0 y 1 con los genes que tengo que filtrar
geneList <- factor(as.integer(geneNames %in% gene_of_interest_list))

# Asigno esa lista a cada gen (pongo header)
names(geneList) <- as.factor(geneNames)

str(geneList)

```

#### GO analysis 

```{r}
# Puedes cambiar ontology y poner BP (biological process), CC (celular component), or MF (molecular function). Cambia el nombre del archivo como se guarda. 

GOdata <- new("topGOdata", ontology = "MF", description = description, allGenes = geneList,
               annot = annFUN.gene2GO, gene2GO = list_of_list_all_genes_GO_term_chr)

description (GOdata)
termStat(GOdata)

test.stat <- new("weight01Count", testStatistic = GOFisherTest, name = "Fisher test")
resultFisher <- getSigGroups(GOdata, test.stat)

head(score(resultFisher))
hist(score(resultFisher))
geneData(resultFisher)

allRes <- GenTable(GOdata, weighted01=resultFisher, orderBy="weight01", ranksOf="weight01", topNodes=50, numChar=300 )

write.table(allRes, paste0(wd, file_output, "_results_GO_MF_Pairwise.txt"), quote = F, row.names = F)

```

###  Pathway enrichment


Intersección con mi archivos de genes de interes

```{r}
## Uno las dos tablas genes de interes, y KO term. 
genes_of_interest_KO <- inner_join(genes_to_KO.list, gene_of_interest) %>%  distinct(V1,V2, .keep_all = TRUE)
write.table(genes_of_interest_KO$V2, paste0(wd, file_output, "_results_KO_Pairwise.txt"), quote = F, row.names = F, col.names = F)

```


## 4. Aquellas unidades con ratios ~1 en la comparación de las dos poblaciones de lynx lynx. 


### Loading my files

```{r}
file_input <- "list_genes_of_interest_lynx_lynx_high_diversity_ratio_1_trimmed_noduplicates.txt"
file_output <- "list_genes_of_interest_lynx_lynx_high_diversity_ratio_1_trimmed_noduplicates"
description <- "Analysis of CDS with a ratio over 1 and high diversity lynx lynx"

```


```{r}

# Genes de interes
gene_of_interest <- as.data.frame(read.table(paste0(wd, file_input), header=F)) %>% mutate (V1=as.character(V1)) 


# Lo convierto en tabla y lo proceso para GO analysis.
gene_of_interest_list <- unlist(gene_of_interest$V1)

# Hago una lista con 0 y 1 con los genes que tengo que filtrar
geneList <- factor(as.integer(geneNames %in% gene_of_interest_list))

# Asigno esa lista a cada gen (pongo header)
names(geneList) <- as.factor(geneNames)

str(geneList)

```

### GO analysis 

```{r}

GOdata <- new("topGOdata", ontology = "BP", description = description, allGenes = geneList,
               annot = annFUN.gene2GO, gene2GO = list_of_list_all_genes_GO_term_chr)

description (GOdata)
termStat(GOdata)

test.stat <- new("weight01Count", testStatistic = GOFisherTest, name = "Fisher test")
resultFisher <- getSigGroups(GOdata, test.stat)

head(score(resultFisher))
hist(score(resultFisher))
geneData(resultFisher)

allRes <- GenTable(GOdata, weighted01=resultFisher, orderBy="weight01", ranksOf="weight01", topNodes=50, numChar=300 )

write.table(allRes, paste0(wd, file_output, "_results_GO_BP.txt"), quote = F, row.names = F)

```


###  Pathway enrichment


Intersección con mi archivos de genes de interes

```{r}
## Uno las dos tablas genes de interes, y KO term. 
genes_of_interest_KO <- inner_join(genes_to_KO.list, gene_of_interest) %>%  distinct(V1,V2, .keep_all = TRUE)
write.table(genes_of_interest_KO$V2, paste0(wd, file_output, "_results_KO.txt"), quote = F, row.names = F, col.names = F)

```






