---
title: "14.GO_enrichment"
output: html_document
---



# 1. Load library and required information

Este es el script donde realizo el GO term annotation & enrichment analysis y el pathway enrichment analysis. 

Lo primero que hago es pedirle a Godoy una lista donde cada gen tenga asociado su GO-term. El me pasa una carpeta donde hay varios archivos.

GO2LYNPC.txt
GO2LYPA.LYPA23C.APPRIS.fatigo.genes.txt
GO2LYPA.LYPA23C.APPRIS.fatigo.txt
GO_enrichment.docx
genes2transcript2go.list.txt
lp23.allgenes.fatigo.txt
lp23.allgenes.goterms
lp23.allgenes.id

En el archivo docx, se explica los pasos que siguió Godoy para hacer el enrichment. Lo que deduzco es que las que se llaman lp23* son listas antiguas, y despues están las GO2LYPA que sólo contienen las isoformas principales. 

Lo distintos archivos son:

GO2LYNPC.txt --> Tiene el nombre del gen, el GO_term, el GO_name y el GO_type; parece que incluye todas las isoformas.
GO2LYPA.LYPA23C.APPRIS.fatigo.genes.txt --> Creo q es el que tengo que usar. Nombre de gen, GO_ACC, GO_name y el GO_type.
GO2LYPA.LYPA23C.APPRIS.fatigo.txt --> Igual pero con proteinas.
GO_enrichment.docx --> Archivo docx donde se explica el procedimiento.
genes2transcript2go.list.txt --> Creo que es la carpeta que tiene el nombre de gen, isoforma principal y proteina y despues GO term y sus características.

Los lp23 son antiguos sin la isoforma principal.

Lo primero que voy a hacer es cargar librerias y nuestros archivos:

## Load information for GO annotation & enrichment

```{r}
library(topGO)
library (plyr)
library(tidyr)
library (dplyr)


# Working directory

wd <- "/Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/Lypa23c_GO_terms_KO_terms_annotation/"

# Tabla con la información de GO term para cada gen.

tabla_all_genes_GO_term_complete_info <- read.table (paste0(wd, "GO2LYPA.LYPA23C.APPRIS.fatigo.genes_minor_modif.txt"), sep="\t", header = T) %>% mutate (GO_ACC=as.character(GO_ACC))
# El minor_modif tiene ' cambiadas por empty space pq si no R entiende que es un string. 


## Modifo el archivo para poder generar un gmt con los nombres de cada GO term, su descripción, y a que genes va asociado. 
# Este archivo lo vamos a necesitar para cargar los datos en CYTOSCAPE. 
tabla_all_genes_GO_term_complete_info_modif <- tabla_all_genes_GO_term_complete_info %>% dplyr::select (-"GO_TYPE") %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(GENE_NAME, collapse=','))  

# Guardo la tabla
write.table(tabla_all_genes_GO_term_complete_info_modif, paste0(wd, "GO2LYPA.LYPA23C.APPRIS.fatigo.genes_modif.NAMES.gmt"), quote = F, row.names = F, col.names = F, sep = "\t" )


# Modificaciones para topGo.
list_of_list_all_genes_GO_term <- dlply(tabla_all_genes_GO_term_complete_info, .(GENE_NAME), function(x)x$GO_ACC)
list_of_list_all_genes_GO_term_chr <- lapply(list_of_list_all_genes_GO_term, function(x) as.character(unlist(x)))

# Guardo los nombres de los genes en cuestión. 
geneNames <- names(list_of_list_all_genes_GO_term_chr)

```

## Load information for pathway enrichment

No hace falta hacer este chunk, pq solo se tiene que hacer una vez.

- Previous work

Voy a seleccionar los KO que son los nombres de proteinas comunes a todos


<!-- ```{bash} -->

<!-- cd /GRUPOS/grupolince/Lyp_annotation_Apr14_final -->

<!-- grep "ko_group" LYPA23C.all.fix.nr.gff3 > all_mRNA -->

<!-- # sanity check -->
<!-- # all ko groups -->
<!-- grep -o '\bko_group=\w*' all_mRNA | wc -l -->
<!-- #10213 -->
<!-- # All parent_id -->
<!-- grep -o '\bParent=\w*' all_mRNA | wc -l -->
<!-- #10213 -->

<!-- # Igual! perfecto, los podemos juntar -->

<!-- paste <(grep -o '\bParent=\w*' all_mRNA) <(grep -o '\bko_group=\w*' all_mRNA) | sed 's/Parent=//g' | sed 's/ko_group=//g' > genes_to_KO.txt -->

<!-- rm all_mRNA -->

<!-- ``` -->

<!-- Me lo bajo a mi carpeta -->

<!-- ```{bash} -->
<!-- scp mlucena@genomics-b.ebd.csic.es:/GRUPOS/grupolince/Lyp_annotation_Apr14_final/genes_to_KO.txt /Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/Lypa23c_GO_terms_KO_terms_annotation -->
<!-- ``` -->

- Load KO information for pathway enrichment

Al final vamos a usar los GO term en lugar de los KO para darle a REACTOME

From reactome webpage: https://reactome.org/userguide/analysis
Entities:
If your data is a single column of identifiers such as UniProt IDs, gene symbols or ChEBI IDs, they are mapped to pathways and over-representation and pathway-topology analyses are run. Over-representation analysis is a statistical (hypergeometric distribution) test that determines whether certain Reactome pathways are over-represented (enriched) in the submitted data. It answers the question ‘Does my list contain more proteins for pathway X than would be expected by chance?’ This test produces a probability score, which is corrected for false discovery rate using the Benjamani-Hochberg method.

Reactions:
Pathway topology analysis considers the connectivity between molecules that is represented by the pathway steps (which we refer to as reactions) in the pathway. It groups all the molecules represented in each reaction as a pathway ‘unit’. If any of these molecules are represented in your query set, this is considered a match to that reaction. This may give a better indication of the proportion of the pathway that matches your data, rather than the number of molecules that are common between your data and the pathway. It may also indicate that your data matches the start, end or a particular branch of a pathway process. This test does not have a probability score.



```{r}
# Archivo de equivalencias KO con genes

# genes_to_KO.list <- read.table("/Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/Lypa23c_GO_terms_KO_terms_annotation/genes_to_KO.txt", header=F)

```


Tengo tablas individuales y las quiero anotar con los GO term correspondientes

Lo uno a la tabla de GO term

# 2. Load files

## 2.1. Watterson

### HD_HD_ratio_1_watterson individual

```{r}
wd <- "/Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/Lypa23c_GO_terms_KO_terms_annotation/"

# Kirov vs Norway

c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson$Gen)

# Kirov vs Poland

c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson$Gen)
                        
# Andujar vs Doñana

c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson <- read.table (paste0(wd,"input_data/c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson$Gen <- sub("P[^P]+$", "", c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson$Gen)
 
```

### HD_HD_ratio_1_watterson in common

```{r}

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson$Gen)
 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson$Gen)
 

```

### ND-D individual

```{r}
wd <- "/Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/Lypa23c_GO_terms_KO_terms_annotation/"

# Kirov vs Norway

c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson$Gen)

# Kirov vs Poland

c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson$Gen)
                        
# Andujar vs Doñana

c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson <- read.table (paste0(wd,"input_data/c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson$Gen <- sub("P[^P]+$", "", c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson$Gen)
 
```

### ND_D in common

```{r}

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson$Gen)
 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson$Gen)
 

```


## 2.2. Pairwise

### HD_HD_ratio_1_pairwise individual

```{r}
wd <- "/Users/marialucenaperez/Owncloud/publico/PhD/WG_diversity/Lypa23c_GO_terms_KO_terms_annotation/"

# Kirov vs Norway

c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise$Gen)

# Kirov vs Poland

c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise$Gen)
                        
# Andujar vs Doñana

c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise <- read.table (paste0(wd,"input_data/c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise$Gen <- sub("P[^P]+$", "", c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise$Gen)
 
```

### HD_HD_ratio_1_pairwise in common

```{r}

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise$Gen)
 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise$Gen)
 

```

### ND-D individual

```{r}
# Kirov vs Norway

c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise$Gen)

# Kirov vs Poland

c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise$Gen)
                        
# Andujar vs Doñana

c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise <- read.table (paste0(wd,"input_data/c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise$Gen <- sub("P[^P]+$", "", c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise$Gen)
 
```

### ND_D in common

```{r}

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise$Gen)
 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise <- read.table (paste0(wd,"input_data/c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise", sep=""), header = T, stringsAsFactors = F) 

# Así quito todo lo que venga despues de la última P. 
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise$Gen <- sub("P[^P]+$", "", c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise$Gen)
 

```

# 3. CYTOSCAPE: ALL GENES:

GO annotation enrichment map.

En esta sección vamos a preparar los archivos para poder visualizarlos en el enrichment map de cytoscape. Esta manera de visualizarlos, es muy útil pues puedes seleccionar qué archivos están en las tres comparaciones, y colorearlas acorde. Además nos devuelve la descripción, y podemos jugar a encontrar si hay alguna descripción interesante, a pesar de no estar enriched. 

## 3.1 Watterson 

### HD_HD_ratio_1_watterson individual

```{r}

# Kirov vs Norway

c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()
write.table(c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )


# Kirov vs Poland

c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile <-  c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 


c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile <- c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

write.table(c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

# Andujar vs Doñana

c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info <- dplyr::inner_join(c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile <-  c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile <- c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

write.table(c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

```

### HD_HD_ratio_1_watterson in common

Transformación de los archivos con varias poblaciones.

```{r}

# Comparaciones en lynx lynx
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

# En principio no lo guardo pq para enrichment map me gusta mas cargar los inviduales y mirarlos en excel: 
#write.table(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile, paste0(wd, "c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )
# Las 3 comparaciones:

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

# En principio no lo guardo pq para enrichment map me gusta mas cargar los inviduales y mirarlos en excel: 
#write.table(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile, paste0(wd, "c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )
```

### ND-D individual

```{r}

# Kirov vs Norway

c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()
write.table(c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_watterson_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )


# Kirov vs Poland

c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info.gProfile <-  c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 


c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info.gProfile <- c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

write.table(c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

# Andujar vs Doñana

c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info <- dplyr::inner_join(c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info.gProfile <-  c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info.gProfile <- c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

write.table(c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

```

### ND_D_ratio_1_watterson in common

Transformación de los archivos con varias poblaciones.

```{r}

# Comparaciones en lynx lynx
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

# En principio no lo guardo pq para enrichment map me gusta mas cargar los inviduales y mirarlos en excel: 
#write.table(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info.gProfile, paste0(wd, "c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_watterson_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )
# Las 3 comparaciones:

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

# En principio no lo guardo pq para enrichment map me gusta mas cargar los inviduales y mirarlos en excel: 
#write.table(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info.gProfile, paste0(wd, "c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_watterson_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )
```
## 3.2. Pairwise

### HD_HD_ratio_1_watterson individual

Ahora los voy a preparar para tener archivos de entrada al programa enrichment map. Este programa nos devuelve unas tablas muy cómodas con las que jugar despues. 

```{r}

# Kirov vs Norway

c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

write.table(c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

# Kirov vs Poland

c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile <-  c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile <- c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

write.table(c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

# Andujar vs Doñana

c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info <- dplyr::inner_join(c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile <-  c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile <- c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

write.table(c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

```


### HD_HD_ratio_1_watterson in common

Transformación de los archivos con varias poblaciones.

```{r}

# Comparaciones en lynx lynx
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

# En principio no lo guardo pq para enrichment map me gusta mas cargar los inviduales y mirarlos en excel: 
#write.table(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile, paste0(wd, "c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )
# Las 3 comparaciones:

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

# En principio no lo guardo pq para enrichment map me gusta mas cargar los inviduales y mirarlos en excel: 
#write.table(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile, paste0(wd, "c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

```


### ND-D individual

```{r}

# Kirov vs Norway

c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()
write.table(c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_ll_ki_n013_c_ll_no_n008_list_CDS_ND_D_pairwise_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )


# Kirov vs Poland

c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info.gProfile <-  c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 


c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info.gProfile <- c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

write.table(c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

# Andujar vs Doñana

c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info <- dplyr::inner_join(c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 

c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info.gProfile <-  c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info.gProfile <- c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

write.table(c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

```

### ND_D_ratio_1_pairwise in common

Transformación de los archivos con varias poblaciones.

```{r}

# Comparaciones en lynx lynx
c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

# En principio no lo guardo pq para enrichment map me gusta mas cargar los inviduales y mirarlos en excel: 
#write.table(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info.gProfile, paste0(wd, "c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_CDS_ND_D_pairwise_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )
# Las 3 comparaciones:

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info <- dplyr::inner_join(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise,                                             tabla_all_genes_GO_term_complete_info, by=c("Gen"="GENE_NAME")) 


c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info.gProfile <-  c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info %>% dplyr::group_by(GO_ACC,GO_NAME) %>% dplyr::summarise(Genes=paste(Gen, collapse=',')) %>% dplyr::mutate(p.Val=1) %>% dplyr::mutate(FDR=1) %>% dplyr::mutate(Phenotype=1) %>% dplyr::rename(GO.ID=GO_ACC, Description=GO_NAME) 

c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info.gProfile <- c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info.gProfile[c("GO.ID","Description","p.Val","FDR","Phenotype","Genes")] %>%  drop_na()

# En principio no lo guardo pq para enrichment map me gusta mas cargar los inviduales y mirarlos en excel: 
#write.table(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info.gProfile, paste0(wd, "c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_CDS_ND_D_pairwise_all_info.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )
```
# 4. Reactome input

Guardo para cada archivo una lista de los GO term. 
```{r}

# Watterson

## Comparaciones_individuales

write.table(c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile[1], paste0(wd, "REACTOME/input_data/c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.GO_terms.txt"), quote = F, row.names = F, col.names = F, sep = "\t" ) 

write.table(c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile[1], paste0(wd, "REACTOME/input_data/c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.GO_terms.txt"), quote = F, row.names = F, col.names = F, sep = "\t" ) 

write.table(c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile[1], paste0(wd, "REACTOME/input_data/c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.GO_terms.txt"), quote = F, row.names = F, col.names = F, sep = "\t" ) 

## Varias comparaciones

write.table(c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile[1], paste0(wd, "REACTOME/input_data/c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.GO_terms.txt"), quote = F, row.names = F, col.names = F, sep = "\t" ) 

write.table(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile[1], paste0(wd, "REACTOME/input_data/c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.GO_terms.txt"), quote = F, row.names = F, col.names = F, sep = "\t" ) 



# Pairwise

## Comparaciones individuales

write.table(c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile[1], paste0(wd, "REACTOME/input_data/c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.GO_terms.txt"), quote = F, row.names = F, col.names = F, sep = "\t" ) 

write.table(c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile[1], paste0(wd, "REACTOME/input_data/c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.GO_terms.txt"), quote = F, row.names = F, col.names = F, sep = "\t" ) 

write.table(c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile[1], paste0(wd, "REACTOME/input_data/c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.GO_terms.txt"), quote = F, row.names = F, col.names = F, sep = "\t" ) 

## Varias comparaciones

write.table(c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile[1], paste0(wd, "REACTOME/input_data/c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.GO_terms.txt"), quote = F, row.names = F, col.names = F, sep = "\t" ) 

write.table(c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.gProfile[1], paste0(wd, "REACTOME/input_data/c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_all_info.GO_terms.txt"), quote = F, row.names = F, col.names = F, sep = "\t" ) 



```

# 5. GO term enrichment 

Hago un loop que hace el enrichment sobre todos los dataframe de interes. 

```{r}

DATAFRAMES = list(
c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson = c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson, c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson = c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson, c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson = c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson, c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson = c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson, c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson = c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson, 
c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise = c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise, c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise = c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise, c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise = c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise, c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise = c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise, c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise = c_ll_ki_n013_c_ll_no_n008_c_ll_ki_n013_c_ll_po_n008_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise)

# Corro en bucle los análisis de enrichemnt:
## Biological procces

for (i in 1:length(DATAFRAMES)) {

print(names(DATAFRAMES)[i])

# Primero genero la lista Gen_list
# Lo convierto en tabla y lo proceso para GO analysis.
gene_of_interest_list <- unlist(DATAFRAMES[[i]]$Gen)
# Hago una lista con 0 y 1 con los genes que tengo que filtrar
geneList <- factor(as.integer(geneNames %in% gene_of_interest_list))
# Asigno esa lista a cada gen (pongo header)
names(geneList) <- as.factor(geneNames)

# GO analysis itself

# Puedes cambiar ontology y poner BP (biological process), CC (celular component), or MF (molecular function). 
GOdata <- new("topGOdata", ontology = "BP", description = names(DATAFRAMES)[i], allGenes = geneList,
               annot = annFUN.gene2GO, gene2GO = list_of_list_all_genes_GO_term_chr)
# description (GOdata)
# termStat(GOdata)
test.stat <- new("weight01Count", testStatistic = GOFisherTest, name = "Fisher test")
resultFisher <- getSigGroups(GOdata, test.stat)
# head(score(resultFisher))
# hist(score(resultFisher))
# geneData(resultFisher)
allRes <- GenTable(GOdata, weighted01=resultFisher, orderBy="weight01", ranksOf="weight01", topNodes=150, numChar=300 )

# Guardo los resultados con el nombre del dataframe
name=names(DATAFRAMES)[i]
assign(paste0(name, "_GO_term_enrichment_BP"), allRes )

write.table(allRes, paste0(wd, "topGO/results/", name, "_results_GO_enrichment_BP.txt"), quote = T, row.names = F)

rm(list = c('gene_of_interest_list','geneList','GOdata','test.stat','resultFisher','allRes'))

}


## Celular component

for (i in 1:length(DATAFRAMES)) {

print(names(DATAFRAMES)[i])

# Primero genero la lista Gen_list
# Lo convierto en tabla y lo proceso para GO analysis.
gene_of_interest_list <- unlist(DATAFRAMES[[i]]$Gen)
# Hago una lista con 0 y 1 con los genes que tengo que filtrar
geneList <- factor(as.integer(geneNames %in% gene_of_interest_list))
# Asigno esa lista a cada gen (pongo header)
names(geneList) <- as.factor(geneNames)

# GO analysis itself
name=names(DATAFRAMES)[i]

# Puedes cambiar ontology y poner BP (biological process), CC (celular component), or MF (molecular function). 
GOdata <- new("topGOdata", ontology ='CC', description = "kk", allGenes = geneList,
               annot = annFUN.gene2GO, gene2GO = list_of_list_all_genes_GO_term_chr)
# description (GOdata)
# termStat(GOdata)
test.stat <- new("weight01Count", testStatistic = GOFisherTest, name = "Fisher test")
resultFisher <- getSigGroups(GOdata, test.stat)
# head(score(resultFisher))
# hist(score(resultFisher))
# geneData(resultFisher)
allRes <- GenTable(GOdata, weighted01=resultFisher, orderBy="weight01", ranksOf="weight01", topNodes=150, numChar=300 )

# Guardo los resultados con el nombre del dataframe
assign(paste0(name, "_GO_term_enrichment_CC"), allRes )

write.table(allRes, paste0(wd, "topGO/results/", name, "_results_GO_enrichment_CC.txt"), quote = T, row.names = F)

rm(list = c('gene_of_interest_list','geneList','GOdata','test.stat','resultFisher','allRes'))

}


## Molecular funtion

for (i in 1:length(DATAFRAMES)) {

print(names(DATAFRAMES)[i])

# Primero genero la lista Gen_list
# Lo convierto en tabla y lo proceso para GO analysis.
gene_of_interest_list <- unlist(DATAFRAMES[[i]]$Gen)
# Hago una lista con 0 y 1 con los genes que tengo que filtrar
geneList <- factor(as.integer(geneNames %in% gene_of_interest_list))
# Asigno esa lista a cada gen (pongo header)
names(geneList) <- as.factor(geneNames)

# GO analysis itself
name=names(DATAFRAMES)[i]

# Puedes cambiar ontology y poner BP (biological process), CC (celular component), or MF (molecular function). 
GOdata <- new("topGOdata", ontology ='MF', description = "kk", allGenes = geneList,
               annot = annFUN.gene2GO, gene2GO = list_of_list_all_genes_GO_term_chr)
# description (GOdata)
# termStat(GOdata)
test.stat <- new("weight01Count", testStatistic = GOFisherTest, name = "Fisher test")
resultFisher <- getSigGroups(GOdata, test.stat)
# head(score(resultFisher))
# hist(score(resultFisher))
# geneData(resultFisher)
allRes <- GenTable(GOdata, weighted01=resultFisher, orderBy="weight01", ranksOf="weight01", topNodes=150, numChar=300 )

# Guardo los resultados con el nombre del dataframe
assign(paste0(name, "_GO_term_enrichment_MF"), allRes )

write.table(allRes, paste0(wd, "topGO/results/", name, "_results_GO_enrichment_MF.txt"), quote = T, row.names = F)

rm(list = c('gene_of_interest_list','geneList','GOdata','test.stat','resultFisher','allRes'))

}


```

## 5.1 GO term in common


```{r}
# Ahora seleccionamos los GO term que están enriched en varios dataset.


c_ll_ki_n013_c_ll_no_n008_pairwise_PLUS_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_results_GO_enrichment_BP <- dplyr::inner_join (c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term, c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term, by=c("GO.ID","Term")) 
# %>% filter(., weighted01.x<0.05 ) %>% filter(., weighted01.y<0.05 ) 


write.table(c_ll_ki_n013_c_ll_no_n008_pairwise_PLUS_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_results_GO_enrichment_BP, paste0(wd, "topGO/results/c_ll_ki_n013_c_ll_no_n008_pairwise_PLUS_c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_results_GO_enrichment_BP.txt", sep=""), quote = T, row.names = F)



c_ll_ki_n013_c_ll_no_n008_PLUS_c_ll_ki_n013_c_ll_po_n008_PLUS_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term <- dplyr::inner_join (c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term, c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term, by=c("GO.ID","Term")) %>% dplyr::inner_join(., c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term, by=c("GO.ID","Term")) 
# %>% filter(., weighted01<0.05 ) %>% filter(., weighted01.x<0.05 ) %>% filter(., weighted01.y<0.05 ) 

write.table(c_ll_ki_n013_c_ll_no_n008_PLUS_c_ll_ki_n013_c_ll_po_n008_PLUS_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term, paste0(wd, "topGO/results/c_ll_ki_n013_c_ll_no_n008_PLUS_c_ll_ki_n013_c_ll_po_n008_PLUS_c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term.txt", sep=""), quote = T, row.names = F)




```

# 6. CYTOSCAPE GO enriched!

Ahora voy a genera los archivos necesarios para correr CYTOSCAPE usando los p valores del enrichment test. 

```{r}
# Uno la tabla con nuestro p valores a la tabla que ya habiamos construido usando nuestros genes con los GO term a los que se asocian. 

# c_ll_ki vs c_ll_no
c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment.gProfile <- rbind (c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment_BP, c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment_CC, c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment_MF) %>%  inner_join(., as.data.frame(c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile), by=c("GO.ID"="GO.ID", "Term"="Description")) %>% dplyr::select( GO.ID, Description=Term, p.Val=weighted01, FDR, Phenotype, Genes)
write.table(c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_ll_ki_n013_c_ll_no_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )

# c_ll_ki vs c_ll_po
c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment.gProfile <- rbind (c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment_BP, c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment_CC, c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment_MF) %>%  inner_join(., as.data.frame(c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile), by=c("GO.ID"="GO.ID", "Term"="Description")) %>% dplyr::select( GO.ID, Description=Term, p.Val=weighted01, FDR, Phenotype, Genes)
write.table(c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_ll_ki_n013_c_ll_po_n008_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )


# c_lp_sm vs c_lp_do
c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment.gProfile <- rbind (c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment_BP, c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment_CC, c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment_MF) %>%  inner_join(., as.data.frame(c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_watterson_all_info.gProfile), by=c("GO.ID"="GO.ID", "Term"="Description")) %>% dplyr::select( GO.ID, Description=Term, p.Val=weighted01, FDR, Phenotype, Genes)
write.table(c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment.gProfile, paste0(wd, "CYTOSCAPE/input_data/c_lp_sm_n019_c_lp_do_n012_list_of_CDS_HD_HD_ratio_1_pairwise_GO_term_enrichment.gProfile.txt"), quote = F, row.names = F, col.names = F, sep = "\t" )


```


## Otras notas

From REACTOME paper

The Kyoto Encyclopedia of Genes and Genomes, or KEGG (Kanehisa et al., 2004) features an extensive set of biological pathway charts. Like HumanCyc, KEGG focuses on intermediate metabolism rather than higher-level pathways. Its data model differs fundamentally from Reactome’s by representing the motivating force of all reactions in the form of catalyst activities via Enzyme Commission EC numbers. Because there is not a one-to-one mapping between EC activity and polypeptide, it can be problematic to relate a protein represented in SwissProt to a reaction represented in KEGG.

-- Otros:

1. There is the Gene Ontology.  Its terms are used (primarily)
for the annotation of gene products.  Both the Ontology and the
annotations contributed by the members of the GO Consortium database
are available from the GO site.

2. There is the KEGG Orthology, available from the KEGG site.
This is _both_ an ontology, seen, for example, by
opening KO up to its 3rd level: http://www.genome.ad.jp/dbget-bin/get_htext?KO+-s+F+-f+F+C
_and_ annotations of classes of gene product, seen if it is opened up
to level 4:
http://www.genome.ad.jp/dbget-bin/get_htext?KO+-s+F+-f+F+D







