---
title: "pipeline_PCA"
output: html_document
---

```{r, engine=bash, eval=FALSE}

####################################################################################
#
#      								Population Structure (PCA, MDS(ngsDist))
#
#										María Lucena Pérez
#
####################################################################################

# Info from ANGSD page, and github Fumagalli/ngsTools/TUTORIAL.md
# One solution would be to perform a Principal Component Analysis (PCA) or a Multidimensional Scaling (MDS) or some clustering based on genetic distances among samples. 
# We are here showing how to perform these analyses using ngsTools/ANGSD in case of low-depth data.


#######################################################################################################################################
#### Principal Component Analysis 
#######################################################################################################################################


# Back to our example, we are now performing a principal component analyses (PCA) without relying on called genotypes, 
# but rather by taking their uncertainty into account. 
# More specifically, the next program we are going to use (ngsTools) takes as input genotype probabilities in binary format, 
# so we need to specify -doGeno 32. Also, we are using a HWE-based prior with -doPost 1. 
# Recalling also our choice for data filtering, our command line is:

# Change directory

cd /home/mlucena/ANGSD_intergenic_analysis
# I want to produce PCAs for the whole dataset, each specie, and within iberian with and without historical samples. 
# I think that per population PCAs make no sense

#To launch one by one ( because otherwise it's too slow I'll launch single pops in a and groups in b)
#Lanzo cada uno en una sesión de screen diferente y con un log distinto ( mediante script). lanzo el script conpleto dandole como input el csv que acabo de crear. Ejemplo:

screen -S pca_intergenic_european_oriental
script pca_intergenic_european_oriental.log

POP="c_ll_ka-c_ll_og-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya"  # <--CHANGE POP HERE




# Set paths to software and some commom files
ANGSD=/opt/angsd
NGSTOOLS=/opt/ngsTools_19092016/
THREADS=50                     # no. of computer cores used by bwa and samtools. 20 = OK, >20 = ask people first!
REF="/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23_without_repetitive_transposable_low_complexity.fa"
ANC="/home/GRUPOS/grupolince/reference_genomes/lynx_rufus_genome/lr1_ref_norecal.fa"
FILTER1=" -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -baq 1 -C 50 "
FILTER2=" -minMapQ 30 -minQ 20 -doCounts 1 "


read POP mean sd mean_truncated sd_truncated maxDepth minDepth maxDepth_truncated minDepth_truncated < /home/mlucena/ANGSD_analysis/"$POP"_n*_mean_sd_depthGlobal_lynx_per_pop_mean_folds_0.95.csv


echo $POP
# We don't need regionfile because we are using a bam file with a subset of the data.

N_IND=$(echo ${POP: -3} )

MIN_IND=$(expr $N_IND / 1) # We stablish a sharp threshold, usually we divided into 2. 


echo $MIN_IND
### Calling genotype likelihoods: 
 
$ANGSD/angsd -P $THREADS -b /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP.intergenic.bamlist -ref $REF -out $POP.intergenic \
$FILTER1 $FILTER2 \
-minInd $MIN_IND -setMaxDepth $maxDepth -setMinDepth $minDepth \
-GL 1 -doMajorMinor 1 -doMaf 1 -skipTriallelic 1 \
-SNP_pval 1e-3 \
-doGeno 32 -doPost 1




### URGENTE ULTIMA ACTUALIZACION!!

# -minInd [int]
# Only keep sites with at least minIndDepth (default is 1) from at least [int] individuals
# -minIndDepth [int]
# Only works with the -minInd filter. Change the minimum depth the individuals must have in order to keep the site. Default is 1.


# SNP_pval = #ind*(2allelos/ind) = x alelos ; 1/x = y  # esto sería un filtro aceptable. En mi caso el dataset más pequeño (h_lypa_all) es de 55 indivíduos (1/(2*55)=9e-3) y el más grande (x_lyxx_all) de 189 (1/(2*189)=2.6e-3)
# Para el PCA vamos a usar 1e-3 porque es un filtro suuuper light.

# Unzip the results (but you cannot open it since it is in binary format)

gunzip $POP.intergenic.geno.gz

### ngsCovar

# We are going to use ngsCovar, which estimates the covariance matrix between individuals based on genotype probabilities. 
# Then this matrix will be decomposed into principal components which will be investigated for population structure analyses. 
# Note that although ngsCovar can account for SNPs uncertanity, we find that it is faster to perform a light SNP filtering first 
# (as we did using -SNP_pval 1e-3) than using all sites.

# If you type:

# $NGSTOOLS/ngsPopGen/ngsCovar

# you will see a list of possible options. 
# For instance, we need to define how many sites we have. To retrieve such value, we can inspect the file with allele frequencies:

#less -S $POP.mafs.gz
N_SITES=`zcat $POP.intergenic.mafs.gz | tail -n+2 | wc -l`
echo $N_SITES

# In my case I have 272064 sites. 
# Using my intergenic region, and minind=74, I have 114097 sites. 

# Now we can perform a PCA by estimating the covariance matrix first:
# $$$Do we want to add the SFS file??

$NGSTOOLS/ngsPopGen/ngsCovar -probfile $POP.*geno -outfile $POP.covar -nind $N_IND  -nsites $N_SITES -call 0 -norm 0 &> /dev/null

#mv c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-h_ll_ba_n074.covar #c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-h_ll_ba_n074.intergenic.covar

# with the options -call 0 meaning that we do not perform genotype calling 
# and -norm 0 that we are not normalising by allele frequency. 
# The latter may give more weight to low frequency variants which are harder to estimate.

# Look at the output file:

#less -S $POP.covar

# which represents a matrix of NxN with N individuals giving the covariance. Note that this matrix is symmetric.

# Finally, we perform an eigenvector decomposition and plot the resulting map:

# create a cluster-like file defining the labelling (population) for each sample

#Example: Rscript -e 'write.table(cbind(seq(1,60),rep(1,60),c(rep("LWK",20),rep("TSI",20),rep("PEL",20))), row.names=F, sep=" ", col.names=c("FID","IID","CLUSTER"), file="Results/ALL.clst", quote=F)'

echo "FID" > $POP.ALL1.clst.borrar
seq $N_IND >> $POP.ALL1.clst.borrar
echo "IID" > $POP.ALL2.clst.borrar
# yes "1" | head -n $N_IND >> $POP.ALL2.clst.borrar
cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 2,3,4,5 | rev  >> $POP.ALL2.clst.borrar
echo "CLUSTER" > $POP.ALL3.clst.borrar
cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 3,4,5 | rev  >> $POP.ALL3.clst.borrar
paste $POP.ALL1.clst.borrar $POP.ALL2.clst.borrar $POP.ALL3.clst.borrar> $POP.ALL.clst

rm $POP.*clst.borrar

# mv c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-h_ll_ba_n074.ALL.clst c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-h_ll_ba_n074.intergenic.ALL.clst

# Para distinguir que se ha hecho con mi intergénico. 

# run and plot

# Rscript $NGSTOOLS/Scripts/plotPCA.R -i Results/ALL.covar -c 1-2 -a Results/ALL.clst -o Results/ALL.pca.pdf
# Rscript $NGSTOOLS/Scripts/plotPCA.R -i $POP.covar -c 1-2 -a $POP.ALL.clst -o $POP.ALL.pca.pdf #--> No corre porque no puedo instalar librerías así que lo hago en mi propio ordenador
# evince Results/ALL.pca.pdf

# where the parameter -c 1-2 specifies that we are plotting only the first and second component. On the screen, you will see a series of numbers. These are the percentage of explained variance for each component.

scp mlucena@genomics-b.ebd.csic.es:/home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_PCA/*.ALL.clst /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/
scp mlucena@genomics-b.ebd.csic.es:/home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_PCA/*.covar /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/
# scp mlucena@genomics-a.ebd.csic.es:/opt/ngsTools_19092016/Scripts/plotPCA.R /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA


```


```{r "setup", include=FALSE}

library(methods)
library(optparse)
library(ggplot2)
library(dplyr)
library(threejs)
library(RColorBrewer)
require("knitr")

# knitr::opts_knit$set(root.dir = normalizePath("/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/test_intergenic_elena"))
# getwd()
# opts_knit$set(root.dir = normalizePath("/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/test_intergenic_elena"))

WORKING_DIR="/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/"
# samples_info <-read.csv("/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/test_intergenic_elena/c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-c_lp_do-c_lp_sm-h_ll_ba.ALL.clst")
# pop= c("c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-c_lp_do-c_lp_sm-h_ll_ba")
pop= c("c_ll_ka-c_ll_og-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya_n030")

# 
###########################################################################################
#for (pop in pops) 
#{

# Annotation file is in plink cluster format

#################################################################################

# Read input file
covar <- read.table(paste(WORKING_DIR,pop, ".covar", sep=''), stringsAsFact=F);

# Read annot file
annot <- read.table(paste(WORKING_DIR,pop, ".ALL.clst", sep=""), header=TRUE) 

# Parse components to analyze
# comp <- as.numeric(strsplit("component1-component2", "-", fixed=TRUE)[[1]])
# comp <- as.numeric(strsplit(opt$comp, "-", fixed=TRUE)[[1]])

comp <- c(1,2)

comp <- c(2,3)

comp <- c(1,3)


# Eigenvalues
eig <- eigen(covar, symm=TRUE);
eig$val <- eig$val/sum(eig$val);
cat(signif(eig$val, digits=3)*100,"\n");

# Plot
PC <- as.data.frame(eig$vectors)
colnames(PC) <- gsub("V", "PC", colnames(PC))
PC$Pop <- factor(annot$CLUSTER)
PC$Ind <- factor(annot$IID)
# PC <- PC %>% filter(Ind == "UA")



PC <- PC %>%  mutate(., 
         Subspecies = ifelse (Pop == "c_ll_po" | Pop == "c_ll_ki" | Pop == "c_ll_la" | Pop == "c_ll_no", "L.l.lynx",        
                          ifelse(Pop == "c_ll_ba" | Pop == "h_ll_ba", "L.l.balcanicus" ,
                          ifelse(Pop == "c_ll_cr","L.l.carpathicus",
                          ifelse(Pop == "c_ll_to" | Pop == "c_ll_tu" | Pop == "c_ll_ka", "L.l.kozlovi", 
                          ifelse(Pop == "c_ll_og", "L.l.isabellinus", 
                          ifelse(Pop == "c_ll_vl", "L.l.stroganovi", 
                          ifelse(Pop == "c_ll_ya", "L.l.wrangeli",     
                                 NA)))))))) %>% 
  mutate (., Populations =  ifelse (Pop == "c_ll_po", "Poland",
                          ifelse (Pop == "c_ll_ki", "Kirov",
                          ifelse (Pop == "c_ll_la", "Latvia",
                          ifelse (Pop == "c_ll_no", "Norway",
                          ifelse (Pop == "c_ll_ba" | Pop == "h_ll_ba", "Balkans" ,
                          ifelse (Pop == "c_ll_cr","Carpathians",
                          ifelse (Pop == "c_ll_to", "Tov",
                          ifelse (Pop == "c_ll_tu", "Tuva",
                          ifelse (Pop == "c_ll_ka", "Khentii-Aimag", 
                          ifelse (Pop == "c_ll_og", "Omnogovi", 
                          ifelse (Pop == "c_ll_vl", "Vladivostok", 
                          ifelse (Pop == "c_ll_ya", "Yakutia",
                          ifelse (Pop == "c_lp_sm", "Sierra Morena",
                          ifelse (Pop == "c_lp_do", "Doñana", NA))))))))))))))) %>% 
    mutate (., color =  ifelse (Populations == "Poland", brewer.pal(12,"Paired")[4], 
          ifelse (Populations == "Balkans", brewer.pal(12,"Paired")[2], 
          ifelse (Populations == "Carpathians", brewer.pal(12,"Paired")[1], 
          ifelse (Populations == "Kirov", brewer.pal(12,"Paired")[3], 
          ifelse (Populations == "Latvia", brewer.pal(12,"Paired")[9], 
          ifelse (Populations == "Norway", brewer.pal(12,"Paired")[10], 
          ifelse (Populations == "Tov", brewer.pal(12,"Paired")[5], 
          ifelse (Populations ==  "Tuva", brewer.pal(12,"Paired")[6], 
          ifelse (Populations == "Khentii-Aimag", brewer.pal(12,"Paired")[7], 
          ifelse (Populations == "Omnogovi", brewer.pal(12,"Paired")[8], 
          ifelse (Populations == "Vladivostok", brewer.pal(12,"Paired")[11], 
          ifelse (Populations == "Yakutia", brewer.pal(12,"Paired")[12],
          ifelse (Populations == "Sierra Morena", brewer.pal(8, "Greys") [5],
          ifelse (Populations == "Doñana", brewer.pal(8, "Greys") [8], NA))))))))))))))) %>% 
  mutate (., seq = ifelse (Ind == "c_ll_vl_0112" | Ind == "c_ll_vl_0112" | Ind == "c_ll_ya_0146" | Ind == "c_ll_cr_0212" | Ind == "c_ll_ki_0090" , "MACROGEN", "CNAG"))

                           
                           
                           
cols <- c("Poland"=brewer.pal(12,"Paired")[4], 
          "Balkans"=brewer.pal(12,"Paired")[2], 
          "Carpathians"=brewer.pal(12,"Paired")[1], 
          "Kirov"=brewer.pal(12,"Paired")[3], 
          "Latvia"=brewer.pal(12,"Paired")[9], 
          "Norway"=brewer.pal(12,"Paired")[10], 
          "Tov"=brewer.pal(12,"Paired")[5], 
          "Tuva"=brewer.pal(12,"Paired")[6], 
          "Khentii-Aimag"=brewer.pal(12,"Paired")[7], 
          "Omnogovi"=brewer.pal(12,"Paired")[8], 
          "Vladivostok"=brewer.pal(12,"Paired")[11], 
          "Yakutia"=brewer.pal(12,"Paired")[12],
          "Sierra Morena"=brewer.pal(8, "Greys") [5],
          "Doñana"=brewer.pal(8, "Greys") [8])

title <- paste("PC",comp[1]," (",signif(eig$val[comp[1]], digits=3)*100,"%)"," / PC",comp[2]," (",signif(eig$val[comp[2]], digits=3)*100,"%)",sep="",collapse="")

x_axis = paste("PC",comp[1],sep="")
y_axis = paste("PC",comp[2],sep="")

# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, color="Pop_label", shape="Pop"), size = 3) + 
#   scale_shape_manual(values=1:15) + ggtitle(title)
# # ggsave(paste(WORKING_DIR,pop, "_pca.TEST.eps", sep=""),units="mm",  width=270, height=190)
# # unlink("Rplots.pdf", force=TRUE)
# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, color="Pop", shape="Pop"), size = 3) + 
#     scale_colour_manual(values = cols)+
#   scale_shape_manual(values=1:15) + ggtitle(title)
# 
# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Pop", shape="subspecies"), size = 3) + 
#   scale_colour_manual(values = cols)+
#   scale_shape_manual(values=1:7)

ggplot() + 
  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Populations"), size = 5) + 
  scale_colour_manual(values = cols)+ ggtitle(title) + theme_classic()
 ggsave(paste(WORKING_DIR,pop, "_pca_pop_PC1_PC2.pdf", sep=""),units="mm",  width=270, height=190)
 
#ggplot() + 
#  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Populations", shape="seq"), size = 5) + 
#  scale_colour_manual(values = cols)+ ggtitle(title) + theme_classic()
 

# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, shape="subspecies"), size = 3) + 
#   scale_shape_manual(values=1:7)

ggplot() + 
  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Subspecies"), size = 5)  +
  ggtitle(title) + theme_classic()
  ggsave(paste(WORKING_DIR,pop, "_pca_subspecies_PC1_PC2.pdf", sep=""),units="mm",  width=270, height=190)

# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="subspecies"), size = 3)  +
#   scale_shape_manual(values=1:7)
# 
# ggsave(paste(WORKING_DIR,pop, "_pca.TEST.eps", sep=""),plot=a, units="mm",  width=270, height=190)
# 
# ggsave(paste(WORKING_DIR,pop, "_pca2.TEST.eps", sep=""),plot=b, units="mm",  width=270, height=190)
  
  
  

###########################################################################################
# 3D plot
###########################################################################################

scatterplot3js(PC$PC1, PC$PC2, PC$PC3, 
                col=PC$color,  labels=PC$Ind,
               size=0.7, grid=FALSE, renderer = "canvas", axis =F
)



```