---
title: "pipeline_PCA"
output: html_document
---

# Population Structure (PCA, MDS(ngsDist))
María Lucena Pérez


Info from ANGSD page, and github Fumagalli/ngsTools/TUTORIAL.md
One solution would be to perform a Principal Component Analysis (PCA) or a Multidimensional Scaling (MDS) or some clustering based on genetic distances among samples. 
We are here showing how to perform these analyses using ngsTools/ANGSD in case of low-depth data.

Principal Component Analysis 

Poblaciones 


Both sps, subsample, only intergenic:

c_lp_do-c_lp_sm-c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n111

Lynx lynx, subsample, only intergenic:

c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080

Lynx lynx, oriental populations:

c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n030

Lynx lynx, wo balcans
c_ll_no-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n077

```{bash}

# We are now performing a principal component analyses (PCA) without relying on called genotypes, but rather by taking their uncertainty into account. 
# More specifically, the next program we are going to use (ngsTools) takes as input genotype probabilities in binary format, 
# so we need to specify -doGeno 32. Also, we are using a HWE-based prior with -doPost 1. 
# Recalling also our choice for data filtering, our command line is:

# Change directory

cd /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_PCA

# I want to produce PCAs for the whole dataset, each specie, and within iberian with and without historical samples. 
# I think that per population PCAs make no sense

#To launch one by one ( because otherwise it's too slow I'll launch single pops in a and groups in b)
#Lanzo cada uno en una sesión de screen diferente y con un log distinto ( mediante script). lanzo el script conpleto dandole como input el csv que acabo de crear. Ejemplo:

screen -S pca_lynx_lynx_wobalcans
script pca_lynx_lynx_wobalcans.log

POP="c_ll_no-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl"  # <--CHANGE POP HERE


read POP mean sd mean_truncated sd_truncated maxDepth minDepth maxDepth_truncated minDepth_truncated < /home/mlucena/ANGSD_analysis/depth_calculus/"$POP"_n*_mean_sd_depthGlobal_lynx_per_pop_mean_folds_0.95.csv


# Set paths to software and some commom files

ANGSD=/opt/angsd/angsd
NGSTOOLS="/opt/ngsTools"
THREADS=20                     # no. of computer cores used by bwa and samtools. 20 = OK, >20 = ask people first!
REF="/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23_without_repetitive_transposable_low_complexity.fa"
ANC="/home/GRUPOS/grupolince/reference_genomes/lynx_rufus_genome/c_lr_zz_0001_recal1.fa"
FILTER1=" -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -baq 1 -C 50 "
FILTER2=" -minMapQ 30 -minQ 20 -doCounts 1 "
N_IND=$(echo ${POP: -3} )
MIN_IND=$(expr $N_IND / 2) 
N_ALLELES=$(expr $N_IND \* 2)
SNP_pval_filter=$(echo "1/(2*$N_IND)" | bc -l | awk '{printf "%.10f\n", $0}')
# SNP_pval = #ind*(2allelos/ind) = x alelos ; 1/x = y  # esto sería un filtro aceptable. En mi caso el dataset más pequeño (h_lypa_all) es de 55 indivíduos (1/(2*55)=9e-3) y el más grande (x_lyxx_all) de 189 (1/(2*189)=2.6e-3)


# Sanity checks:

echo $POP
echo $N_IND
echo $MIN_IND
echo $maxDepth
echo $minDepth
echo $SNP_pval_filter


### Calling genotype likelihoods: 
 
$ANGSD/angsd -P $THREADS -b /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP.intergenic.bamlist -ref $REF -out $POP.intergenic \
$FILTER1 $FILTER2 \
-minInd $MIN_IND -setMaxDepth $maxDepth -setMinDepth $minDepth  \
-GL 1 -doMajorMinor 1 -doMaf 1 -skipTriallelic 1 \
-SNP_pval $SNP_pval_filter \
-doGeno 32 -doPost 1


# -setMinDepthInd

# doPost 1 uses the estimates perstie allele frequency as a rpior for genotyped proportions, assuming HWE. 


### URGENTE ULTIMA ACTUALIZACION!!

# -minInd [int]
# Only keep sites with at least minIndDepth (default is 1) from at least [int] individuals
# -minIndDepth [int]
# Only works with the -minInd filter. Change the minimum depth the individuals must have in order to keep the site. Default is 1.


# SNP_pval = #ind*(2allelos/ind) = x alelos ; 1/x = y  # esto sería un filtro aceptable. En mi caso el dataset más pequeño (h_lypa_all) es de 77 indivíduos (1/(2*77)=9e-3) y el más grande (x_lyxx_all) de 189 (1/(2*189)=2.6e-3)
# Para el PCA vamos a usar 1e-3 porque es un filtro suuuper light.

### OJO! NO ES TAN SUPER LIGHT!!!


# Unzip the results (but you cannot open it since it is in binary format)

gunzip $POP.intergenic.geno.gz

### ngsCovar

# We are going to use ngsCovar, which estimates the covariance matrix between individuals based on genotype probabilities. 
# Then this matrix will be decomposed into principal components which will be investigated for population structure analyses. 
# Note that although ngsCovar can account for SNPs uncertanity, we find that it is faster to perform a light SNP filtering first 
# (as we did using -SNP_pval 1e-3) than using all sites.

# If you type:

# $NGSTOOLS/ngsPopGen/ngsCovar

# you will see a list of possible options. 
# For instance, we need to define how many sites we have. To retrieve such value, we can inspect the file with allele frequencies:

#less -S $POP.mafs.gz
N_SITES=`zcat $POP.intergenic.mafs.gz | tail -n+2 | wc -l` 

# Hago tail -n+2 porque tengo cabecero. 

echo $N_SITES 


# I have 3636950 sites lynx lynx + lynx pardinus. Oct 2017.
# I have 2096552 sites only lynx lynx. Oct 2017.
# I have 1773532 sites only lynx lynx oriental. Oct 2017.
# I have 1488015 sites only lynx lynx west. Nov 2017.



# Does it make sense?


# Now we can perform a PCA by estimating the covariance matrix first:

$NGSTOOLS/ngsPopGen/ngsCovar -probfile $POP.*geno -outfile $POP.covar -nind $N_IND  -nsites $N_SITES -call 0 -norm 0 &> /dev/null

#-minmaf FLOAT: ignore sites below this threshold of minor allele frequency

# with the options -call 0 meaning that we do not perform genotype calling 
# and -norm 0 that we are not normalising by allele frequency. 
# The latter may give more weight to low frequency variants which are harder to estimate.

# Look at the output file:

#less -S $POP.covar

# which represents a matrix of NxN with N individuals giving the covariance. Note that this matrix is symmetric.

# Finally, we perform an eigenvector decomposition and plot the resulting map:

# create a cluster-like file defining the labelling (population) for each sample

#Example: Rscript -e 'write.table(cbind(seq(1,60),rep(1,60),c(rep("LWK",20),rep("TSI",20),rep("PEL",20))), row.names=F, sep=" ", col.names=c("FID","IID","CLUSTER"), file="Results/ALL.clst", quote=F)'

echo "FID" > $POP.ALL1.clst.borrar
seq $N_IND >> $POP.ALL1.clst.borrar
echo "IID" > $POP.ALL2.clst.borrar
# yes "1" | head -n $N_IND >> $POP.ALL2.clst.borrar
cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 2,3,4,5 | rev  >> $POP.ALL2.clst.borrar
echo "CLUSTER" > $POP.ALL3.clst.borrar
cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 3,4,5 | rev  >> $POP.ALL3.clst.borrar
paste $POP.ALL1.clst.borrar $POP.ALL2.clst.borrar $POP.ALL3.clst.borrar> $POP.ALL.clst

rm $POP.*clst.borrar

# run and plot

# Rscript $NGSTOOLS/Scripts/plotPCA.R -i Results/ALL.covar -c 1-2 -a Results/ALL.clst -o Results/ALL.pca.pdf
# Rscript $NGSTOOLS/Scripts/plotPCA.R -i $POP.covar -c 1-2 -a $POP.ALL.clst -o $POP.ALL.pca.pdf #--> No corre porque no puedo instalar librerías así que lo hago en mi propio ordenador
# evince Results/ALL.pca.pdf

# where the parameter -c 1-2 specifies that we are plotting only the first and second component. On the screen, you will see a series of numbers. These are the percentage of explained variance for each component.

```

## Copying to Dropbox folder


```{bash}
scp mlucena@genomics-b.ebd.csic.es:/home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_PCA/*.ALL.clst /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_PCA/

scp mlucena@genomics-b.ebd.csic.es:/home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_PCA/*.covar /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_PCA/

# scp mlucena@genomics-a.ebd.csic.es:/opt/ngsTools_19092016/Scripts/plotPCA.R /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA


```


## R representation

```{r "setup", include=FALSE}
install.packages("optparse")
library(methods)
library(optparse)
library(ggplot2)
library(dplyr)
library(threejs)
library(RColorBrewer)
require("knitr")

# knitr::opts_knit$set(root.dir = normalizePath("/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/test_intergenic_elena"))
# getwd()
# opts_knit$set(root.dir = normalizePath("/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/test_intergenic_elena"))

WORKING_DIR="/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_PCA/"
# samples_info <-read.csv("/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/test_intergenic_elena/c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-c_lp_do-c_lp_sm-h_ll_ba.ALL.clst")
# pop=c("c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080")
pop=c("c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur_n050")
# pop= c("c_lp_do-c_lp_sm-c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n111")
pop=c("c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n030")

# 
###########################################################################################
#for (pop in pops) 
#{

# Annotation file is in plink cluster format

#################################################################################

# Read input file
covar <- read.table(paste(WORKING_DIR,pop, ".covar", sep=''), stringsAsFact=F);

# Read annot file
annot <- read.table(paste(WORKING_DIR,pop, ".ALL.clst", sep=""), header=TRUE) 

# Parse components to analyze
# comp <- as.numeric(strsplit("component1-component2", "-", fixed=TRUE)[[1]])
# comp <- as.numeric(strsplit(opt$comp, "-", fixed=TRUE)[[1]])

# comp <- c(1,2)
# comp <- c(2,3)
# comp <- c(1,3)


# Eigenvalues
eig <- eigen(covar, symm=TRUE);
eig$val <- eig$val/sum(eig$val);
cat(signif(eig$val, digits=3)*100,"\n");
eig$values


# Plot
PC <- as.data.frame(eig$vectors)
colnames(PC) <- gsub("V", "PC", colnames(PC))
PC$Pop <- factor(annot$CLUSTER)
PC$Ind <- factor(annot$IID)
# PC <- PC %>% filter(Ind == "UA")



PC <- PC %>%  mutate(., 
         Subspecies = ifelse (Pop == "c_ll_po" | Pop == "c_ll_ki" | Pop == "c_ll_la" | Pop == "c_ll_no"  | Pop == "c_ll_ur", "L.l.lynx",        
                          ifelse(Pop == "c_ll_ba" | Pop == "h_ll_ba", "L.l.balcanicus" ,
                          ifelse(Pop == "c_ll_cr","L.l.carpathicus",
                          ifelse(Pop == "c_ll_to" | Pop == "c_ll_tu" | Pop == "c_ll_ka", "L.l.kozlovi", 
                          ifelse(Pop == "c_ll_og", "L.l.isabellinus", 
                          ifelse(Pop == "c_ll_vl", "L.l.stroganovi", 
                          ifelse(Pop == "c_ll_ya", "L.l.wrangeli",     
                                 NA)))))))) %>% 
  mutate (., Populations =  ifelse (Pop == "c_ll_po", "Bialowieza",
                          ifelse (Pop == "c_ll_ur", "Urals",
                          ifelse (Pop == "c_ll_ki", "Kirov",
                          ifelse (Pop == "c_ll_la", "Latvia",
                          ifelse (Pop == "c_ll_no", "Norway",
                          ifelse (Pop == "c_ll_ba" | Pop == "h_ll_ba", "Balkans" ,
                          ifelse (Pop == "c_ll_cr","Carpathians",
                          ifelse (Pop == "c_ll_to", "Töv",
                          ifelse (Pop == "c_ll_tu", "Tuva",
                          ifelse (Pop == "c_ll_ka", "Khentii-Aimag", 
                          ifelse (Pop == "c_ll_og", "Ömnögovi", 
                          ifelse (Pop == "c_ll_vl", "Vladivostok", 
                          ifelse (Pop == "c_ll_ya", "Yakutia",
                          ifelse (Pop == "c_lp_sm", "Sierra Morena",
                          ifelse (Pop == "c_lp_do", "Doñana", NA)))))))))))))))) %>% 
    mutate (., color =  ifelse (Populations == "Bialowieza", brewer.pal(12,"Paired")[4], 
                      ifelse (Populations == "Urals", brewer.pal(11,"BrBG")[9], 
                      ifelse (Populations == "Balkans", brewer.pal(12,"Paired")[10], 
                      ifelse (Populations == "Carpathians", brewer.pal(12,"Paired")[9], 
                      ifelse (Populations == "Kirov", brewer.pal(12,"Paired")[1], 
                      ifelse (Populations == "Latvia", brewer.pal(12,"Paired")[3], 
                      ifelse (Populations == "Norway", brewer.pal(12,"Paired")[2], 
                      ifelse (Populations == "Töv", brewer.pal(12,"Paired")[12], 
                      ifelse (Populations == "Tuva", brewer.pal(12,"Paired")[8], 
                      ifelse (Populations == "Khentii-Aimag", brewer.pal(12,"Paired")[7], 
                      ifelse (Populations == "Ömnögovi", brewer.pal(12,"Paired")[11], 
                      ifelse (Populations == "Vladivostok", brewer.pal(12,"Paired")[5], 
                      ifelse (Populations == "Yakutia", brewer.pal(12,"Paired")[6],
                      ifelse (Populations == "Sierra Morena", brewer.pal(8, "Greys") [5],
                      ifelse (Populations == "Doñana", brewer.pal(8, "Greys") [8], NA)))))))))))))))) %>% 
  mutate (., seq = ifelse (Ind == "c_ll_vl_0112" | Ind == "c_ll_vl_0112" | Ind == "c_ll_ya_0146" | Ind == "c_ll_cr_0212" | Ind == "c_ll_ki_0090" , "MACROGEN", "CNAG"))

                           
                           
                           
cols <- c("Bialowieza"=brewer.pal(12,"Paired")[4], 
          "Urals"=brewer.pal(11,"BrBG")[9], 
          "Balkans"=brewer.pal(12,"Paired")[10], 
          "Carpathians"=brewer.pal(12,"Paired")[9], 
          "Kirov"=brewer.pal(12,"Paired")[1], 
          "Latvia"=brewer.pal(12,"Paired")[3], 
          "Norway"=brewer.pal(12,"Paired")[2], 
          "Töv"=brewer.pal(12,"Paired")[12], 
          "Tuva"=brewer.pal(12,"Paired")[8], 
          "Khentii-Aimag"=brewer.pal(12,"Paired")[7], 
          "Ömnögovi"=brewer.pal(12,"Paired")[11], 
          "Vladivostok"=brewer.pal(12,"Paired")[5], 
          "Yakutia"=brewer.pal(12,"Paired")[6],
          "Sierra Morena"=brewer.pal(8, "Greys") [5],
          "Doñana"=brewer.pal(8, "Greys") [8])

PC$Populations <- factor (PC$Populations, levels=c("Norway", "Balkans", "Carpathians", "Bialowieza", "Latvia", "Kirov", "Urals", "Tuva", "Töv", "Khentii-Aimag", "Ömnögovi",  "Yakutia", "Vladivostok"))


title <- paste("PC",comp[1]," (",signif(eig$val[comp[1]], digits=3)*100,"%)"," / PC",comp[2]," (",signif(eig$val[comp[2]], digits=3)*100,"%)",sep="",collapse="")

x_axis = paste("PC",comp[1],sep="")
y_axis = paste("PC",comp[2],sep="")

# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, color="Pop_label", shape="Pop"), size = 3) + 
#   scale_shape_manual(values=1:15) + ggtitle(title)
# # ggsave(paste(WORKING_DIR,pop, "_pca.TEST.eps", sep=""),units="mm",  width=270, height=190)
# # unlink("Rplots.pdf", force=TRUE)
# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, color="Pop", shape="Pop"), size = 3) + 
#     scale_colour_manual(values = cols)+
#   scale_shape_manual(values=1:15) + ggtitle(title)
# 
# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Pop", shape="subspecies"), size = 3) + 
#   scale_colour_manual(values = cols)+
#   scale_shape_manual(values=1:7)

ggplot() + 
geom_point(data=PC, aes_string(x=x_axis, y=y_axis, color="Populations"), size = 5) + 
scale_colour_manual(values = cols) + ggtitle(title) + theme_classic() 
ggsave(paste(WORKING_DIR,pop, "_pca_pop_PC1_PC2.pdf", sep=""),units="mm",  width=270, height=190)
 





 ggplot() + 
  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Populations"), size = 5) + 
  scale_colour_manual(values = cols)+ ggtitle(title) + theme_classic() +
  scale_y_reverse()
 ggsave(paste(WORKING_DIR,pop, "_pca_pop_PC1_PC2_reverse_axis.pdf", sep=""),units="mm",  width=270, height=190)

 
 
 
 
 # SECEM plot #
 ggplot() + 
  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, color="Populations"), size = 5) + 
#  scale_colour_manual(values = cols)+ ggtitle(title) + theme_classic() +
   scale_color_manual("Población",labels = c("Noruega","Balcanes", "Carpatos", "Bialowieza","Letonia","Kirov","Urales","Tuva", "Töv","Khentii","Ömnögovi", "Yakutia", "Vladivostok"), values = cols) + theme_classic() + ggtitle(title) +
   scale_y_reverse() +
theme (legend.text=element_text(size=18),legend.title = element_text(size=23), axis.title = element_text(size = 20), axis.text = element_text(size=18), title=element_text(size=23))
 
ggsave(paste(WORKING_DIR,pop, "_pca_pop_PC1_PC2_reverse_axis_spanish.pdf", sep=""),  width=270, height=190,units="mm")



 
#ggplot() + 
#  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Populations", shape="seq"), size = 5) + 
#  scale_colour_manual(values = cols)+ ggtitle(title) + theme_classic()

# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, shape="subspecies"), size = 3) + 
#   scale_shape_manual(values=1:7)

ggplot() + 
  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Subspecies"), size = 5)  +
  ggtitle(title) + theme_classic() + theme(text = element_text(size=20)) 
  ggsave(paste(WORKING_DIR,pop, "_pca_subspecies_PC1_PC2.pdf", sep=""),units="mm",  width=270, height=190)

  
  # Article
   ggplot() + 
  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Populations"), size = 5) + 
  scale_colour_manual(values = cols)+ ggtitle(title) + theme_classic() +
  scale_y_reverse() +
     theme(legend.position="top")
   
 ggsave(paste(WORKING_DIR,pop, "_pca_pop_PC1_PC2_reverse_axis.pdf", sep=""),units="mm",  width=270, height=190,  useDingbats=FALSE)


  
# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="subspecies"), size = 3)  +
#   scale_shape_manual(values=1:7)
# 
# ggsave(paste(WORKING_DIR,pop, "_pca.TEST.eps", sep=""),plot=a, units="mm",  width=270, height=190)
# 
# ggsave(paste(WORKING_DIR,pop, "_pca2.TEST.eps", sep=""),plot=b, units="mm",  width=270, height=190)
  

###########################################################################################
# 3D plot
###########################################################################################

scatterplot3js(PC$PC1, PC$PC2, PC$PC3, 
                col=PC$color,  labels=PC$Ind,
               size=0.7, grid=FALSE, renderer = "canvas", axis =F)

scatterplot3js(PC$PC1, PC$PC2, PC$PC3, 
                col=PC$color,  labels=PC$Ind,
               size=0.7)



```

## Adding lynx lynx form genome project to know procedence.

Hago el bamlist a mano y no corro la cobertura porque no voy a aplicar ningún filtro. Es sólo para tener esta información.

```{bash}

mkdir /home/mlucena/ANGSD_analysis/ancestry_ll23A
cd /home/mlucena/ANGSD_analysis/ancestry_ll23A

nano c_ll_zz-c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n081.bamlist

/home/mlucena/grupolince/lynx_genomes_5x/raw_data/genome_project_raw_data_lynx_lynx/c_ll_zz_023A_sorted_rmdup_sorted_indelrealigner.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_no_0075_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_no_0076_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_no_0077_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_no_0078_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_no_0079_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_no_0080_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_no_0081_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_no_0082_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ba_0216_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/h_ll_ba_0214_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/h_ll_ba_0215_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_cr_0205_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_cr_0206_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_cr_0207_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_cr_0208_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_cr_0209_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_cr_0212_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_po_0001_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_po_0002_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_po_0003_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_po_0011_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_po_0014_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_po_0019_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_po_0105_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_po_0106_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_la_0044_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_la_0047_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_la_0048_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_la_0052_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_la_0053_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_la_0054_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0090_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0091_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0092_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0093_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0094_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0095_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0096_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0097_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0098_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0099_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0100_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0101_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ki_0102_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ur_0194_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ur_0195_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ur_0196_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ur_0199_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ur_0200_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ur_0203_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_tu_0153_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_tu_0157_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_tu_0158_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_tu_0159_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_tu_0165_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_tu_0166_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_to_0190_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_to_0191_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ka_0184_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ka_0186_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ka_0188_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ka_0189_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_og_0181_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_og_0187_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ya_0138_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ya_0139_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ya_0140_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ya_0142_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ya_0143_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ya_0145_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ya_0146_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_ya_0147_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_vl_0107_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_vl_0108_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_vl_0109_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_vl_0110_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_vl_0112_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_vl_0113_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_vl_0128_intergenic.bam
/home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final/BAM_intergenic/c_ll_vl_0132_intergenic.bam



POP=c_ll_zz-c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n081

ANGSD=/opt/angsd/angsd
NGSTOOLS="/opt/ngsTools"
THREADS=20                     # no. of computer cores used by bwa and samtools. 20 = OK, >20 = ask people first!
REF="/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23_without_repetitive_transposable_low_complexity.fa"
ANC="/home/GRUPOS/grupolince/reference_genomes/lynx_rufus_genome/c_lr_zz_0001_recal1.fa"
FILTER1=" -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -baq 1 -C 50 "
FILTER2=" -minMapQ 30 -minQ 20 -doCounts 1 "
N_IND=$(echo ${POP: -3} )
MIN_IND=$(expr $N_IND / 2) 
REGIONFILE="/home/mlucena/ANGSD_analysis/depth_calculus/no_genes_Lypa_10000longest_center_final_slop20_dot.rf"




# Sanity checks:

echo $POP
echo $N_IND
echo $MIN_IND
echo $maxDepth
echo $minDepth


### Calling genotype likelihoods: 
 
$ANGSD/angsd -P $THREADS -b $POP.bamlist -ref $REF -out $POP.rf \
$FILTER1 $FILTER2 \
-rf $REGIONFILE \
-GL 1 -doMajorMinor 1 -doMaf 1 -skipTriallelic 1 \
-doGeno 32 -doPost 1


# -setMinDepthInd

# doPost 1 uses the estimates perstie allele frequency as a rpior for genotyped proportions, assuming HWE. 


### URGENTE ULTIMA ACTUALIZACION!!

# -minInd [int]
# Only keep sites with at least minIndDepth (default is 1) from at least [int] individuals
# -minIndDepth [int]
# Only works with the -minInd filter. Change the minimum depth the individuals must have in order to keep the site. Default is 1.


# SNP_pval = #ind*(2allelos/ind) = x alelos ; 1/x = y  # esto sería un filtro aceptable. En mi caso el dataset más pequeño (h_lypa_all) es de 55 indivíduos (1/(2*55)=9e-3) y el más grande (x_lyxx_all) de 189 (1/(2*189)=2.6e-3)
# Para el PCA vamos a usar 1e-3 porque es un filtro suuuper light.

### OJO! NO ES TAN SUPER LIGHT!!!


# Unzip the results (but you cannot open it since it is in binary format)

gunzip $POP.rf.geno.gz

### ngsCovar

# We are going to use ngsCovar, which estimates the covariance matrix between individuals based on genotype probabilities. 
# Then this matrix will be decomposed into principal components which will be investigated for population structure analyses. 
# Note that although ngsCovar can account for SNPs uncertanity, we find that it is faster to perform a light SNP filtering first 
# (as we did using -SNP_pval 1e-3) than using all sites.

# If you type:

# $NGSTOOLS/ngsPopGen/ngsCovar

# you will see a list of possible options. 
# For instance, we need to define how many sites we have. To retrieve such value, we can inspect the file with allele frequencies:

#less -S $POP.mafs.gz
N_SITES=`zcat $POP.rf.mafs.gz | tail -n+2 | wc -l`
echo $N_SITES 

# 1640662 ¡¡¡¡¡¡Son muchos!!!!!

# Does it make sense?


# Now we can perform a PCA by estimating the covariance matrix first:

$NGSTOOLS/ngsPopGen/ngsCovar -probfile $POP.*geno -outfile $POP.covar -nind $N_IND  -nsites $N_SITES -call 0 -norm 0 &> /dev/null

#-minmaf FLOAT: ignore sites below this threshold of minor allele frequency

# with the options -call 0 meaning that we do not perform genotype calling 
# and -norm 0 that we are not normalising by allele frequency. 
# The latter may give more weight to low frequency variants which are harder to estimate.

# Look at the output file:

#less -S $POP.covar

# which represents a matrix of NxN with N individuals giving the covariance. Note that this matrix is symmetric.

# Finally, we perform an eigenvector decomposition and plot the resulting map:

# create a cluster-like file defining the labelling (population) for each sample

#Example: Rscript -e 'write.table(cbind(seq(1,60),rep(1,60),c(rep("LWK",20),rep("TSI",20),rep("PEL",20))), row.names=F, sep=" ", col.names=c("FID","IID","CLUSTER"), file="Results/ALL.clst", quote=F)'

echo "FID" > $POP.ALL1.clst.borrar
seq $N_IND >> $POP.ALL1.clst.borrar
echo "IID" > $POP.ALL2.clst.borrar
# yes "1" | head -n $N_IND >> $POP.ALL2.clst.borrar
cat $POP.bamlist | rev | cut -d "/"  -f 1 | rev | colrm 13  >> $POP.ALL2.clst.borrar
echo "CLUSTER" > $POP.ALL3.clst.borrar
cat $POP.bamlist | rev | cut -d "/"  -f 1 | rev | colrm 8  >> $POP.ALL3.clst.borrar
paste $POP.ALL1.clst.borrar $POP.ALL2.clst.borrar $POP.ALL3.clst.borrar> $POP.ALL.clst

rm $POP.*clst.borrar

# run and plot

# Rscript $NGSTOOLS/Scripts/plotPCA.R -i Results/ALL.covar -c 1-2 -a Results/ALL.clst -o Results/ALL.pca.pdf
# Rscript $NGSTOOLS/Scripts/plotPCA.R -i $POP.covar -c 1-2 -a $POP.ALL.clst -o $POP.ALL.pca.pdf #--> No corre porque no puedo instalar librerías así que lo hago en mi propio ordenador
# evince Results/ALL.pca.pdf

# where the parameter -c 1-2 specifies that we are plotting only the first and second component. On the screen, you will see a series of numbers. These are the percentage of explained variance for each component.

```

## Copying to Dropbox folder


```{bash}
scp mlucena@genomics-b.ebd.csic.es://home/mlucena/ANGSD_analysis/ancestry_ll23A/*.ALL.clst  /Users/marialucenaperez/Dropbox/PhD/PCA_ll23A_ancestry/
scp mlucena@genomics-b.ebd.csic.es:/home/mlucena/ANGSD_analysis/ancestry_ll23A/*.covar /Users/marialucenaperez/Dropbox/PhD/PCA_ll23A_ancestry

```


## R representation

```{r "setup", include=FALSE}

library(methods)
library(optparse)
library(ggplot2)
library(dplyr)
library(threejs)
library(RColorBrewer)
require("knitr")

# knitr::opts_knit$set(root.dir = normalizePath("/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/test_intergenic_elena"))
# getwd()
# opts_knit$set(root.dir = normalizePath("/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/test_intergenic_elena"))

WORKING_DIR="/Users/marialucenaperez/Dropbox/PhD/PCA_ll23A_ancestry/"

pop=c("c_ll_zz-c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n081")


 
###########################################################################################
#for (pop in pops) 
#{

# Annotation file is in plink cluster format

#################################################################################

# Read input file
covar <- read.table(paste(WORKING_DIR,pop, ".covar", sep=''), stringsAsFact=F);

# Read annot file
annot <- read.table(paste(WORKING_DIR,pop, ".ALL.clst", sep=""), header=TRUE) 

# Parse components to analyze
# comp <- as.numeric(strsplit("component1-component2", "-", fixed=TRUE)[[1]])
# comp <- as.numeric(strsplit(opt$comp, "-", fixed=TRUE)[[1]])

comp <- c(1,2)
# comp <- c(2,3)
# comp <- c(1,3)


# Eigenvalues
eig <- eigen(covar, symm=TRUE);
eig$val <- eig$val/sum(eig$val);
cat(signif(eig$val, digits=3)*100,"\n");

# Plot
PC <- as.data.frame(eig$vectors)
colnames(PC) <- gsub("V", "PC", colnames(PC))
PC$Pop <- factor(annot$CLUSTER)
PC$Ind <- factor(annot$IID)
# PC <- PC %>% filter(Ind == "UA")



PC <- PC %>%  mutate(., 
         Subspecies = ifelse (Pop == "c_ll_po" | Pop == "c_ll_ki" | Pop == "c_ll_la" | Pop == "c_ll_no"  | Pop == "c_ll_ur", "L.l.lynx",        
                          ifelse(Pop == "c_ll_ba" | Pop == "h_ll_ba", "L.l.balcanicus" ,
                          ifelse(Pop == "c_ll_cr","L.l.carpathicus",
                          ifelse(Pop == "c_ll_to" | Pop == "c_ll_tu" | Pop == "c_ll_ka", "L.l.kozlovi", 
                          ifelse(Pop == "c_ll_og", "L.l.isabellinus", 
                          ifelse(Pop == "c_ll_vl", "L.l.stroganovi", 
                          ifelse(Pop == "c_ll_ya", "L.l.wrangeli",     
                                 NA)))))))) %>% 
  mutate (., Populations =  ifelse (Pop == "c_ll_po", "Poland",
                          ifelse (Pop == "c_ll_ur", "Urals",
                          ifelse (Pop == "c_ll_ki", "Kirov",
                          ifelse (Pop == "c_ll_la", "Latvia",
                          ifelse (Pop == "c_ll_no", "Norway",
                          ifelse (Pop == "c_ll_ba" | Pop == "h_ll_ba", "Balkans" ,
                          ifelse (Pop == "c_ll_cr","Carpathians",
                          ifelse (Pop == "c_ll_to", "Tov",
                          ifelse (Pop == "c_ll_tu", "Tuva",
                          ifelse (Pop == "c_ll_ka", "Khentii-Aimag", 
                          ifelse (Pop == "c_ll_og", "Omnogovi", 
                          ifelse (Pop == "c_ll_vl", "Vladivostok", 
                          ifelse (Pop == "c_ll_ya", "Yakutia",
                          ifelse (Pop == "c_ll_zz", "Zoo",NA))))))))))))))) %>% 
    mutate (., color =  ifelse (Populations == "Poland", brewer.pal(12,"Paired")[4], 
                      ifelse (Populations == "Urals", brewer.pal(9,"RdPu")[6], 
                      ifelse (Populations == "Balkans", brewer.pal(12,"Paired")[10], 
                      ifelse (Populations == "Carpathians", brewer.pal(12,"Paired")[9], 
                      ifelse (Populations == "Kirov", brewer.pal(12,"Paired")[1], 
                      ifelse (Populations == "Latvia", brewer.pal(12,"Paired")[3], 
                      ifelse (Populations == "Norway", brewer.pal(12,"Paired")[2], 
                      ifelse (Populations == "Tov", brewer.pal(12,"Paired")[12], 
                      ifelse (Populations == "Tuva", brewer.pal(12,"Paired")[8], 
                      ifelse (Populations == "Khentii-Aimag", brewer.pal(12,"Paired")[7], 
                      ifelse (Populations == "Omnogovi", brewer.pal(12,"Paired")[11], 
                      ifelse (Populations == "Vladivostok", brewer.pal(12,"Paired")[5], 
                      ifelse (Populations == "Yakutia", brewer.pal(12,"Paired")[6],
                      ifelse (Populations == "Zoo", brewer.pal(8, "Greys") [5], NA))))))))))))))) %>% 
  mutate (., seq = ifelse (Ind == "c_ll_vl_0112" | Ind == "c_ll_vl_0112" | Ind == "c_ll_ya_0146" | Ind == "c_ll_cr_0212" | Ind == "c_ll_ki_0090" , "MACROGEN", "CNAG"))

                           
                           
                           
cols <- c("Poland"=brewer.pal(12,"Paired")[4], 
          "Urals"=brewer.pal(9,"RdPu")[6], 
          "Balkans"=brewer.pal(12,"Paired")[10], 
          "Carpathians"=brewer.pal(12,"Paired")[9], 
          "Kirov"=brewer.pal(12,"Paired")[1], 
          "Latvia"=brewer.pal(12,"Paired")[3], 
          "Norway"=brewer.pal(12,"Paired")[2], 
          "Tov"=brewer.pal(12,"Paired")[12], 
          "Tuva"=brewer.pal(12,"Paired")[8], 
          "Khentii-Aimag"=brewer.pal(12,"Paired")[7], 
          "Omnogovi"=brewer.pal(12,"Paired")[11], 
          "Vladivostok"=brewer.pal(12,"Paired")[5], 
          "Yakutia"=brewer.pal(12,"Paired")[6],
          "Zoo"=brewer.pal(8, "Greys") [5])

title <- paste("PC",comp[1]," (",signif(eig$val[comp[1]], digits=3)*100,"%)"," / PC",comp[2]," (",signif(eig$val[comp[2]], digits=3)*100,"%)",sep="",collapse="")

x_axis = paste("PC",comp[1],sep="")
y_axis = paste("PC",comp[2],sep="")

# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, color="Pop_label", shape="Pop"), size = 3) + 
#   scale_shape_manual(values=1:15) + ggtitle(title)
# # ggsave(paste(WORKING_DIR,pop, "_pca.TEST.eps", sep=""),units="mm",  width=270, height=190)
# # unlink("Rplots.pdf", force=TRUE)
# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, color="Pop", shape="Pop"), size = 3) + 
#     scale_colour_manual(values = cols)+
#   scale_shape_manual(values=1:15) + ggtitle(title)
# 
# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Pop", shape="subspecies"), size = 3) + 
#   scale_colour_manual(values = cols)+
#   scale_shape_manual(values=1:7)

ggplot() + 
  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Populations"), size = 5) + 
  scale_colour_manual(values = cols)+ ggtitle(title) + theme_classic()
 ggsave(paste(WORKING_DIR,pop, "_pca_pop_PC1_PC2.pdf", sep=""),units="mm",  width=270, height=190)

#ggplot() + 
#  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Populations", shape="seq"), size = 5) + 
#  scale_colour_manual(values = cols)+ ggtitle(title) + theme_classic()

# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, shape="subspecies"), size = 3) + 
#   scale_shape_manual(values=1:7)

#ggplot() + 
#  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Subspecies"), size = 5)  +
#  ggtitle(title) + theme_classic()
#  ggsave(paste(WORKING_DIR,pop, "_pca_subspecies_PC1_PC2.pdf", sep=""),units="mm",  width=270, height=190)

# ggplot() + 
#   geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="subspecies"), size = 3)  +
#   scale_shape_manual(values=1:7)
# 
# ggsave(paste(WORKING_DIR,pop, "_pca.TEST.eps", sep=""),plot=a, units="mm",  width=270, height=190)
# 
# ggsave(paste(WORKING_DIR,pop, "_pca2.TEST.eps", sep=""),plot=b, units="mm",  width=270, height=190)
  

###########################################################################################
# 3D plot
###########################################################################################

scatterplot3js(PC$PC1, PC$PC2, PC$PC3, 
                col=PC$color,  labels=PC$Ind,
               size=0.7, grid=FALSE, renderer = "canvas", axis =F)



```


```


