---
title: "4.FST_intergenic"
output: html_document
---


# Population Genetic Differenciation (Fst)

Our program can estimate fst between populations. 
And has been generalized to give all pairwise fst estimates if you supply the command with multiple populations.
if you supply 3 populations, the program will also output the pbs statistic.

NB we have removed the very unusefull unweighted fst estimator in the output, and have included a header. The output example below will be updated at some point.

The procedure is

1 - Use angsd for calculating saf files for each population
2 - Use realSFS to calculate 2d sfs for each pair
3 - Use the above calculated 2dsfs as priors jointly with all safs from step1 to calculate fst binary files
4 - Use realSFS to extract the the fst values from the fst


## 1 - Estimate multidimensional SFS: Calculate all pairwise 2dsfs's

An important issue when doing this is to be sure that we are comparing the exactly same corresponding sites between populations. 
ANGSD does that automatically and considers only a set of overlapping sites. 

First calculate per pop saf for each population. This is already done.


```{r, engine=bash, eval=FALSE}

#SCRIPT TO GENERATE ALL POSSIBLE ITERATIONS
nano pops_combinations.R
========================================================================================================================
#!/usr/bin/env Rscript
library(dplyr)
library(magrittr)
f <- file("stdin")
open(f)
pops <- scan(f, what = "list")
combinations <- as.data.frame(t(combn(pops, 2))) %>% mutate (combination=paste(V1,V2,sep="/")) %>% select(combination)
for (combination in combinations) {
cat(combination)
}
========================================================================================================================
chmod 555 pops_combinations.R


screen -S creating_realSFSs_and_Fst_intergenic
script pairwise_real2DSFSs_and_Fst_intergenic.log

RUTA=/home/mlucena/ANGSD_analysis 
ANGSD="/opt/angsd/angsd"
NGSTOOLS="/opt/angsd/angsd/misc"
THREADS=15
SFS_FOLDER=$RUTA/intergenic_analysis/intergenic_sfs
FST_FOLDER=$RUTA/intergenic_analysis/intergenic_fst

#Get all populations name
cd $SFS_FOLDER
POPS=($(ls *.unfolded-lr.postprob.thetas.gz | sed -e 's/.unfolded-lr.postprob.thetas.gz//g'  | sort | uniq))
cd $FST_FOLDER
#Get all possible pairwise combinations
POP_PAIRS=($(echo ${POPS[@]} | Rscript pops_combinations.R))

# I have 28 combinations:

c_ll_cr_n006/c_ll_ki_n013
c_ll_cr_n006/c_ll_la_n006
c_ll_cr_n006/c_ll_no_n008
c_ll_cr_n006/c_ll_po_n008
c_ll_cr_n006/c_ll_vl_n008
c_ll_cr_n006/c_ll_ya_n008
c_ll_cr_n006/x_ll_ba_n003
c_ll_ki_n013/c_ll_la_n006
c_ll_ki_n013/c_ll_no_n008
c_ll_ki_n013/c_ll_po_n008
c_ll_ki_n013/c_ll_vl_n008
c_ll_ki_n013/c_ll_ya_n008
c_ll_ki_n013/x_ll_ba_n003
c_ll_la_n006/c_ll_no_n008
c_ll_la_n006/c_ll_po_n008
c_ll_la_n006/c_ll_vl_n008
c_ll_la_n006/c_ll_ya_n008
c_ll_la_n006/x_ll_ba_n003
c_ll_no_n008/c_ll_po_n008
c_ll_no_n008/c_ll_vl_n008
c_ll_no_n008/c_ll_ya_n008
c_ll_no_n008/x_ll_ba_n003
c_ll_po_n008/c_ll_vl_n008
c_ll_po_n008/c_ll_ya_n008
c_ll_po_n008/x_ll_ba_n003
c_ll_vl_n008/c_ll_ya_n008
c_ll_vl_n008/x_ll_ba_n003
c_ll_ya_n008/x_ll_ba_n003





#Iterate over combinations
for POP_PAIR in  "${POP_PAIRS[@]}"
do
POP1=($(echo $POP_PAIR | cut -d "/"  -f 1))
POP2=($(echo $POP_PAIR | cut -d "/"  -f 2)) 
echo ">$POP_PAIR"
#Skip iteration if file exists
if [ -f  $POP1.$POP2.unfolded-lr.saf.idx.ml  ]
  then
    echo "PAIR: $POP_PAIR exists."
    continue
fi
echo "--->realSFS 2 dimensional $POP1 and $POP2"
$NGSTOOLS/realSFS -P $THREADS $SFS_FOLDER/$POP1.unfolded-lr.saf.idx $SFS_FOLDER/$POP2.unfolded-lr.saf.idx  > $POP1.$POP2.unfolded-lr.saf.idx.ml

# The output file is a flatten matrix, where each value is the count of sites with the corresponding joint frequency ordered as [0,0] [0,1] and so on.
# less -S KIR.NOR.unfolded.ml.sfs
# No need to plot it, as it a ML estimation. 

# Prepare the fst for easy analysis
echo "-------------------->FST  $POP1 and $POP2"
$NGSTOOLS/realSFS fst index -P $THREADS $SFS_FOLDER/$POP1.unfolded-lr.saf.idx $SFS_FOLDER/$POP2.unfolded-lr.saf.idx  -sfs $POP1.$POP2.unfolded-lr.saf.idx.ml -fstout $POP1.$POP2.unfolded-lr
echo "--------------------------------->FST  $POP1 and $POP2 printing"
$NGSTOOLS/realSFS fst print $POP1.$POP2.unfolded-lr.fst.idx > $POP1.$POP2.unfolded-lr.fst.print
# Ojo!!! el archivo print ocupa muchÃ­simo!!!!!!
done

```



```{r, engine=bash, eval=FALSE}
#Get the global estimate

echo "POP1,POP2,FSTunweight,FSTweight" > global.fsts.csv
for POP_PAIR in  ${POP_PAIRS[@]}
do
POP1=($(echo $POP_PAIR | cut -d "/"  -f 1))
POP2=($(echo $POP_PAIR | cut -d "/"  -f 2)) 
FST=($( $NGSTOOLS/realSFS fst stats $POP1.$POP2.unfolded-lr.fst.idx | tr ' ' ',' ))
echo "$POP1,$POP2,$FST" >> global.fsts.csv
done



# you can have a look at their values:
#$NGSTOOLS/realSFS fst print KIR.NOR.BIA.unfolded.pbs.fst.idx | less -S

# where columns are: chromosome, postisition, (a), (a+b) values for the three FST comparisons, where FST is defined as a/(a+b).
# Note that FST on multiple SNPs is calculated as sum(a)/sum(a+b).


# Now, you could do an sliding-window analysis.
# * If you give only 2 pop in input, only the FST will be calculated. 
#### All this analysis could be estimated using NGStools (ngs2dSFS plus ngsFST) but it would need to:

# 1. Define the subset of overlapping sites as it does not compute it automatically as it does ANGSD.
# 2. It could required more memory than ANGSD and may be slower. 
# 3. The methods behind both programs are very similar. 
# 4. ngsTools implements a very simple estimator of the SFS, where the most likely joint allele frequency is recorded at each site. 
# 	 This tends to overestimate singletons, for instance. If you have enough sites, then it is recommended to use the 2DSFS estimated in ANGSD instead. 
# 5. ANGSD can compute a ML estimate of the 2D-SFS which should be preferred when many sites are available. 
# 	 However, ANGSD output file should be transformed (from log to un-log and from space-separated to tab-separated) before being used in ngsFST.


```

 # Copying to local
```{bash}
scp mlucena@genomics-b.ebd.csic.es:/home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_fst/*.fsts.csv /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_fst/
```




```{r}
library(dplyr)
library(tidyr)
library(magrittr)
library(ggplot2)
library(ape)

wd<-"/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_fst/"

#Look for all global fts table 

fst_table = list.files(path = wd, pattern="*fsts.csv")    


# Define function to plot trees

plot_FST_nj_weighted_unrooted_tree <- function(wd,df,filter_pattern,subset_name ){
first.pop <- read.csv( paste0(wd,df) , stringsAsFact=F, sep=",", header = TRUE) %>%
    mutate(pop1.n=(nchar(POP1)-4)/8,pop2.n=(nchar(POP2)-4)/8)%>%
    filter(pop1.n<2, pop2.n<2, grepl(filter_pattern,POP1),grepl(filter_pattern,POP2)) %>% 
    select(POP1) %>% slice(c(1))%>% as.character()
  last.pop<- read.csv( paste0(wd,df) , stringsAsFact=F, sep=",", header = TRUE) %>%
    mutate(pop1.n=(nchar(POP1)-4)/8,pop2.n=(nchar(POP2)-4)/8)%>%
    filter(pop1.n<2, pop2.n<2, grepl(filter_pattern,POP1),grepl(filter_pattern,POP2)) %>% 
    select(POP2) %>% slice(n())%>% as.character()
 
  fst.w_matrix<- read.csv( paste0(wd,df) , stringsAsFact=F, sep=",", header = TRUE) %>%
    rbind(c(first.pop,first.pop,NA,NA),
          c(last.pop,last.pop,NA,NA)) %>%
    mutate(pop1.n=(nchar(POP1)-4)/8,pop2.n=(nchar(POP2)-4)/8)%>%
    filter(pop1.n<2, pop2.n<2, grepl(filter_pattern,POP1),grepl(filter_pattern,POP2)) %>% 
    select(POP1,POP2,FSTweight)%>% 
    spread(POP2,FSTweight)%>%
    tibble::remove_rownames() %>% tibble::column_to_rownames("POP1")
  
  ## plot Neighbouring-Joining Tree
 
    pdf(paste0(wd,subset_name, gsub(".fsts.csv", "_weighted.pdf",df)), paper =  "a4" )
    plot(nj(as.dist(t(fst.w_matrix))), main=gsub(".csv", "_Fst_w", df),sub="neighbor joining",type="unrooted")
    dev.off()
}


# Format tables intro matrices and Plot NJtrees


for (i in 1:length (fst_table)){
  #Load table,  keep only the values of single populations 
  #(I migth be interested in the other comparissons but later)
  plot_FST_nj_weighted_unrooted_tree(wd,fst_table[i],"_ll_","lynx_lynx_fst_test") 
}


```

