---
title: "4.FST_intergenic"
output: html_document
---

```{r, engine=bash, eval=FALSE}


########################################################################################################################################################################
#
#      						Population Genetic Differenciation (Fst)
#
#											07/03/2017		Elena Marmesat
#
########################################################################################################################################################################


# Our program can estimate fst between populations. 
# And has been generalized to give all pairwise fst estimates if you supply the command with multiple populations.
# if you supply 3 populations, the program will also output the pbs statistic.

# NB we have removed the very unusefull unweighted fst estimator in the output, and have included a header. The output example below will be updated at some point.

# The procedure is

# 1 - Use angsd for calculating saf files for each population
# 2 - Use realSFS to calculate 2d sfs for each pair
# 3 - Use the above calculated 2dsfs as priors jointly with all safs from step1 to calculate fst binary files
# 4 - Use realSFS to extract the the fst values from the fst


#### 1 - Estimate multidimensional SFS 

# An important issue when doing this is to be sure that we are comparing the exactly same corresponding sites between populations. 
# ANGSD does that automatically and considers only a set of overlapping sites. 

# First calculate per pop saf for each population. This is already done in the script ANGSD-3_global.sh. So we will use KIR.saf, NOR.saf, BIA.saf (ML SAF).


#SCRIPT TO GENERATE ALL POSSIBLE ITERATIONS
nano pops_combinations.R
========================================================================================================================
#!/usr/bin/env Rscript
library(dplyr)
library(magrittr)
f <- file("stdin")
open(f)
pops <- scan(f, what = "list")
combinations <- as.data.frame(t(combn(pops, 2))) %>% mutate (combination=paste(V1,V2,sep="/")) %>% select(combination)
for (combination in combinations) {
cat(combination)
}
========================================================================================================================
chmod 555 pops_combinations.R


######################################################################################
#                            Specific target region file                             #
######################################################################################
#
declare -A REGIONS=(	["neutral"]="/home/emarmesat/grupolince/immunocapture/capture/neutral/no_genes_Lypa_10000longest_center_final" \
			["immunome"]="/home/emarmesat/grupolince/immunocapture/capture/functional/target-genes2_antigen-immune_cds_merged_size80" \
			["rand-cds"]="/home/emarmesat/grupolince/immunocapture/capture/functional/random_exons80_norep_notarget"  )


#SITES FILES
#format regions files to the angsd format.
#If you want to use bed files, you need to convert to the native angsd format.
#Bedfiles are chr pos1 pos2 value, and are zero indexed. Furthermore pos1 is included, but pos2 is not included. 
for REGION  in ${!REGIONS[@]}
do
echo "--->$REGION"

awk '{print $1"\t"$2+1"\t"$3}' ${REGIONS[$REGION]}.bed | sed 's/lp23s/lp23.s/' > ${REGIONS[$REGION]}.sites
#then index
angsd sites index ${REGIONS[$REGION]}.sites

#RF FILES (lp23.s00001:3574877-3575027) --> $$ al final no puedo usar -rf, no soy capax de que corra en combinación con -sites, pero -sites debería ser suficiente
awk -v OFS='' '{print $1,":",$2+1,"-",$3}' ${REGIONS[$REGION]}.bed > ${REGIONS[$REGION]}.rf
done

######################################################################################
#                           Calculate all pairwise 2dsfs's                           #
######################################################################################

FST_FOLDER=/home/emarmesat/grupolince/immunocapture/ansgd/fst
mkdir $FST_FOLDER
screen -S creating_realSFSs_and_Fst
script pairwise_real2DSFSs_and_Fst.log
ANGSD="/opt/angsd"
NGSTOOLS="/opt/angsd/misc"
THREADS=15
SFS_FOLDER=/home/emarmesat/grupolince/immunocapture/ansgd/sfs
FST_FOLDER=/home/emarmesat/grupolince/immunocapture/ansgd/fst
declare -A REGIONS=(	["neutral"]="/home/emarmesat/grupolince/immunocapture/capture/neutral/no_genes_Lypa_10000longest_center_final" ["immunome"]="/home/emarmesat/grupolince/immunocapture/capture/functional/target-genes2_antigen-immune_cds_merged_size80" ["rand-cds"]="/home/emarmesat/grupolince/immunocapture/capture/functional/random_exons80_norep_notarget"  )



#Get all populations name
cd $SFS_FOLDER
POPS=($(ls *.unfolded-lr.postprob.thetas.gz | sed -e 's/.unfolded-lr.postprob.thetas.gz//g'  | sort | uniq))
cd $FST_FOLDER
#Get all possible pairwise combinations
POP_PAIRS=($(echo ${POPS[@]} | Rscript pops_combinations.R))

#Iterate over combinations
for POP_PAIR in  "${POP_PAIRS[@]}"
do
POP1=($(echo $POP_PAIR | cut -d "/"  -f 1))
POP2=($(echo $POP_PAIR | cut -d "/"  -f 2)) 
echo ">$POP_PAIR"
#Skip iteration if file exists
if [ -f  $POP1.$POP2.unfolded-lr.$REGION.saf.idx.ml  ]
  then
    echo "PAIR: $POP_PAIR exists."
    continue
fi
#Compute Fsts for all region types
for REGION  in ${!REGIONS[@]}
do
echo "--->realSFS 2 dimensional $POP1 and $POP2"
$NGSTOOLS/realSFS -P $THREADS -sites ${REGIONS[$REGION]}.sites $SFS_FOLDER/$POP1.unfolded-lr.saf.idx $SFS_FOLDER/$POP2.unfolded-lr.saf.idx  > $POP1.$POP2.unfolded-lr.$REGION.saf.idx.ml

# The output file is a flatten matrix, where each value is the count of sites with the corresponding joint frequency ordered as [0,0] [0,1] and so on.
#less -S KIR.NOR.unfolded.ml.sfs
# No need to plot it, as it a ML estimation. 

# Prepare the fst for easy analysis
echo "-------------------->FST  $POP1 and $POP2"
$NGSTOOLS/realSFS fst index -P $THREADS  -sites ${REGIONS[$REGION]}.sites $SFS_FOLDER/$POP1.unfolded-lr.saf.idx $SFS_FOLDER/$POP2.unfolded-lr.saf.idx  -sfs $POP1.$POP2.unfolded-lr.$REGION.saf.idx.ml  -fstout $POP1.$POP2.unfolded-lr.$REGION
echo "--------------------------------->FST  $POP1 and $POP2 printing"
$NGSTOOLS/realSFS fst print $POP1.$POP2.unfolded-lr.$REGION.fst.idx > $POP1.$POP2.unfolded-lr.$REGION.fst.print
 
done
done


#Get the global estimate
for REGION  in ${!REGIONS[@]}
do
echo $REGION
echo "POP1,POP2,FSTunweight,FSTweight" > global.$REGION.fsts.csv
for POP_PAIR in  ${POP_PAIRS[@]}
do
POP1=($(echo $POP_PAIR | cut -d "/"  -f 1))
POP2=($(echo $POP_PAIR | cut -d "/"  -f 2)) 
FST=($( $NGSTOOLS/realSFS fst stats $POP1.$POP2.unfolded-lr.$REGION.fst.idx | tr ' ' ',' ))
echo "$POP1,$POP2,$FST" >> global.$REGION.fsts.csv
done
done


# you can have a look at their values:
#$NGSTOOLS/realSFS fst print KIR.NOR.BIA.unfolded.pbs.fst.idx | less -S

# where columns are: chromosome, postisition, (a), (a+b) values for the three FST comparisons, where FST is defined as a/(a+b).
# Note that FST on multiple SNPs is calculated as sum(a)/sum(a+b).


# Now, you could do an sliding-window analysis.
# * If you give only 2 pop in input, only the FST will be calculated. 
#### All this analysis could be estimated using NGStools (ngs2dSFS plus ngsFST) but it would need to:

# 1. Define the subset of overlapping sites as it does not compute it automatically as it does ANGSD.
# 2. It could required more memory than ANGSD and may be slower. 
# 3. The methods behind both programs are very similar. 
# 4. ngsTools implements a very simple estimator of the SFS, where the most likely joint allele frequency is recorded at each site. 
# 	 This tends to overestimate singletons, for instance. If you have enough sites, then it is recommended to use the 2DSFS estimated in ANGSD instead. 
# 5. ANGSD can compute a ML estimate of the 2D-SFS which should be preferred when many sites are available. 
# 	 However, ANGSD output file should be transformed (from log to un-log and from space-separated to tab-separated) before being used in ngsFST.


```