---
title: "5.Admixture"
output: html_document
---

# Generating Admixture plot

## Converting our data in beagle format:


```{r, engine=bash, eval=FALSE}

ANGSD=/opt/angsd/angsd
RUTA=/home/mlucena/ANGSD_analysis 
POP="c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-h_ll_ba"
POP_MODIF="x_ll_xx"


# Running GL:

$ANGSD/angsd -GL 1 -out "$POP".genolike -nThreads 40 -doGlf 2 -doMajorMinor 1 -SNP_pval 1e-6 -doMaf 1  -bam $RUTA/intergenic_analysis/"$POP"_n*.intergenic.bamlist

for i in $POP*
do
echo $i
scp $i ${i/$POP/$POP_MODIF} 
done

# Imputation in beagle (optional):

# java -jar /home/mlucena/software/beagle.jar like="$POP_MODIF".genolike.beagle.gz out=

# out se deja sin nada pq por defecto pone cosas
```

Population format might be a problem as beagle does not admit long names! 
Check if it is beagle, cause then is not a problem. 

## Running admixture

```{r, engine=bash, eval=FALSE}

for i in `seq 1 12`;
do
echo $i
/home/mlucena/software/NGSadmix -likes "$POP_MODIF".genolike.beagle.gz -K "$i" -P 20 -o "$POP_MODIF"_K"$i" -minMaf 0.05 
done    

```

./NGSadmix
Arguments:
	-likes Beagle likelihood filename
	-K Number of ancestral populations
Optional:
	-fname Ancestral population frequencies
	-qname Admixture proportions
	-o Prefix for output files
	-printInfo print ID and mean maf for the SNPs that were analysed
Setup:
	-seed Seed for initial guess in EM
	-P Number of threads
	-method If 0 no acceleration of EM algorithm
	-misTol Tolerance for considering site as missing
Stop chriteria:
	-tolLike50 Loglikelihood difference in 50 iterations
	-tol Tolerance for convergence
	-dymBound Use dymamic boundaries (1: yes (default) 0: no)
	-maxiter Maximum number of EM iterations
Filtering
	-minMaf Minimum minor allele frequency
	-minLrt Minimum likelihood ratio value for maf>0
	-minInd Minumum number of informative individuals
	
	
## Plotting results

First we need a file with our population information:

```{bash}

cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 3,4,5 | rev  > $POP_MODIF.column1.borrar
cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 2,3,4,5 | rev  > $POP_MODIF.column2.borrar

paste $POP_MODIF.column1.borrar $POP_MODIF.column2.borrar > $POP_MODIF.info

rm *borrar
```


### Creating a plot

```{bash}
nano Admix_plot.R
```

```{r}
library(RColorBrewer)
args <- commandArgs(TRUE)
print(args)
admix<-t(as.matrix(read.table(args[1])))
pop<-read.table(args[2],as.is=T)
admix<-admix[,order(pop[,1])]
pop<-pop[order(pop[,1]),]
n <- args[3]
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
pie(rep(1,n), col=sample(col_vector, n))
#setEPS()
# postscript(paste0(args[4],"_K",args[3],".eps"))
pdf(paste0(args[4],"_K",args[3],".pdf"))
barplot(admix,col=col_vector,space=0,xlab="Individuals",ylab="Admixture")
text(tapply(1:nrow(pop),pop[,1],mean),-0.1,unique(pop[,1]),xpd=T,srt=60)
dev.off()

```

```{bash}
for i in `seq 1 12`;
do
echo $i
Rscript Admix.R "$POP_MODIF"_K"$i".qopt "$POP_MODIF".info $i $POP_MODIF
done
```

```{r}

library(RColorBrewer)

admix<-t(as.matrix(read.table("/Users/marialucenaperez/ownCloud/publico/Eurasian_lynx_phylogeography/lynx_lynx_ADMIXTURE/x_ll_xx_K2.qopt"))) 
pop<-read.table("/Users/marialucenaperez/ownCloud/publico/Eurasian_lynx_phylogeography/lynx_lynx_ADMIXTURE/x_ll_xx.info",as.is=T)
admix<-admix[,order(pop[,1])]
pop<-pop[order(pop[,1]),]

n <- 2
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
pie(rep(1,n), col=sample(col_vector, n))

h<-barplot(admix,space=0,xlab="Individuals",ylab="Admixture")
text(tapply(1:nrow(pop),pop[,1],mean),-0.1,unique(pop[,1]),xpd=T,srt=60)

```



## Test solo con la region de Elena porque sospecho que ha pasado algo durante el proceso de creaciÃ³n del anterior archivo. 

```{bash}

# Test
cd /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_ADMIXTURE/test

REGIONFILE="/home/mlucena/ANGSD_analysis/depth_calculus/no_genes_Lypa_10000longest_center_final_slop20_dot.rf"
POP="c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-h_ll_ba"
ANGSD=/opt/angsd/angsd


$ANGSD/angsd -GL 1 -out "$POP".TEST.genolike -nThreads 40 -doGlf 2 -doMajorMinor 1 -SNP_pval 1e-6 -doMaf 1  -rf $REGIONFILE  -bam /home/mlucena/ANGSD_analysis/whole_genome_analysis/$POP*.bamlist

java -jar /home/mlucena/software/beagle.jar like=lynx_lynx.TEST.genolike.beagle.gz out=lynx_lynx.TEST.beagleOutput

```


Command line: java -Xmx27305m -jar beagle.jar
  like=lynx_lynx.TEST.genolike.beagle.gz
  out=lynx_lynx.TEST.beagleOutput

number of markers    =    4718
number of haplotypes =     148 (lynx_lynx.TEST.genolike.beagle.gz)

Phasing: iteration 1

Phasing: iteration 2

Phasing: iteration 3

Phasing: iteration 4

Phasing: iteration 5

Phasing: iteration 6

Phasing: iteration 7

Phasing: iteration 8

Phasing: iteration 9

Phasing: iteration 10


```{bash}

/home/mlucena/software/NGSadmix -likes lynx_lynx.TEST.genolike.beagle.gz -K 1 -P 40 -o "lynx_lynx.TEST" -minMaf 0.05 

```






