---
title: "5.Admixture"
output: html_document
---



# Generating Admixture plot

## Converting our data in beagle format:


```{bash}


POP="c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080"  # <--CHANGE POP HERE


RUTA=/home/mlucena/ANGSD_analysis/depth_calculus 
THREADS=15                    # no. of computer cores used by bwa and samtools. 20 = OK, >20 = ask people first!

read POP mean sd mean_truncated sd_truncated maxDepth minDepth maxDepth_truncated minDepth_truncated < $RUTA/"${POP}"_mean_sd_depthGlobal_lynx_per_pop_mean_folds_0.95.csv

ANGSD="/opt/angsd/angsd"
NGSTOOLS="/opt/angsd/angsd/misc"
REF="/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23_without_repetitive_transposable_low_complexity.fa"
ANC="/home/GRUPOS/grupolince/reference_genomes/lynx_rufus_genome/c_lr_zz_0001_recal1.fa"
FILTER1=" -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -baq 1 -C 50 "
FILTER2=" -minMapQ 30 -minQ 20 -doCounts 1 "
N_IND=$(echo ${POP: -3} )
MIN_IND=$(expr $N_IND / 2)
SNP_PVAL="1e-4"

RUTA=/home/mlucena/ANGSD_analysis 


# Sanity checks:

echo $POP
echo $N_IND
echo $MIN_IND
echo $maxDepth
echo $minDepth
echo $SNP_PVAL

# Running GL:

$ANGSD/angsd -nThreads $THREADS -bam $RUTA/intergenic_analysis/"$POP".intergenic.bamlist -ref $REF  \
-out "$POP".genolike \
$FILTER1 \
$FILTER2 \
-GL 1  -doGlf 2 \
-doMajorMinor 1 -doMaf 1 -SNP_pval $SNP_PVAL -skipTriallelic 1 \
-minInd $MIN_IND -setMaxDepth $maxDepth -setMinDepth $minDepth 


# Parameters chosen:
# FILTER1=" -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -baq 1 -C 50 "
# FILTER2=" -minMapQ 30 -minQ 20 -doCounts 1 "
# GL 1 --> Use samtools
# doGlf 2 --> beagle likelihood file
# doMajorMinor 1 --> Infer major and minor from GL (other options would be: infer major and minor from allele counts, from a file, use reference as major, use ancestral as major)
# doMaf 3 --> frequency (fixed major and minor/fixed major,unknownminor): Both are assumed to be known. Allele freq i obtained using based on GL.  
# -doMajorMinor 1 -doMaf 3. Example for estimating the allele frequencies both while assuming known major and minor allele but also while taking the uncertaincy of the minor allele inference into account. The inference of the major and minor allele is done directly from the genotype likelihood
# -SNP_pval [float]
# The p-value used for calling snaps. see Allele_Frequency_estimation for additional options. 



#################################
# Lo que hacemos con este script es: Primero decidimos que vamos a usar las frecencias alellicas en cuenta para establecer cual es el minor y major allele, usando los Genotipes likelihood. Usando estos major y minor Usamos los GL para obteenr la frecuencia del mayor y el minor. Además también asumimos cierta incertidumbre en el minor. 

#################################

# for i in $POP*
# do
# echo $i
# scp $i ${i/$POP/$POP_MODIF} 
# done

# Imputation in beagle (optional):

# java -jar /home/mlucena/software/beagle.jar like="$POP_MODIF".genolike.beagle.gz out=

# out se deja sin nada pq por defecto pone cosas

zcat c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.mafs.gz | tail -n+2 | wc -l

# 8262604 sites.


```

Population format might be a problem as beagle does not admit long names! 
Check if it is beagle, cause then is not a problem. 

## Running admixture

Corro 10 veces el comando para cada K. 

```{r, engine=bash, eval=FALSE}


for j in `seq 1 10`
do
for i in `seq 1 13`;
do
echo "running RUN" $j "k" $i
/home/mlucena/software/NGSadmix -likes "$POP".genolike.beagle.gz -K "$i" -P 20 -o "$POP"_K"$i"_run"$j" 
done  
done


```


using only SNP with af MAF above 5% (-minMaf 0.05).


./NGSadmix
Arguments:
	-likes Beagle likelihood filename
	-K Number of ancestral populations
Optional:
	-fname Ancestral population frequencies
	-qname Admixture proportions
	-o Prefix for output files
	-printInfo print ID and mean maf for the SNPs that were analysed
Setup:
	-seed Seed for initial guess in EM
	-P Number of threads
	-method If 0 no acceleration of EM algorithm
	-misTol Tolerance for considering site as missing
Stop chriteria:
	-tolLike50 Loglikelihood difference in 50 iterations
	-tol Tolerance for convergence
	-dymBound Use dymamic boundaries (1: yes (default) 0: no)
	-maxiter Maximum number of EM iterations
Filtering
	-minMaf Minimum minor allele frequency
	-minLrt Minimum likelihood ratio value for maf>0
	-minInd Minumum number of informative individuals
	
	
## Likelihood test

Estimating the Best K

The method of Evanno enables the user to identify a single k value, out of a range of K values, which captures the uppermost level of structure. This method was purposed by Evonno et al. in 2005 (Molecular Ecology 14, 2005). In addition, we also use ln(Pr(X|K) values in order to identify the k for which Pr(K=k) is highest (as described in STRUCTURE's manual, section 5.1).

Encontrado en:

https://github.com/alexkrohn/AmargosaVoleTutorials/blob/master/ngsAdmix_tutorial.md


```{bash}

# First I registered the likelihoods:

for i in *log
do
echo $i | cut -d'.' -f 1 | cut -d 'K' -f 2 | cut -d'_' -f1 >> Ks.borrar
grep "best like=" $i | cut -d'=' -f 2 | cut -d' ' -f 1 >> likelihood.borrar
done

paste Ks.borrar likelihood.borrar | sort -n -k2 > Ks_log_prob_table.txt

rm *borrar

```

## Plotting admix results

First we need a file with our population information:

```{bash}

cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 3,4,5 | rev  > $POP.column1.borrar
cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 2,3,4,5 | rev  > $POP.column2.borrar

paste $POP.column1.borrar $POP.column2.borrar > $POP.info

rm *borrar
```


### Creating a plot

Guardo este script para lanzarlo en secuencial para cada una de las ks:

```{bash}
nano Admix_plot.R
```

Script que guardo

```{r}
library(RColorBrewer)
library (dplyr)
args <- commandArgs(TRUE)
print(args)
admix<-t(as.matrix(read.table(args[1])))
pop<-read.table(args[2],as.is=T)


names(pop) <- c("pop", "ind")
pop$pop <- as.factor (pop$pop)
pop$pop <- factor (pop$pop, levels=c("c_ll_no", "h_ll_ba", "c_ll_ba","c_ll_cr",  "c_ll_po", "c_ll_la", "c_ll_ki", "c_ll_ur", "c_ll_vl", "c_ll_ya", "c_ll_tu", "c_ll_og", "c_ll_ka", "c_ll_to"))

admix<-admix[,order(pop[,1])]
pop<-pop[order(pop[,1]),]
n <- args[3]
# qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',] Antes era así, lo que significa que coje aquellas paletas que son cualitativas. como yo quiero que coja paired la primera para que sea como en el PCA, pongo como condición que tenga maxcolors de 12. 
twelve_col_pals = brewer.pal.info[brewer.pal.info$maxcolors == '12',] 
col_vector = unlist(mapply(brewer.pal, twelve_col_pals$maxcolors, rownames(twelve_col_pals)))
pie(rep(1,n), col=sample(col_vector, n))
#setEPS()
# postscript(paste0(args[4],"_K",args[3],".eps"))

pdf(paste0(args[4],"_K",args[3],".pdf"), height=10, width = 30)
barplot(admix,col=col_vector,space=0,xlab="Individuals",ylab="Admixture")
text(tapply(1:nrow(pop),pop[,1],mean),-0.1,unique(pop[,1]),xpd=T,srt=60)
dev.off()


```

Lo lanzo de manera secuencial.

```{bash}

for run in `seq 1 10`;
do

for i in `seq 1 13`;
do
echo $i
Rscript Admix_plot.R "$POP"_K"$i"_run$run.qopt "$POP".info $i $POP"_run"$run
done

done

```

Copiar tabla para correr CLUMPAK

http://clumpak.tau.ac.il/bestK.html

```{bash}

scp mlucena@genomics-b.ebd.csic.es://home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_ADMIXTURE/Ks_log_prob_table.txt /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_ADMIXTURE/2017-10-26_Snp-pval/

scp mlucena@genomics-b.ebd.csic.es://home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_ADMIXTURE/*pdf /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_ADMIXTURE/2017-10-26_Snp-pval/

```




```{bash}

perl BestKByEvanno.pl --i 110 --d /home/mlucena/results --f /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_ADMIXTURE/Ks_log_prob_table.txt  -- inputtype lnprobbyk

```



# Admixture with Maf=1 sites!!!!

Nos hemos dado cuenta que hay muchos más sitios en Admixture que en PCa (mirar script z3-5.Different_number_of_sites), por tanto vamos a intentar filtrar el Admxiture para los sitios del PCA. 

```{bash}

# Primero copiamos todos los sitios del PCA

zcat c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.intergenic.mafs.gz | awk 'NR >1 {print $1"_"$2 }' > ../intergenic_ADMIXTURE/sites_PCA.txt


# Ahora filtramos el archivo beagle que tiene esta pinta

<!-- marker  allele1 allele2 Ind0    Ind0    Ind0    Ind1    Ind1    Ind1    Ind2    Ind2    Ind2    Ind3    Ind3    Ind3    Ind4    Ind4    Ind4    Ind5    Ind5    Ind5    Ind6    Ind6    Ind6    Ind7    Ind7    Ind7 -->
<!-- lp23.s00001_2166        1       3       0.941178        0.058822        0.000000        0.941178        0.058822        0.000000        0.047749        0.951949        0.000302        0.888891        0.111108 -->
<!-- lp23.s00001_2191        1       0       0.941178        0.058822        0.000000        0.800000        0.199997        0.000003        0.941178        0.058822        0.000000        0.888891        0.111109 -->
<!-- lp23.s00001_3926        2       0       0.799962        0.199987        0.000050        0.799902        0.199972        0.000126        0.984616        0.015384        0.000000        0.888886        0.111108 -->
<!-- lp23.s00001_4153        3       1       0.969698        0.030302        0.000000        0.969698        0.030302        0.000000        0.799998        0.199996        0.000007        0.941178        0.058822 -->


# Y tiene:
zcat c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.beagle.gz | wc -l
# 8262605

# Lo filtarmos usando el archivo de sitios que acabamos de crear.

zgrep  -wFf sites_PCA.txt c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.beagle.gz > c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.beagle.modif

wc -l c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.beagle.modif
# 2096552 c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.beagle.modif

wc -l sites_PCA.txt 
# 2096552 sites_PCA.txt

awk '{print $1}' c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.beagle.modif | diff - sites_PCA.txt 
# Nada! perfecto!

gzip c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.beagle.modif
's/\|/\t/g'

sed 's/_/\t/g' sites_PCA.txt > sites_PCA2.txt 

zgrep  -wFf sites_PCA2.txt c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.mafs.gz > c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.mafs.modif

wc -l  c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.mafs.modif
# 2096552

gzip c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.mafs.modif

rm sites_PCA2.txt
rm sites_PCA.txt

```

Ahora creo un README que tiene todos estos datos, y los guardo con el nombre original que debieran, también modifico el .arg, para que no nos confundamos nunca. 



```{bash}
mv c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.beagle.modif.gz c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.beagle.gz


mv c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.mafs.modif.gz c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.mafs.gz

```


## Running admixture

Corro 10 veces el comando para cada K. 

```{r, engine=bash, eval=FALSE}


for j in `seq 1 10`
do
for i in `seq 1 13`;
do
echo "running RUN" $j "k" $i
/home/mlucena/software/NGSadmix -likes "$POP".genolike.beagle.gz -K "$i" -P 20 -o "$POP"_K"$i"_run"$j" 
done  
done


```


using only SNP with af MAF above 5% (-minMaf 0.05).


./NGSadmix
Arguments:
	-likes Beagle likelihood filename
	-K Number of ancestral populations
Optional:
	-fname Ancestral population frequencies
	-qname Admixture proportions
	-o Prefix for output files
	-printInfo print ID and mean maf for the SNPs that were analysed
Setup:
	-seed Seed for initial guess in EM
	-P Number of threads
	-method If 0 no acceleration of EM algorithm
	-misTol Tolerance for considering site as missing
Stop chriteria:
	-tolLike50 Loglikelihood difference in 50 iterations
	-tol Tolerance for convergence
	-dymBound Use dymamic boundaries (1: yes (default) 0: no)
	-maxiter Maximum number of EM iterations
Filtering
	-minMaf Minimum minor allele frequency
	-minLrt Minimum likelihood ratio value for maf>0
	-minInd Minumum number of informative individuals
	
	
## Likelihood test

Estimating the Best K

The method of Evanno enables the user to identify a single k value, out of a range of K values, which captures the uppermost level of structure. This method was purposed by Evonno et al. in 2005 (Molecular Ecology 14, 2005). In addition, we also use ln(Pr(X|K) values in order to identify the k for which Pr(K=k) is highest (as described in STRUCTURE's manual, section 5.1).

Encontrado en:

https://github.com/alexkrohn/AmargosaVoleTutorials/blob/master/ngsAdmix_tutorial.md


```{bash}

# First I registered the likelihoods:

for i in *log
do
echo $i | cut -d'.' -f 1 | cut -d 'K' -f 2 | cut -d'_' -f1 >> Ks.borrar
grep "best like=" $i | cut -d'=' -f 2 | cut -d' ' -f 1 >> likelihood.borrar
done

paste Ks.borrar likelihood.borrar | sort -n -k2 > Ks_log_prob_table.txt

rm *borrar

```

## Plotting admix results

First we need a file with our population information:

```{bash}

cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 3,4,5 | rev  > $POP.column1.borrar
cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 2,3,4,5 | rev  > $POP.column2.borrar

paste $POP.column1.borrar $POP.column2.borrar > $POP.info

rm *borrar
```


### Creating a plot

Guardo este script para lanzarlo en secuencial para cada una de las ks:

```{bash}
nano Admix_plot.R
```

Script que guardo

```{r}
library(RColorBrewer)
library (dplyr)
args <- commandArgs(TRUE)
print(args)
admix<-t(as.matrix(read.table(args[1])))
pop<-read.table(args[2],as.is=T)

names(pop) <- c("pop", "ind")
pop$pop <- as.factor (pop$pop)
pop$pop <- factor (pop$pop, levels=c("c_ll_no", "h_ll_ba", "c_ll_ba","c_ll_cr",  "c_ll_po", "c_ll_la", "c_ll_ki", "c_ll_ur", "c_ll_vl", "c_ll_ya", "c_ll_tu", "c_ll_og", "c_ll_ka", "c_ll_to"))

admix<-admix[,order(pop[,1])]
pop<-pop[order(pop[,1]),]
n <- args[3]
# qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',] Antes era así, lo que significa que coje aquellas paletas que son cualitativas. como yo quiero que coja paired la primera para que sea como en el PCA, pongo como condición que tenga maxcolors de 12. 
twelve_col_pals = brewer.pal.info[brewer.pal.info$maxcolors == '12',] 
col_vector = unlist(mapply(brewer.pal, twelve_col_pals$maxcolors, rownames(twelve_col_pals)))
pie(rep(1,n), col=sample(col_vector, n))
#setEPS()
# postscript(paste0(args[4],"_K",args[3],".eps"))

pdf(paste0(args[4],"_K",args[3],".pdf"), height=10, width = 30)
barplot(admix,col=col_vector,space=0,xlab="Individuals",ylab="Admixture")
text(tapply(1:nrow(pop),pop[,1],mean),-0.1,unique(pop[,1]),xpd=T,srt=60)
dev.off()


```

Lo lanzo de manera secuencial.

```{bash}

for run in `seq 1 10`;
do

for i in `seq 1 13`;
do
echo $i
Rscript Admix_plot.R "$POP"_K"$i"_run$run.qopt "$POP".info $i $POP"_run"$run
done

done

```

Copiar tabla para correr CLUMPAK

http://clumpak.tau.ac.il/bestK.html

```{bash}

scp mlucena@genomics-b.ebd.csic.es://home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_ADMIXTURE/Ks_log_prob_table.txt /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_ADMIXTURE/2017-10-16_Snp-pval_/

scp mlucena@genomics-b.ebd.csic.es://home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_ADMIXTURE/*pdf /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_ADMIXTURE/2017-10-16_Snp-pval_/

```


test

```{r}

library(RColorBrewer)
library (dplyr)
library("reshape2")
library(ggplot2)
library(lemon)


wd <-"/Users/marialucenaperez/Desktop/"
# args <- commandArgs(TRUE)
# print(args)

admix<-t(as.matrix(read.table(paste (wd,"c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080_K9_run8.qopt", sep=""))))

pop<-read.table(paste(wd,"c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.info", sep =""))

names(pop) <- c("pop", "ind")
pop$pop <- as.factor (pop$pop)
pop$pop <- factor (pop$pop, levels=c("c_ll_no", "h_ll_ba", "c_ll_ba","c_ll_cr",  "c_ll_po", "c_ll_la", "c_ll_ki", "c_ll_ur", "c_ll_tu", "c_ll_og", "c_ll_ka", "c_ll_to", "c_ll_ya", "c_ll_vl"))



admix<-admix[,order(pop[,1])]
pop <-pop[order(pop[,1]),] 
pop$Var2 <- seq.int(nrow(pop)) 
pop <- pop %>% mutate (Population =  
                          ifelse (pop == "c_ll_po", "Poland",
                          ifelse (pop == "c_ll_ur", "Urals",
                          ifelse (pop == "c_ll_ki", "Kirov",
                          ifelse (pop == "c_ll_la", "Latvia",
                          ifelse (pop == "c_ll_no", "Norway",
                          ifelse (pop == "c_ll_ba" | pop == "h_ll_ba", "Balkans" ,
                          ifelse (pop == "c_ll_cr","Carpathians",
                          ifelse (pop == "c_ll_to", "Centr_Mong",
                          ifelse (pop == "c_ll_tu", "Tuva",
                          ifelse (pop == "c_ll_ka", "Khentiy_Mong", 
                          ifelse (pop == "c_ll_og", "South_Gobi", 
                          ifelse (pop == "c_ll_vl", "Vladivostok", 
                          ifelse (pop == "c_ll_ya", "Yakutia",
                          ifelse (pop == "c_ll_zz", "Zoo",NA)))))))))))))))  %>% rename(Pop=pop)
tmp <- melt(admix)  %>% full_join(., pop, by="Var2") 

tmp$ind <- factor (tmp$ind, levels=c("c_ll_no_0075", "c_ll_no_0076", "c_ll_no_0077", "c_ll_no_0078", "c_ll_no_0079", "c_ll_no_0080", "c_ll_no_0081", "c_ll_no_0082", "c_ll_ba_0216", "h_ll_ba_0214", "h_ll_ba_0215", "c_ll_cr_0205", "c_ll_cr_0206", "c_ll_cr_0207", "c_ll_cr_0208", "c_ll_cr_0209", "c_ll_cr_0212", "c_ll_po_0001", "c_ll_po_0002", "c_ll_po_0003", "c_ll_po_0011", "c_ll_po_0014", "c_ll_po_0019", "c_ll_po_0105", "c_ll_po_0106", "c_ll_la_0044", "c_ll_la_0047", "c_ll_la_0048", "c_ll_la_0052", "c_ll_la_0053", "c_ll_la_0054", "c_ll_ki_0090", "c_ll_ki_0091", "c_ll_ki_0092", "c_ll_ki_0093", "c_ll_ki_0094", "c_ll_ki_0095", "c_ll_ki_0096", "c_ll_ki_0097", "c_ll_ki_0098", "c_ll_ki_0099", "c_ll_ki_0100", "c_ll_ki_0101", "c_ll_ki_0102", "c_ll_ur_0194", "c_ll_ur_0195", "c_ll_ur_0196", "c_ll_ur_0199", "c_ll_ur_0200", "c_ll_ur_0203", "c_ll_tu_0153", "c_ll_tu_0157", "c_ll_tu_0158", "c_ll_tu_0159", "c_ll_tu_0165", "c_ll_tu_0166", "c_ll_to_0190", "c_ll_to_0191", "c_ll_ka_0184", "c_ll_ka_0186", "c_ll_ka_0188", "c_ll_ka_0189", "c_ll_og_0181", "c_ll_og_0187", "c_ll_ya_0138", "c_ll_ya_0139", "c_ll_ya_0140", "c_ll_ya_0142", "c_ll_ya_0143", "c_ll_ya_0145", "c_ll_ya_0146", "c_ll_ya_0147", "c_ll_vl_0107", "c_ll_vl_0108", "c_ll_vl_0109", "c_ll_vl_0110", "c_ll_vl_0112", "c_ll_vl_0113", "c_ll_vl_0128", "c_ll_vl_0132"))


# n <- 24 # Esto servía porque aquí venían las k y así hacia una paleta adecuada a la k que tenía, en mi caso, ahora no hace falta. 
# qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',] Antes era así, lo que significa que coje aquellas paletas que son cualitativas. como yo quiero que coja paired la primera para que sea como en el PCA, pongo como condición que tenga maxcolors de 12. 
twelve_col_pals = brewer.pal.info[brewer.pal.info$maxcolors == '12',] 
col_vector = unlist(mapply(brewer.pal, twelve_col_pals$maxcolors, rownames(twelve_col_pals)))
# pie(rep(1,n), col=sample(col_vector, n))
my_paired_colours <- as.vector(col_vector)

x_breaks <-tapply(1:nrow(pop),pop[,4],mean)
x_breaks_to_dataframe <- data.frame(Population=names(x_breaks), breaks=x_breaks) 

#setEPS()
# postscript(paste0(args[4],"_K",args[3],".eps"))

# pdf(paste0("K_8.pdf"), height=10, width = 30)

admix_plot <- ggplot() + 
   geom_bar (data = tmp, (aes(x=ind, y=value, fill=Var1)), stat = "identity", position = "fill",  width=1) +
   scale_fill_manual(values=my_paired_colours) +
   scale_y_continuous(breaks = seq(0, 1, by = 0.2)) +
   # geom_text (data=x_breaks_to_dataframe, aes(x=breaks, y=1.2, size=20, label=Population), angle = 45) +
   theme_classic() +
   theme(plot.margin = unit(c(3,3,3,3), "cm")) 

for (i in 1:length(x_breaks_to_dataframe$breaks)) {
admix_plot <- admix_plot + annotation_custom(
      grob = textGrob(label = x_breaks_to_dataframe$Population[i], hjust = 1, gp = gpar(cex = 1.5)),
      ymin = 1.5,      # Vertical position of the textGrob
      ymax = 1.5,
      xmin = x_breaks_to_dataframe$breaks[i],         # Note: The grobs are positioned outside the plot area
      xmax = x_breaks_to_dataframe$breaks[i])
}
    
```


