---
title: "6. Genetic distance"
output: html_document
---



```{bash}

cd /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_genetic_distances

ANGSD=/opt/angsd/angsd
RUTA=/home/mlucena/ANGSD_analysis 
NGSTOOLS=/opt/ngsTools_19092016


POP="c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080"

# Running GL:

# En teoría tendría que calcular el archivo de GL con -doGlf 2 -doMajorMinor 1 -doMaf 1. Sin embargo, este ya está calculado puesto que lo usé para NGSAdmix. 
# Por tanto, voy a copiarlo directamente a mi carpeta y todo lo que está ahora mismo "comentado" NO lo he hecho. 

<!-- SNP_PVAL="1e-4" -->
<!-- $ANGSD/angsd -GL 1 -out "$POP"_.genolike -nThreads 40 -doGlf 2 -doMajorMinor 1 -SNP_pval $SNP_PVAL -doMaf 1 -bam $RUTA/intergenic_analysis/"$POP"_n*.intergenic.bamlist -->
<!-- SAMPLE="c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080.genolike.beagle.gz" -->
<!-- # SAMPLE="$POP"_SingletonSNP_pval_Filtered.genolike.beagle.gz -->
<!-- 
<!-- cat $RUTA/intergenic_analysis/"$POP"_n*.intergenic.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 2,3,4,5 | rev | sed 's/ /n/' > LABELS.txt -->


scp /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_ADMIXTURE/*genolike.beagle.gz /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_genetic_distances

scp /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_ADMIXTURE/*genolike.mafs.gz /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_genetic_distances

scp /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_ADMIXTURE/*.info /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_genetic_distances


N_SITES=`zcat $POP.genolike.mafs.gz | wc -l`
# No tiene cabecero así que no hago tail -n+2

awk '{print $2}' $POP.info > labels.txt


echo $N_SITES 
# 2096552; Perfecto! Igual que PCA y NGSAdmix! :)


# Calculo distancia.

$NGSTOOLS/ngsDist --n_threads 10 --seed 12345 --geno $POP.genolike.beagle.gz --probs --n_ind 80 --n_sites $N_SITES --labels labels.txt --call_geno --out $POP.dist


# Ninguno de estos porque ya he hecho la selección de sitios que quiero en el maf. 
# --call_thresh $SNP_PVAL 
# --N_thresh 0.3 

# ../ngsDist --n_threads 10 --seed 12345 --verbose 1 --geno testA_32.geno --probs --n_ind $N_IND --n_sites $N_SITES --labels testA.labels --n_boot_rep 5 --boot_block_size 10 --call_geno --N_thresh 0.3 --call_thresh 0.9 --out testA_32B-10CGf.dist


```

ERROR: [call_geno] missing data threshold must be smaller than calling genotype threshold!

--geno FILE: input file with genotypes, genotype likelihoods or genotype posterior probabilities.
--n_ind INT: sample size (number of individuals).
--n_sites INT: number of sites in input file.
--tot_sites INT: total number of sites in dataset.
--labels FILE: labels, one per line, of the input sequences.
--probs: is the input genotype probabilities (likelihoods or posteriors)?
--log_scale: Ii the input in log-scale?.
--call_geno: call genotypes before running analyses.
--N_thresh DOUBLE: minimum threshold to consider site; missing data if otherwise (assumes -call_geno)
--call_thresh DOUBLE: minimum threshold to call genotype; left as is if otherwise (assumes -call_geno)
--pairwise_del: pairwise deletion of missing data.
--avg_nuc_dist: use average number of nucleotide differences as distance (by default, ngsDist uses genotype distances based on allele frequency differences). Only pairs of heterozygous positions are actually affected when using this option, with their distance being 0.5 (instead of 0 by default).
--indep_geno: assume independence between genotypes?
--n_boot_rep INT: number of bootstrap replicates [0].
--boot_block_size INT: block size for bootstrapping [1].
--out FILE: output file name.
--n_threads INT: number of threads to use. [1]
--version: prints program version and exits.
--verbose INT: selects verbosity level. [1]
--seed INT: random number generator seed (only for the bootstrap analysis).



```{r}

data <- read.table("/Users/marialucenaperez/Desktop/c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-h_ll_ba.genolike.beagle.gz.dist", sep ="\t", header=F, dec=".")

# preubecinas anteriores:
# m <- as.dist(matrix)
# hmap(m)
# matrix <- read.table("/Users/marialucenaperez/Desktop/prueba.csv", sep =";", header=T, dec=".")
# a <- heatmap(as.matrix(matrix),symm = T, keep.dendro=F)
# rotate <- function() t(apply(x, 2, rev))


# Script from http://sebastianraschka.com/Articles/heatmaps_in_r.html
#########################################################
### A) Installing and loading required packages
#########################################################

if (!require("gplots")) {
   install.packages("gplots", dependencies = TRUE)
   library(gplots)
   }
if (!require("RColorBrewer")) {
   install.packages("RColorBrewer", dependencies = TRUE)
   library(RColorBrewer)
   }

#########################################################
### B) Reading in data and transform it into matrix format
#########################################################

rnames <- data[,1]                            # assign labels in column 1 to "rnames"
mat_data <- data.matrix(data[,2:ncol(data)])  # transform column 2-5 into a matrix
rownames(mat_data) <- rnames                  # assign row names


#########################################################
### C) Customizing and plotting the heat map
#########################################################

# creates a own color palette from red to green
my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)

# (optional) defines the color breaks manually for a "skewed" color transition
col_breaks = c(seq(-1,0,length=100),  # for red
  seq(0.01,0.8,length=100),           # for yellow
  seq(0.81,1,length=100))             # for green

# creates a 5 x 5 inch image
png("../images/heatmaps_in_r.png",    # create PNG for the heat map        
  width = 5*300,        # 5 x 300 pixels
  height = 5*300,
  res = 300,            # 300 pixels per inch
  pointsize = 8)        # smaller font size

heatmap.2(mat_data,
  symm = T,
  trace="none",         # turns off trace lines inside the heat map
  main = "Distance matrix", # heat map title
  density.info="none",  # turns off density plot inside color legend
  col=my_palette,       # use on color palette defined earlier
  Rowv="NA",            # turn off column clustering
  Colv="NA")
  #   margins =c(12,9))     # widens margins around plot
  # cellnote = mat_data,  # same data set for cell labels
  # notecol="black",      # change font color of cell labels to black
  # dendrogram="row")#,     # only draw a row dendrogram
  # breaks=col_breaks)    # enable color transition at specified limits

dev.off()   



```



