---
title: "6. Genetic distance"
output: html_document
---



```{bash}

cd /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_genetic_distances

SAMPLE="c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-h_ll_ba.genolike.beagle.gz"

NUMBER_OF_SITES=($(zcat $SAMPLE | echo `wc -l` -1 | bc)) # --> 5437960 

test=c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-h_ll_ba.genolike.beagle.gz

/opt/ngsTools_19092016/ngsDist/ngsDist --geno /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_ADMIXTURE/$SAMPLE --n_ind 73 --n_sites NUMBER_OF_SITES --out $SAMPLE.genetic_distances_test

../ngsDist --n_threads 10 --seed 12345 --verbose 1 --geno testA_32.geno --probs --n_ind $N_IND --n_sites $N_SITES --labels testA.labels                                                                                  --out testA_32.dist


```

--geno FILE: input file with genotypes, genotype likelihoods or genotype posterior probabilities.
--n_ind INT: sample size (number of individuals).
--n_sites INT: number of sites in input file.
--tot_sites INT: total number of sites in dataset.
--labels FILE: labels, one per line, of the input sequences.
--probs: is the input genotype probabilities (likelihoods or posteriors)?
--log_scale: Ii the input in log-scale?.
--call_geno: call genotypes before running analyses.
--N_thresh DOUBLE: minimum threshold to consider site; missing data if otherwise (assumes -call_geno)
--call_thresh DOUBLE: minimum threshold to call genotype; left as is if otherwise (assumes -call_geno)
--pairwise_del: pairwise deletion of missing data.
--avg_nuc_dist: use average number of nucleotide differences as distance (by default, ngsDist uses genotype distances based on allele frequency differences). Only pairs of heterozygous positions are actually affected when using this option, with their distance being 0.5 (instead of 0 by default).
--indep_geno: assume independence between genotypes?
--n_boot_rep INT: number of bootstrap replicates [0].
--boot_block_size INT: block size for bootstrapping [1].
--out FILE: output file name.
--n_threads INT: number of threads to use. [1]
--version: prints program version and exits.
--verbose INT: selects verbosity level. [1]
--seed INT: random number generator seed (only for the bootstrap analysis).