---
title: "Untitled"
output: html_document
---


# Generating Dxy 

## Introduction and populations:

I am going to run MAF & Dxy in my target populations.

Populations:

c_ll_no -->
c_ll_po -->
c_ll_ki -->

c_lp_do --> lanzado 
c_lp_sm --> lanzado


## Global variables:

```{r, engine=bash, eval=FALSE}

RUTA=/home/mlucena/ANGSD_analysis 
cd $RUTA/whole_genome_analysis/dxy

#To launch one by one

POP="c_lp_sm_n019"  # <--CHANGE POP HERE

screen -S "$POP"_Dxy
# screen -S "$POP"_thetas

POP="c_lp_sm_n019"  # <--CHANGE POP HERE

script "$POP"_Dxy.log
# script "$POP"_thetas.log

POP="c_lp_sm_n019"  # <--CHANGE POP HERE

THREADS=15                     # no. of computer cores used by bwa and samtools. 20 = OK, >20 = ask people first!

RUTA=/home/mlucena/ANGSD_analysis/depth_calculus 
read POP mean sd mean_truncated sd_truncated maxDepth minDepth maxDepth_truncated minDepth_truncated < $RUTA/"${POP}"_mean_sd_depthGlobal_lynx_per_pop_mean_folds_0.95.csv

ANGSD="/opt/angsd/angsd"
NGSTOOLS="/opt/angsd/angsd/misc"
REF="/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23_without_repetitive_transposable_low_complexity.fa"
ANC="/home/GRUPOS/grupolince/reference_genomes/lynx_rufus_genome/c_lr_zz_0001_recal1.fa"
FILTER1=" -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -baq 1 -C 50 "
FILTER2=" -minMapQ 30 -minQ 20 -doCounts 1 "
N_IND=$(echo ${POP: -3} )
MIN_IND=$(expr $N_IND / 2)
RUTA=/home/mlucena/ANGSD_analysis 
SNP_PVAL=rrrrr


# Sanity checks:

echo $POP
echo $N_IND
echo $MIN_IND
echo $maxDepth_truncated
echo $minDepth_truncated

```

## Calculating MAF

```{bash}


$ANGSD/angsd -nThreads $THREADS -bam $RUTA/whole_genome_analysis/"$POP".bamlist -ref $REF -anc $ANC \
-out "$POP".genolike \
$FILTER1 \
$FILTER2 \
-GL 1 -doMajorMinor 1 -doMaf 2  \
-minInd  $MIN_IND -setMaxDepth $maxDepth_truncated -setMinDepth $minDepth_truncated \
-SNP_pval $SNP_PVAL -skipTriallelic 1


# -doMaf 2 Esto te saca el known y el unknown: http://www.popgen.dk/angsd/index.php/Allele_Frequencies
## Cuidado porque el minor y major puede no coincidir, y todas las cuentas que hagas despues estarán mal. Puedes forzarlo haciendo Maf 4

```


######### OJO!!!!!######### OJO!!!!!######### OJO!!!!!######### OJO!!!!!######### OJO!!!!!

# Cuidado porque el minor y major puede no coincidir, y todas las cuentas que hagas despues estarán mal. Puedes forzarlo haciendo Maf 4

######### OJO!!!!!######### OJO!!!!!######### OJO!!!!!######### OJO!!!!!######### OJO!!!!!


We can now calculate some summary statistics, namely number of segregating sites, expected heterozygosity, number of fixed differences and dxy. Please note that the latter 2 statistics have not been properly tested and they are shown to be For instance, dxy been shown to be over-estimated and should be used only for inspecting the distribution and not to make inferences based on its absolute values. In case you want to estimate dxy, you can find in ngsTools/ngsPopGen/scripts folder a perl script written by Nagarjun Vijay and a R script written by [Joshua Penalba][https://joshuapenalba.com/] which calculate Dxy from ANGSD allele frequency files. Please see that script for additional help.

--> We are using Nagarjun Vijay script. 

```{bash}


POP1=c_lp_do_n012.genolike.mafs.gz
POP2=c_lp_sm_n019.genolike.mafs.gz

# Esta es la pinta del archivo maf
# chromo	position	major	minor	ref	anc	unknownEM	pu-EM	nInd
# lp23.s00001	42503	G	T	G	G	0.051070	6.026475e-02	11
# lp23.s00001	62369	G	A	G	G	0.097718	1.830133e-02	12
# lp23.s00001	62377	G	T	G	G	0.092205	3.373843e-02	12
# lp23.s00001	94627	T	A	T	A	0.233673	1.110223e-16	12
# lp23.s00001	94676	A	T	A	T	0.301259	3.219647e-15	12
# lp23.s00001	110613	C	A	N	C	0.055303	3.387711e-02	12
# lp23.s00001	112776	A	G	A	A	0.044053	1.246238e-02	12
# lp23.s00001	118717	A	G	A	A	0.311334	0.000000e+00	12
# lp23.s00001	120814	G	C	G	N	0.421426	0.000000e+00	12


zcat c_lp_do_n012.genolike.mafs.gz | awk -v OFS='\t' '{printf ("%s\t%s\t%s\t%s\t%s\t%f\t%s\t%s\n", $1, $2, $3, $4, $5, (1-$7), $7, $9)}'  > prueba_1
zcat c_lp_sm_n019.genolike.mafs.gz | awk -v OFS='\t' '{printf ("%s\t%s\t%s\t%s\t%s\t%f\t%s\t%s\n", $1, $2, $3, $4, $5, (1-$7), $7, $9)}'  > prueba_2

# Tendría que filtrarlo para pu-EM

printf ("%.8f\n",n1-n2)}

POP1=c_lp_do_n012.genolike.mafs.gz
POP2=c_lp_sm_n019.genolike.mafs.gz

./getDxy.pl --pop1maf $POP1 --pop2maf $POP2 --nInd 31

./getDxy.pl --pop1maf prueba_1 --pop2maf prueba_2 --minInd 13


```

## getDxy.pl --> Nagarjun Vijay

```{perl}
#!/usr/bin/perl

#use strict;
use warnings;
use Getopt::Long;

my $pop1maf;
my $pop2maf;
my $minInd;

&GetOptions( 'pop1maf=s' => \$pop1maf,
             'pop2maf=s' => \$pop2maf,
             'minInd=i' => \$minInd,
			 );

my $usage = <<_EOUSAGE_;
#########################################################################################
# getDxy.pl --pop1maf <FILE> --pop2maf <FILE> --nInd <Integer>
# Required:
#  --pop1maf        a text maf file output from ANGSD with apporpriate filtering for Population 1
#  --pop2maf        a text maf file output from ANGSD with apporpriate filtering for Population 2
#  --minInd         minimum number of individuals required per population
#
#This script assumes equal and corresponding lines in the two maf files. User needs to filter for variable sites using the SNP pval before running this script.
#
#
#Following columns need to be present:
#
#chromo	position	major	minor	ref	knownEM	unknownEM	nInd
#
#
#Dxy is reported only for the sites included in the MAF file. While calculating the value per window, the correct number of sites has to be used.
#
#Example command to run the script
#perl getDxy.pl --pop1maf pop1.pop1_pop2.genotypes.mafs.txt --pop2maf pop2.pop1_pop2.genotypes.mafs.txt --minInd 5
###########################################################################################
_EOUSAGE_

	;
 
if (! defined $pop1maf) {print $usage;exit;}
if (! defined $pop2maf) {print $usage;exit;}
if (! defined $minInd) {print $usage;exit;}

open POP1MAF, $pop1maf or die $!;
open POP2MAF, $pop2maf or die $!;

my $line1;
my $line2;

my $dxy=0;
#read in the header and do nothing
my $header1=<POP1MAF>;my $header2=<POP2MAF>;

print "chromo\tposition\tDxy\n";

while($line1=<POP1MAF>){
        #read in maf from pop1
        chomp $line1;my @parts=split('\t',$line1);
        
        #read in maf from pop2
        $line2=<POP2MAF>;chomp $line2;my @parts2=split('\t',$line2);
        
if(($parts[6]>=$minInd)&&($parts2[6]>=$minInd)){#use only sites that are covered by atleast $minInd individuals in each population

        #if($parts3[4]>0.99999999){#use only sites with pvar >0.99999999, same as criteria used for fst

                if(($parts[2]=~/$parts2[2]/)&&($parts[3]=~/$parts2[3]/)){#check if the major and minor allele are matching

                $dxy=$parts[4]*(1-$parts2[4])+($parts2[4]*(1-$parts[4]));

                print "$parts[0]\t$parts[1]\t$dxy\n"; # print scaffold, position and per site dxy

                }

                if(($parts[2]=~/$parts2[3]/)&&($parts[3]=~/$parts2[2]/)){#check if the major and minor allele are NOT matching

                $dxy=($parts[4]*$parts2[4])+((1-$parts2[4])*(1-$parts[4]));

                print "$parts[0]\t$parts[1]\t$dxy\n";# print scaffold, position and per site dxy

                }
}}
close POP1MAF;
close POP2MAF;
```


## Dxy from mafs files --> Joshua Penalba

```{r}
#########################################
#                                       #
#   Calculates Dxy from mafs files      #
#                                       #
#   Author: Joshua Penalba              #
#   Date: 22 Oct 2016                   #
#                                       #
#########################################


# NOTES
# * Prior to calculating Dxy the following steps are recommended:
#   1. Run ANGSD with all populations with a -SNP_pval and -skipTriallelic flags.
#   2. Rerun ANGSD per population 
#       Use the -sites flag with a file corresponding to the recovered SNPs.
#       This will guarantee that sites with an allele fixed in one population is still included.
#       Remove the -SNP_pval flag.
#       IMPORTANT: Include an outgroup reference to polarize alleles.
#   3. Gunzip the resulting mafs files.
# 
# * Make sure the totLen only includes the chromosomes being analyzed.
# * minInd flag not added, assuming already considered in the ANGSD run.
# * Test for matching major and minor alleles not included as it would filter out sequencing errors. 
#   This has been accounted for in the allele frequency calculations.
#   This filter may give an underestimate of dxy.
# * Per site Dxy of ~0 could be common if the alternate alleles are present in a population other than the two being included in the calculation.


### Creating an argument parser
library("optparse")

option_list = list(
  make_option(c("-p","--popA"), type="character",default=NULL,help="path to unzipped mafs file for pop 1",metavar="character"),
  make_option(c("-q","--popB"), type="character",default=NULL,help="path to unzipped mafs file for pop 2",metavar="character"),
  make_option(c("-t","--totLen"), type="numeric",default=NULL,help="total sequence length for global per site Dxy estimate [optional]",metavar="numeric")
)
opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

### Troubleshooting input
if(is.null(opt$popA) | is.null(opt$popB)){
  print_help(opt_parser)
  stop("One or more of the mafs paths are missing", call.=FALSE)
}

if(grepl('.gz$',opt$popA) | grepl('.gz$',opt$popB)){
  print_help(opt_parser)
  stop("One or more of the mafs is gzipped.", call.=FALSE)
}

if(is.null(opt$totLen)){
  print("Total length not supplied. The output will not be a per site estimate.")
}



### Reading data in
allfreqA <- read.table("prueba_1",sep='\t',row.names=NULL, header=T)
allfreqB <- read.table("prueba_2",sep='\t',row.names=NULL, header=T)


### Manipulating the table and print dxy table
allfreq <- merge(allfreqA, allfreqB, by=c("chromo","position"))
allfreq <- allfreq[order(allfreq$chromo, allfreq$position),]
# -> Actual dxy calculation
allfreq <- transform(allfreq, dxy=(knownEM.x*(1-knownEM.y))+(knownEM.y*(1-knownEM.x)))
write.table(allfreq[,c("chromo","position","dxy")], file="Dxy_persite.txt",quote=FALSE, row.names=FALSE, sep='\t')
print('Created Dxy_persite.txt')




# Print global dxy
print(paste0('Global dxy is: ',sum(allfreq$dxy)))
if(!is.null(opt$totLen)){
  print(paste0('Global per site Dxy is: ',sum(allfreq$dxy)/opt$totLen))
}


print(paste0('Global per site Dxy is: ',sum(allfreq$dxy)/407194))

```


 "Global dxy is: 155416.991753343"
> len(allfreq)
Error in len(allfreq) : no se pudo encontrar la función "len"
> lenght(allfreq)
Error in lenght(allfreq) : no se pudo encontrar la función "lenght"
> length(allfreq)
[1] 15
> print allfreq
Error: unexpected symbol in "print allfreq"
> print (allfreq)
^C
> nrow(all)
all                    all.equal.environment  all.equal.language     all.equal.raw          allfreqA               
all.equal              all.equal.envRefClass  all.equal.list         all.names              allfreqB               
all.equal.character    all.equal.factor       all.equal.numeric      all.vars               allGenerics            
all.equal.default      all.equal.formula      all.equal.POSIXt       allfreq                allNames               
> nrow(allfreq)
[1] 407194
> print(paste0('Global per site Dxy is: ',sum(allfreq$dxy)/407194))
[1] "Global per site Dxy is: 0.38167800054358"


# Divergence --> substitutions


OJO! he movido algunos de los archivos a una ruta anterior --> check "7b.Divergence"

VCF creado por Dani. Email:

Hola María,

En el servidor B, en 

/home/GRUPOS/grupolince/lynx_genomes_5x/VCFs_Dani 

he creado:

un archivo con la lista de posiciones en c_ll_ki_0090 y c_lr_zz_0001 (c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions), 
de ahí otro archivo con la clasificación de cada posición (c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions_class), 
y ahora estoy generando un tercero sin NAs, INDELs o multialélicos (c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions_class_callable_universe). 

Lo que quedaría lo podemos definir como el universo callable.
Puedes usar este tercero para tus cálculos por ventana. 

Tengo que hacer ventanas de 50000pb contando el número de posiciones callable y las sustituciones en estas. 

Debo reportar:

1. Scaffold
2. Win start, win end
3. Win center
4. Number of substitutions
5. Number of callable sites

¿Cómo es el archivo?

Scaffold Position Reference Alternative c_ll_ki_0090 c_lr_zz_0001 Class
lp23.s00001	1	C	.	0/0	0/0	 Invariable
lp23.s00001	2	T	.	0/0	0/0	 Invariable
lp23.s00001	3	T	.	0/0	0/0	 Invariable
lp23.s00001	4	G	.	0/0	0/0	 Invariable
lp23.s00001	5	T	.	0/0	0/0	 Invariable
lp23.s00001	7	G	.	0/0	0/0	 Invariable
lp23.s00001	8	T	.	0/0	0/0	 Invariable
lp23.s00001	9	T	.	0/0	0/0	 Invariable
lp23.s00001	10	G	.	0/0	0/0	 Invariable

# OJO! si lo quieres volver a correr lo suyo es general unas ventanas con todas las posibles y correrlo sobre eso en lugar de sobre el de una población concreta. Nosotros lo hicimos sobre carpatos y kirov, pero como faltaban 4 ventanas las hemos añadido. Ver adding new windows al final. 


```{bash}

# Primero lo hago cero based.Y convierto todas las "SUSBSTITUTIONS" en 1 y lo demás en 0, para que me sea muy comodo hacer el slidding window. 

mkdir -p /home/mlucena/divergence/intergenic_analysis

WD="/home/mlucena/divergence/intergenic_analysis"
WINDOW_DIRECTORY="/home/mlucena/datos/intermediate_ANGSD_files/intergenic_analysis/intergenic_sfs"
FILE_DANI="/home/GRUPOS/grupolince/lynx_genomes_5x/VCFs_Dani/c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions_class_callable_universe"
FILE_DANI_MODIF="c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions_class_callable_universe_summarized"

# Modifico el archivo de Dani:
tail -n+2  $FILE_DANI | awk '{if ($7=="Substitution") print $1, $2-1, "1"; else print $1, $2-1, "0"}' > $WD/$FILE_DANI_MODIF


# Nombre de un archivos por ventanas:
# c_ll_ya_n008.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions

# Lo que hago es coger el archivo de diversidad por ventanas de un par de poblaciones (en este caso kirov, tuva y cr, con la idea de juntarlos luego para que no falte ninguna ventana) y correr este script definiendo el rango que tienen que coger usando ese rango de ventanas estudiadas para esas poblaciones. Idealmente si estuvieran cubiertas completamente todas las ventanas en todas las poblaciones los archivos de salida serían idénticos, puesto que la divergencia se establece usando la poblacion de kirov-rufus. 

cd $WINDOW_DIRECTORY

# POPS USADAS

# POP=c_ll_cr_n006
POP=c_ll_ki_n013
# POP=c_ll_tu_n006

screen -S $POP.divergence

# POP=c_ll_cr_n006
POP=c_ll_ki_n013
# POP=c_ll_tu_n006

WD="/home/mlucena/divergence/intergenic_analysis"
WINDOW_DIRECTORY="/home/mlucena/datos/intermediate_ANGSD_files/intergenic_analysis/intergenic_sfs"
FILE_DANI="/home/GRUPOS/grupolince/lynx_genomes_5x/VCFs_Dani/c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions_class_callable_universe"
FILE_DANI_MODIF="c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions_class_callable_universe_summarized"
cd $WD
echo $POP
rm $POP.divergence_with_rufus


# Lo que en awk llamo sum son los callable sites y los que llamos count son las substituciones.

while read CHR WinStart WinStop WinCenter tW tP tF tH tL Tajima fuf fud fayh zeng nSites WinNumber AUTOSOMICpositions Xpositions

do

echo "Window: " "$CHR" "$WinStart"-"$WinStop"

grep $CHR $FILE_DANI_MODIF | awk -v CHR="$CHR" -v WinStart="$WinStart" -v WinStop="$WinStop" '$2>WinStop { exit } $2>=WinStart && $2<=WinStop{sum+=$3;count++;}END{print CHR, WinStart, WinStop, count, sum, sum/count} ' >> $POP.divergence_with_rufus

done < <(tail -n+2  "$WINDOW_DIRECTORY"/"$POP".unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions)

# Cuando acabe borrar: $FILE_DANI_MODIF

# Sanity checks:

diff c_ll_cr_n006.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions c_ll_ki_n013.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions

diff c_ll_cr_n006.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions c_ll_tu_n006.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions

diff c_ll_ki_n013.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions c_ll_tu_n006.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions



# Si tiene todos los campos que debiera podemos hacer un join con el archivo de diversidad.


```

Copio a local

```{bash}
mkdir /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_sfs/divergence_window

scp  mlucena@genomics-b.ebd.csic.es:/home/mlucena/divergence/intergenic_analysis/*divergence_with_rufus /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_sfs/divergence_window
```

### R comprobaciones

```{r}
library(dplyr)
library(ggplot2)
library (tidyr)

wd <- "/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_sfs/divergence_window/"


divergence_cr <- read.table(paste (wd,"c_ll_cr_n006.divergence_with_rufus", sep =""), sep=" ", dec=".", header=F) 

divergence_ki <- read.table(paste (wd,"c_ll_ki_n013.divergence_with_rufus", sep =""), sep=" ", dec=".", header=F) 


full_join_divergence <- full_join(divergence_cr, divergence_ki, by =c("V1","V2","V3"))

```

### How I should have done it

Observamos que cuando hay datos, los valores son los mismo, pero inesperadamente hay ventanas que no están cubiertas en kirov con respecto a carpatos y al contrario. 
La he cagado y tendría que haber sacado un archivo con todas las ventanas de este modo:

```{bash}
cd /home/mlucena/datos/intermediate_ANGSD_files/intergenic_analysis/intergenic_sfs

cat *.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions | awk '{print $1,$2,$3}' | sort -k1,1 -k2,2n -k3,3n | uniq 

# o con makewindow (la herramienta de bedtools).

```



De todas formas voy a ver si juntando kirov y carpatos tengo todas o casi todas las ventanas posibles:

¿Cuantas ventanas tendría si junto todo?

```{bash}
cat *.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions | awk '{print $1,$2,$3}' | sort -k1,1 -k2,2n -k3,3n | uniq | wc -l
#41440
```

¿Y si junto kirov y carpatos?

```{bash}
 cat c_ll_ki_n013.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions c_ll_cr_n006.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions | awk '{print $1,$2,$3}' | sort -k1,1 -k2,2n -k3,3n | uniq | wc -l
# 41436
```

Me faltan 4 ventanas. ¿cuales son?

```{bash}
diff <(cat *.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions | awk '{print $1,$2,$3}' | sort -k1,1 -k2,2n -k3,3n | uniq) <(cat c_ll_ki_n013.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions c_ll_cr_n006.unfolded-lr.postprob.thetasWindow_50000.50000.gz.pestPG.with_x_autosomic_positions | awk '{print $1,$2,$3}' | sort -k1,1 -k2,2n -k3,3n | uniq)

5097d5096
< lp23.s00280 100000 150000
15341d15339
< lp23.s10659 200000 250000
35673d35670
< lp23.s31439 400000 450000
36171d36167
< lp23.s31555 100000 150000


```

Puedo calcular la divergencia en estas ventanas y añadirlo a posteriori.

Primero voy a crear un archivo en comun con las ventanas de carpatos y kirov sobre el que añadir esta información. Lo voy a hacer en R que me parece más facil y luego lo llevo al servidor.

### Divergence both kirov and carpathians.

```{r}
library(dplyr)
library(ggplot2)
library (tidyr)

wd <- "/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_sfs/divergence_window/"


divergence_cr <- read.table(paste (wd,"c_ll_cr_n006.divergence_with_rufus", sep =""), sep=" ", dec=".", header=F) 

divergence_ki <- read.table(paste (wd,"c_ll_ki_n013.divergence_with_rufus", sep =""), sep=" ", dec=".", header=F) 


full_join_divergence <- full_join(divergence_cr, divergence_ki, by =c("V1","V2","V3"))

# Esto de pmax coge el valor máximo. Como yo quiero hacer un merged y en ambas columnnas el valor simpre es el mismo excepto cuando en alguna de las dos no hay valor, me valdría esta aproximación.

c_ll_ki_n013_c_ll_cr_n006_divergence <- full_join_divergence %>%  mutate (callable_sites = pmax(V4.y, V4.x, na.rm=TRUE)) %>% mutate (substitutions = pmax(V5.y, V5.x, na.rm=TRUE)) %>% mutate (divergence = pmax(V6.y, V6.x, na.rm=TRUE)) %>% rename(scaffold=V1,winstart=V2,winend=V3) %>% select(scaffold, winstart, winend, callable_sites, substitutions, divergence)


write.table(c_ll_ki_n013_c_ll_cr_n006_divergence, paste(wd, "c_ll_ki_n013-c_ll_cr_n006.divergence_with_rufus", sep=""),quote=F, row.names = F, sep=' ')


```


La copio al servidor:

```{bash}

scp /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_sfs/divergence_window/c_ll_ki_n013-c_ll_cr_n006.divergence_with_rufus mlucena@genomics-b.ebd.csic.es://home/mlucena/divergence/intergenic_analysis/




```



# Adding new windows.

```{bash}

cd /home/mlucena/divergence/intergenic_analysis
scp c_ll_ki_n013-c_ll_cr_n006.divergence_with_rufus c_ll.divergence_with_rufus

screen -S ventana_1
FILE_DANI_MODIF="c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions_class_callable_universe_summarized"
POP=c_ll
CHR=lp23.s00280
WinStart=100000
WinStop=150000

screen -S ventana_2
FILE_DANI_MODIF="c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions_class_callable_universe_summarized"
POP=c_ll
CHR=lp23.s10659
WinStart=200000
WinStop=250000

screen -S ventana_3
FILE_DANI_MODIF="c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions_class_callable_universe_summarized"
POP=c_ll
CHR=lp23.s31439 
WinStart=400000
WinStop=450000

screen -S ventana_4
FILE_DANI_MODIF="c_ll_ki_0090_plus_c_lr_zz_0001_recal_round-1_25x_SNPs.positions_class_callable_universe_summarized"
POP=c_ll
CHR=lp23.s31555
WinStart=100000
WinStop=150000


grep $CHR $FILE_DANI_MODIF | awk -v CHR="$CHR" -v WinStart="$WinStart" -v WinStop="$WinStop" '$2>WinStop { exit } $2>=WinStart && $2<=WinStop{sum+=$3;count++;}END{print CHR, WinStart, WinStop, count, sum, sum/count} ' >> $POP.divergence_with_rufus

```

Ahora podria eliminar los otros archivos.

```{bash}
cd /home/mlucena/divergence/intergenic_analysis

rm c_ll_cr_n006.divergence_with_rufus
rm c_ll_ki_n013.divergence_with_rufus
rm c_ll_ki_n013-c_ll_cr_n006.divergence_with_rufus

```







# Download last divergence file

```{bash}
scp  mlucena@genomics-b.ebd.csic.es://home/mlucena/divergence/intergenic_analysis/c_ll.divergence_with_rufus /Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/intergenic_analysis/analysis_sfs/divergence_window/
```


