---
title: "8.Relatedness"
output: html_document
---


# Relatedness

http://www.popgen.dk/software/index.php/NgsRelate

Assume we have file containing paths to 100 BAM/CRAM files; one line per BAN/CRAM file. Then we can use ANGSD to estimate frequencies and calculate genotype likelihoods while doing SNP calling and in the end produce the the input files needed for the NgsRelate program as follows:

Maria note: This frequencies make sense if they are coming from a populations. Therefore we set some populations to be able to calculate those frecuencies. 

```{bash}

mkdir /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_relatedness

cd /home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_relatedness/

POP="c_ll_ya_n008"
POP="c_ll_cr_n006"
POP="c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080"  # <--CHANGE POP HERE


THREADS=15 
read POP mean sd mean_truncated sd_truncated maxDepth minDepth maxDepth_truncated minDepth_truncated < /home/mlucena/ANGSD_analysis/depth_calculus/${POP}_mean_sd_depthGlobal_lynx_per_pop_mean_folds_0.95.csv

ANGSD="/opt/angsd/angsd"
NGSTOOLS="/opt/angsd/angsd/misc"
REF="/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23_without_repetitive_transposable_low_complexity.fa"
ANC="/home/GRUPOS/grupolince/reference_genomes/lynx_rufus_genome/c_lr_zz_0001_recal1.fa"
FILTER1=" -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -baq 1 -C 50 "
FILTER2=" -minMapQ 30 -minQ 20 -doCounts 1 "
N_IND=$(echo ${POP: -3} )
MIN_IND=$(expr $N_IND / 2)
# REGIONFILE="/home/mlucena/ANGSD_analysis/depth_calculus/no_genes_Lypa_10000longest_center_final_slop20_dot.rf"


SNP_PVAL=1e-4

# Sanity checks:

echo $POP
echo $N_IND
echo $MIN_IND
echo $maxDepth
echo $minDepth
echo $SNP_PVAL

### First we generate a file with allele frequencies (angsdput.mafs.gz) and a file with genotype likelihoods (angsdput.glf.gz).

$ANGSD/angsd -P $THREADS -b /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP.intergenic.bamlist -ref $REF -out $POP.intergenic \
$FILTER1 $FILTER2 \
-GL 1 -doMajorMinor 1 -doMaf 1 -skipTriallelic 1 \
-SNP_pval $SNP_PVAL \
-minmaf 0.05 -doGlf 3 \
-minInd $MIN_IND -setMaxDepth $maxDepth -setMinDepth $minDepth 
# -rf $REGIONFILE \

cat /home/mlucena/ANGSD_analysis/intergenic_analysis/$POP*.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 2,3,4,5 | rev  > id

# Modifico el id de h_ba a mano y lo cambio por c_ba para que funcione luego la comparación a pares.

### Then we extract the frequency column from the allele frequency file and remove the header (to make it in the format NgsRelate needs)
zcat $POP.intergenic.mafs.gz | cut -f6 | sed 1d > $POP.freq


# Once we have these files we can use NgsRelate to estimate relatedness between any pairs of individuals. E.g. if we want to estimate relatedness between the first two individuals (numbered from 0, so 0 and 1) we can do it using the following command:

/home/mlucena/ngsRelate -g $POP.intergenic.glf.gz -n $N_IND -f $POP.freq -z id > $POP.gl.res


cat $POP.gl.res | awk '{if (substr($3, 1, 7)==substr($4, 1, 7)) print $0}' > $POP.perpop.gl.res

# Here we specify the name of our file with genotype likelihoods after the option "-g", the number of individuals in the file after the option "-n", the name of the file with allele frequencies after the option "-f" and the number of the two individuals after the options "-a" and "-b" . If -a and -b are not specified NgsRelate will loop through all pairs of individuals in the input file.

# NEW: Note that if you want you also input a file with the IDs of the individuals (on ID per line) in the same order as in the file 'filelist' used to make the genotype likelihoods. If you do the output will also contain these IDs and not just the numbers of the samples (one can actually just use that exact file, however the IDs then tend to be a bit long). This can be done with the optional flag -z followed by the filename.

# Historically, several summary statistics have been used, such as the kinship coefficient θ, however almost all of these statistics can be calculated from  R=(k0,k1,k2), where km is the fraction of genome in which the two individuals share m alleles IBD

# Relationship	      K_0	  K_1	  K_2
# mono-zygotic twin	    0 	  0	    1 
# Parent-Offspring	    0 	  1	    0 
# Full siblings	      0.25  0.5  	 0.25 
# Half siblings	      0.5  	0.5 	  0 
# First cousins	      0.75  0.25 	  0  
# Unrelated	            1 	  0 	  0  


# The first two columns contain the information of about what two individuals was used for the analysis. The third column contains information about how many sites were used in the analysis. The following three columns are the maximum likelihood (ML) estimates of the relatedness coefficients. The seventh column is the log of the likelihood of the ML estimate. The eigth column is the number of iterations of the maximization algorithm that was used to find the MLE, and finally the ninth column is fraction of non-missing sites, i.e. the fraction of sites where data was available for both individuals, and where the minor allele frequency (MAF) above the threshold (default is 0.05 but the user may specify a different threshold). Note that in some cases nIter is -1. This indicates that values on the boundary of the parameter space had a higher likelihood than the values achieved using the EM-algorithm (ML methods sometimes have trouble finding the ML estimate when it is on the boundary of the parameter space, and we therefore test the boundary values explicitly and output these if these have the highest likelihood).



scp mlucena@genomics-b.ebd.csic.es:/home/mlucena/ANGSD_analysis/intergenic_analysis/intergenic_relatedness/*.res  /Users/marialucenaperez/Dropbox/phd/contemporary/ANGSD/intergenic_analysis/analysis_relatedness 

```

```{r}
library("dplyr")

POP="c_ll_ya_n008"
POP="c_ll_cr_n006"

wd <- "/Users/marialucenaperez/Dropbox/phd/contemporary/ANGSD/intergenic_analysis/analysis_relatedness/"
res <- read.csv(paste(wd,POP,".gl.res", sep=""), sep="\t", dec=".")
df_relatedness <- res %>% mutate (k_total=((k0*0)+(k1*0.5)+(k2*1))) %>% arrange(., desc(k_total)) %>% filter (k_total!=0) %>% select (-a, -b, -loglh)
write.table(df_relatedness, paste(wd,POP,"_relatedness.csv", sep=""), row.names = F, quote = F ,dec=",", sep=";")




```

