---
title: "PCA_Karolina_Scaffold_outlier"
output: html_document
---



```{bash}

# Create directory

mkdir /home/mlucena/grupolince/PCA_outlier_Karolina


# I want to produce PCAs for the whole dataset, each specie, and within iberian with and without historical samples. 
# I think that per population PCAs make no sense

#To launch one by one ( because otherwise it's too slow I'll launch single pops in a and groups in b)
#Lanzo cada uno en una sesión de screen diferente y con un log distinto ( mediante script). lanzo el script conpleto dandole como input el csv que acabo de crear. Ejemplo:

screen -S pca__lynx_lynx_oriental_Karolina

POP="c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl"  # <--CHANGE POP HERE

cd /home/mlucena/grupolince/lynx_genomes_5x/BAM_files_final 
ls -d $PWD/*_ll_*recal_round-1.bam | grep -v ll_ba_ >  /home/mlucena/grupolince/PCA_outlier_Karolina/ll.bamlist

read POP mean sd mean_truncated sd_truncated maxDepth minDepth maxDepth_truncated minDepth_truncated < /home/mlucena/ANGSD_analysis/depth_calculus/"$POP"_n*_mean_sd_depthGlobal_lynx_per_pop_mean_folds_0.95.csv

cd /home/mlucena/grupolince/PCA_outlier_Karolina/

# Set paths to software and some commom files

ANGSD=/opt/angsd/angsd
NGSTOOLS="/opt/ngsTools"
THREADS=20                     # no. of computer cores used by bwa and samtools. 20 = OK, >20 = ask people first!
REF="/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23_without_repetitive_transposable_low_complexity.fa"
ANC="/home/GRUPOS/grupolince/reference_genomes/lynx_rufus_genome/c_lr_zz_0001_recal1.fa"
FILTER1=" -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -baq 1 -C 50 "
FILTER2=" -minMapQ 30 -minQ 20 -doCounts 1 "
N_IND=$(echo ${POP: -3} )
MIN_IND=$(expr $N_IND / 2) 


SNP_PVAL="1e-4"



# Sanity checks:

echo $POP
echo $N_IND
echo $MIN_IND
echo $maxDepth
echo $minDepth
echo $SNP_PVAL


# ojo! Hemos aplicado filtros para todo aunque no incluyendo balcanes!

### Calling genotype likelihoods: 
 
$ANGSD/angsd -P $THREADS -b ll.bamlist -ref $REF -out $POP.intergenic \
$FILTER1 $FILTER2 \
-r lp23.s15646:3000000-3875154 \
-minInd $MIN_IND -setMaxDepth $maxDepth -setMinDepth $minDepth  \
-GL 1 -doMajorMinor 1 -doMaf 1 -skipTriallelic 1 \
-SNP_pval $SNP_PVAL \
-doGeno 32 -doPost 1

# Unzip the results (but you cannot open it since it is in binary format)

gunzip $POP.intergenic.geno.gz

### ngsCovar


#less -S $POP.mafs.gz
N_SITES=`zcat $POP.intergenic.mafs.gz | tail -n+2 | wc -l` 

# Hago tail -n+2 porque tengo cabecero. 

echo $N_SITES 


# Now we can perform a PCA by estimating the covariance matrix first:
# Change number of individuals
N_IND=77

$NGSTOOLS/ngsPopGen/ngsCovar -probfile $POP.*geno -outfile $POP.covar -nind $N_IND  -nsites $N_SITES -call 0 -norm 0 &> /dev/null


echo "FID" > $POP.ALL1.clst.borrar
seq $N_IND >> $POP.ALL1.clst.borrar
echo "IID" > $POP.ALL2.clst.borrar
# yes "1" | head -n $N_IND >> $POP.ALL2.clst.borrar
cat ll.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 3,4,5,6 | rev  >> $POP.ALL2.clst.borrar
echo "CLUSTER" > $POP.ALL3.clst.borrar
cat ll.bamlist | rev | cut -d "/"  -f 1 | cut -d "_" -f 4,5,6 | rev  >> $POP.ALL3.clst.borrar
paste $POP.ALL1.clst.borrar $POP.ALL2.clst.borrar $POP.ALL3.clst.borrar> $POP.ALL.clst

rm $POP.*clst.borrar


```


## Copying to Dropbox folder


```{bash}

scp mlucena@genomics-b.ebd.csic.es:/home/mlucena/grupolince/PCA_outlier_Karolina/*.ALL.clst /Users/marialucenaperez/Owncloud/publico/test_karolina
scp mlucena@genomics-b.ebd.csic.es:/home/mlucena/grupolince/PCA_outlier_Karolina/*.covar /Users/marialucenaperez/Owncloud/publico/test_karolina

```


## R representation

```{r "setup", include=FALSE}

library(methods)
library(optparse)
library(ggplot2)
library(dplyr)
library(threejs)
library(RColorBrewer)
require("knitr")


WORKING_DIR="/Users/marialucenaperez/Owncloud/publico/test_karolina/"
# samples_info <-read.csv("/Users/marialucenaperez/Dropbox/PhD/contemporary/ANGSD/analysis_PCA/test_intergenic_elena/c_ll_ba-c_ll_cr-c_ll_ka-c_ll_ki-c_ll_la-c_ll_no-c_ll_og-c_ll_po-c_ll_to-c_ll_tu-c_ll_vl-c_ll_ya-c_lp_do-c_lp_sm-h_ll_ba.ALL.clst")
pop=c("c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080")
# OJO!!!! Aunqeu se llama así, en realidad no tiene balcanes!!!
# 
###########################################################################################
#for (pop in pops) 
#{

# Annotation file is in plink cluster format

#################################################################################

# Read input file
covar <- read.table(paste(WORKING_DIR,pop, ".covar", sep=''), stringsAsFact=F);

# Read annot file
annot <- read.table(paste(WORKING_DIR,pop, ".ALL.clst", sep=""), header=TRUE) 

# Parse components to analyze
# comp <- as.numeric(strsplit("component1-component2", "-", fixed=TRUE)[[1]])
# comp <- as.numeric(strsplit(opt$comp, "-", fixed=TRUE)[[1]])

# comp <- c(1,2)
# comp <- c(2,3)
# comp <- c(1,3)


# Eigenvalues
eig <- eigen(covar, symm=TRUE);
eig$val <- eig$val/sum(eig$val);
cat(signif(eig$val, digits=3)*100,"\n");
eig$values


# Plot
PC <- as.data.frame(eig$vectors)
colnames(PC) <- gsub("V", "PC", colnames(PC))
PC$Pop <- factor(annot$CLUSTER)
PC$Ind <- factor(annot$IID)
# PC <- PC %>% filter(Ind == "UA")



PC <- PC %>%  mutate(., 
         Subspecies = ifelse (Pop == "c_ll_po" | Pop == "c_ll_ki" | Pop == "c_ll_la" | Pop == "c_ll_no"  | Pop == "c_ll_ur", "L.l.lynx",        
                          ifelse(Pop == "c_ll_ba" | Pop == "h_ll_ba", "L.l.balcanicus" ,
                          ifelse(Pop == "c_ll_cr","L.l.carpathicus",
                          ifelse(Pop == "c_ll_to" | Pop == "c_ll_tu" | Pop == "c_ll_ka", "L.l.kozlovi", 
                          ifelse(Pop == "c_ll_og", "L.l.isabellinus", 
                          ifelse(Pop == "c_ll_vl", "L.l.stroganovi", 
                          ifelse(Pop == "c_ll_ya", "L.l.wrangeli",     
                                 NA)))))))) %>% 
  mutate (., Populations =  ifelse (Pop == "c_ll_po", "Bialowieza",
                          ifelse (Pop == "c_ll_ur", "Urals",
                          ifelse (Pop == "c_ll_ki", "Kirov",
                          ifelse (Pop == "c_ll_la", "Latvia",
                          ifelse (Pop == "c_ll_no", "Norway",
                          ifelse (Pop == "c_ll_ba" | Pop == "h_ll_ba", "Balkans" ,
                          ifelse (Pop == "c_ll_cr","Carpathians",
                          ifelse (Pop == "c_ll_to", "Töv",
                          ifelse (Pop == "c_ll_tu", "Tuva",
                          ifelse (Pop == "c_ll_ka", "Khentii-Aimag", 
                          ifelse (Pop == "c_ll_og", "Ömnögovi", 
                          ifelse (Pop == "c_ll_vl", "Vladivostok", 
                          ifelse (Pop == "c_ll_ya", "Yakutia",
                          ifelse (Pop == "c_lp_sm", "Sierra Morena",
                          ifelse (Pop == "c_lp_do", "Doñana", NA)))))))))))))))) %>% 
    mutate (., color =  ifelse (Populations == "Bialowieza", brewer.pal(12,"Paired")[4], 
                      ifelse (Populations == "Urals", brewer.pal(11,"BrBG")[9], 
                      ifelse (Populations == "Balkans", brewer.pal(12,"Paired")[10], 
                      ifelse (Populations == "Carpathians", brewer.pal(12,"Paired")[9], 
                      ifelse (Populations == "Kirov", brewer.pal(12,"Paired")[1], 
                      ifelse (Populations == "Latvia", brewer.pal(12,"Paired")[3], 
                      ifelse (Populations == "Norway", brewer.pal(12,"Paired")[2], 
                      ifelse (Populations == "Töv", brewer.pal(12,"Paired")[12], 
                      ifelse (Populations == "Tuva", brewer.pal(12,"Paired")[8], 
                      ifelse (Populations == "Khentii-Aimag", brewer.pal(12,"Paired")[7], 
                      ifelse (Populations == "Ömnögovi", brewer.pal(12,"Paired")[11], 
                      ifelse (Populations == "Vladivostok", brewer.pal(12,"Paired")[5], 
                      ifelse (Populations == "Yakutia", brewer.pal(12,"Paired")[6],
                      ifelse (Populations == "Sierra Morena", brewer.pal(8, "Greys") [5],
                      ifelse (Populations == "Doñana", brewer.pal(8, "Greys") [8], NA)))))))))))))))) %>% 
  mutate (., seq = ifelse (Ind == "c_ll_vl_0112" | Ind == "c_ll_vl_0112" | Ind == "c_ll_ya_0146" | Ind == "c_ll_cr_0212" | Ind == "c_ll_ki_0090" , "MACROGEN", "CNAG"))

                           
                           
                           
cols <- c("Bialowieza"=brewer.pal(12,"Paired")[4], 
          "Urals"=brewer.pal(11,"BrBG")[9], 
          "Balkans"=brewer.pal(12,"Paired")[10], 
          "Carpathians"=brewer.pal(12,"Paired")[9], 
          "Kirov"=brewer.pal(12,"Paired")[1], 
          "Latvia"=brewer.pal(12,"Paired")[3], 
          "Norway"=brewer.pal(12,"Paired")[2], 
          "Töv"=brewer.pal(12,"Paired")[12], 
          "Tuva"=brewer.pal(12,"Paired")[8], 
          "Khentii-Aimag"=brewer.pal(12,"Paired")[7], 
          "Ömnögovi"=brewer.pal(12,"Paired")[11], 
          "Vladivostok"=brewer.pal(12,"Paired")[5], 
          "Yakutia"=brewer.pal(12,"Paired")[6],
          "Sierra Morena"=brewer.pal(8, "Greys") [5],
          "Doñana"=brewer.pal(8, "Greys") [8])

# PC$Populations <- factor (PC$Populations, levels=c("Norway", "Balkans", "Carpathians", "Bialowieza", "Latvia", "Kirov", "Urals", "Tuva", "Töv", "Khentii-Aimag", "Ömnögovi",  "Yakutia", "Vladivostok"))


title <- paste("PC",comp[1]," (",signif(eig$val[comp[1]], digits=3)*100,"%)"," / PC",comp[2]," (",signif(eig$val[comp[2]], digits=3)*100,"%)",sep="",collapse="")

x_axis = paste("PC",comp[1],sep="")
y_axis = paste("PC",comp[2],sep="")



ggplot() + 
geom_point(data=PC, aes_string(x=x_axis, y=y_axis, color="Populations"), size = 5) + 
scale_colour_manual(values = cols) + ggtitle(title) + theme_classic() 
ggsave(paste(WORKING_DIR,pop, "_pca_pop_PC1_PC2.pdf", sep=""),units="mm",  width=270, height=190)
 

  # Article
   ggplot() + 
  geom_point(data=PC, aes_string(x=x_axis, y=y_axis, colour="Populations"), size = 5) + 
  scale_colour_manual(values = cols)+ ggtitle(title) + theme_classic() +
  scale_y_reverse() +
     theme(legend.position="top")
   
 ggsave(paste(WORKING_DIR,pop, "_pca_pop_PC1_PC2_reverse_axis.pdf", sep=""),units="mm",  width=270, height=190,  useDingbats=FALSE)


  
###########################################################################################
# 3D plot
###########################################################################################


scatterplot3js(PC$PC1, PC$PC2, PC$PC3, 
                col=PC$color,  labels=PC$Ind,
               size=0.7)



```